<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Master ggtree package suite to handle tree with data.">

<title>(PART*) Part I: Tree data input, output, and manipulation – Data Integration, Manipulation and Visualization of Phylogenetic Trees</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./02_tidytree.html" rel="next">
<link href="./00_author.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-0dd2bd5de344125cf763a379ddc3eb04.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="css/style.css">
</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./01_treeio_importing_tree.html">(PART*) Part I: Tree data input, output, and manipulation</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Data Integration, Manipulation and Visualization of Phylogenetic Trees</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/YuLab-SMU/treedata-book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Data Integration, Manipulation and Visualization of Phylogenetic Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./00_author.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About the Author</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_treeio_importing_tree.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">(PART*) Part I: Tree data input, output, and manipulation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_tidytree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Manipulating Tree with Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_treeio_exporting_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Exporting tree with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_ggtree_visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part II: Tree data visualization and annotation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_ggtree_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Phylogenetic Tree Annotation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_ggtree_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Visual Exploration of Phylogenetic Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_ggtree_tree_with_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Plotting tree with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_ggtree_with_silhouette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Annotating Tree with Silhouette Images and Sub-plots</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_ggtree_tree_objects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">ggtree for Other Tree-like Objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_ggtree_exts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part III: ggtree extensions</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11_ggtree_exts-others.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Other ggtree Extensions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12_ggtree_utilities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part IV: Miscellaneous topics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13_ggtree_gallery.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Gallery of Reproducible Examples</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendix</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-faq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(APPENDIX) Appendix</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Related Tools</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-fig-tab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Figures and Tables</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#chapter1" id="toc-chapter1" class="nav-link active" data-scroll-target="#chapter1">Importing Tree with Data</a>
  <ul class="collapse">
  <li><a href="#pc" id="toc-pc" class="nav-link" data-scroll-target="#pc">Overview of Phylogenetic Tree Construction</a></li>
  <li><a href="#format" id="toc-format" class="nav-link" data-scroll-target="#format">Phylogenetic Tree Formats</a>
  <ul class="collapse">
  <li><a href="#newick-tree-format" id="toc-newick-tree-format" class="nav-link" data-scroll-target="#newick-tree-format">Newick tree format</a></li>
  <li><a href="#nexus-tree-format" id="toc-nexus-tree-format" class="nav-link" data-scroll-target="#nexus-tree-format">NEXUS tree format</a></li>
  <li><a href="#nhxtext" id="toc-nhxtext" class="nav-link" data-scroll-target="#nhxtext">New Hampshire eXtended format</a></li>
  <li><a href="#jplace-format" id="toc-jplace-format" class="nav-link" data-scroll-target="#jplace-format">Jplace format</a></li>
  <li><a href="#software-outputs" id="toc-software-outputs" class="nav-link" data-scroll-target="#software-outputs">Software outputs</a></li>
  </ul></li>
  <li><a href="#getting-tree-data-with-treeio" id="toc-getting-tree-data-with-treeio" class="nav-link" data-scroll-target="#getting-tree-data-with-treeio">Getting Tree Data with <strong>treeio</strong></a>
  <ul class="collapse">
  <li><a href="#overview-of-treeio" id="toc-overview-of-treeio" class="nav-link" data-scroll-target="#overview-of-treeio">Overview of <strong>treeio</strong></a></li>
  <li><a href="#function-demonstration" id="toc-function-demonstration" class="nav-link" data-scroll-target="#function-demonstration">Function demonstration</a></li>
  <li><a href="#as-treedata" id="toc-as-treedata" class="nav-link" data-scroll-target="#as-treedata">Converting other tree-like objects to <code>phylo</code> or <code>treedata</code> objects</a></li>
  <li><a href="#get-treedata-data" id="toc-get-treedata-data" class="nav-link" data-scroll-target="#get-treedata-data">Getting information from <code>treedata</code> object</a></li>
  </ul></li>
  <li><a href="#summary1" id="toc-summary1" class="nav-link" data-scroll-target="#summary1">Summary</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/YuLab-SMU/treedata-book/edit/main/01_treeio_importing_tree.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/blob/main/01_treeio_importing_tree.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">(PART*) Part I: Tree data input, output, and manipulation</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div style="page-break-after: always;"></div>
<section id="chapter1" class="level1">
<h1>Importing Tree with Data</h1>
<section id="pc" class="level2">
<h2 class="anchored" data-anchor-id="pc">Overview of Phylogenetic Tree Construction</h2>
<p>Phylogenetic trees are used to describe genealogical relationships among a group of organisms, which can be constructed based on the genetic sequences of the organisms. A rooted phylogenetic tree represents a model of evolutionary history depicted by ancestor-descendant relationships between tree nodes and clustering of ‘sister’ or ‘cousin’ organisms at a different level of relatedness, as illustrated in Figure @ref(fig:phylogeny). In infectious disease research, phylogenetic trees are usually built from the pathogens’ gene or genome sequences to show which pathogen sample is genetically closer to another sample, providing insights into the underlying unobserved epidemiologic linkage and the potential source of an outbreak.</p>
<p>(ref:phylogenyscap) Components of a phylogenetic tree.</p>
<p>(ref:phylogenycap) <strong>Components of a phylogenetic tree.</strong> External nodes (green circles), also called ‘tips’, represent actual organisms sampled and sequenced (<em>e.g.</em>, viruses in infectious disease research). They are the ‘taxa’ in the terminology of evolutionary biology. The internal nodes (blue circles) represent hypothetical ancestors for the tips. The root (red circle) is the common ancestor of all species in the tree. The horizontal lines are branches and represent evolutionary changes (gray number) measured in a unit of time or genetic divergence. The bar at the bottom provides the scale of these branch lengths.</p>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.
ℹ The deprecated feature was likely used in the ggtree package.
  Please report the issue at &lt;https://github.com/YuLab-SMU/ggtree/issues&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="01_treeio_importing_tree_files/figure-html/phylogeny-1.png" class="img-fluid figure-img" width="576"></p>
<figcaption>(ref:phylogenycap)</figcaption>
</figure>
</div>
</div>
</div>
<p>A phylogenetic tree can be constructed from genetic sequences using distance-based methods or character-based methods. Distance-based methods, including the unweighted pair group method with arithmetic means (UPGMA) and the Neighbor-joining (NJ), are based on the matrix of pairwise genetic distances calculated between sequences. The character-based methods, including maximum parsimony (MP) <span class="citation" data-cites="fitch_toward_1971">(<a href="#ref-fitch_toward_1971" role="doc-biblioref">Fitch, 1971</a>)</span>, maximum likelihood (ML) <span class="citation" data-cites="felsenstein_evolutionary_1981">(<a href="#ref-felsenstein_evolutionary_1981" role="doc-biblioref">Felsenstein, 1981</a>)</span>, and Bayesian Markov Chain Monte Carlo (BMCMC) method <span class="citation" data-cites="rannala_probability_1996">(<a href="#ref-rannala_probability_1996" role="doc-biblioref">Rannala &amp; Yang, 1996</a>)</span>, are based on a mathematical model that describes the evolution of genetic characters and searches for the best phylogenetic tree according to their optimality criteria.</p>
<p>The MP method assumes that the evolutionary change is rare and minimizes the amount of character-state changes (<em>e.g.</em>, number of DNA substitutions). The criterion is similar to Occam’s razor, that the simplest hypothesis that can explain the data is the best hypothesis. Unweighted parsimony assumes mutations across different characters (nucleotides or amino acids) are equally likely, while the weighted method assumes the unequal likelihood of mutations (<em>e.g.</em>, the third codon position is more liable than other codon positions; and the transition mutations have a higher frequency than transversion). The concept of the MP method is straightforward and intuitive, which is a probable reason for its popularity amongst biologists who care more about the research question rather than the computational details of the analysis. However, this method has several disadvantages, in particular, the tree inference can be biased by the well-known systematic error called long-branch attraction (LBA) that incorrectly infer distantly related lineages as closely related <span class="citation" data-cites="felsenstein_cases_1978">(<a href="#ref-felsenstein_cases_1978" role="doc-biblioref">Felsenstein, 1978</a>)</span>. This is because the MP method poorly takes into consideration of many sequence evolution factors (<em>e.g.</em>, reversals and convergence) that are hardly observable from the existing genetic data.</p>
<p>The maximum likelihood (ML) method and Bayesian Markov Chain Monte Carlo (BMCMC) method are the two most commonly used methods in phylogenetic tree construction and are most often used in scientific publications. ML and BMCMC methods require a substitution model of sequence evolution. Different sequence data have different substitution models to formulate the evolutionary process of DNA, codon and amino acid. There are several models for nucleotide substitution, including JC69, K2P, F81, HKY, and GTR <span class="citation" data-cites="arenas_trends_2015">(<a href="#ref-arenas_trends_2015" role="doc-biblioref">Arenas, 2015</a>)</span>. These models can be used in conjunction with the rate variation across sites (denoted as +<span class="math inline">\(\Gamma\)</span>)) <span class="citation" data-cites="yang_maximum_1994">(<a href="#ref-yang_maximum_1994" role="doc-biblioref">Yang, 1994</a>)</span> and the proportion of invariable sites (denoted as +I) <span class="citation" data-cites="shoemaker_evidence_1989">(<a href="#ref-shoemaker_evidence_1989" role="doc-biblioref">Shoemaker &amp; Fitch, 1989</a>)</span>. Previous research <span class="citation" data-cites="lemmon_importance_2004">(<a href="#ref-lemmon_importance_2004" role="doc-biblioref">Lemmon &amp; Moriarty, 2004</a>)</span> had suggested that misspecification of substitution model might bias phylogenetic inference. Procedural testing for the best-fit substitution model is recommended.</p>
<p>The optimal criterion of the ML method is to find the tree that maximizes the likelihood given the sequence data. The procedure of the ML method is simple: calculating the likelihood of a tree and optimizing its topology and branches (and the substitution model parameters, if not fixed) until the best tree is found. Heuristic search, such as those implemented in <a href="http://www.atgc-montpellier.fr/phyml/"><strong>PhyML</strong></a> and <a href="http://evomics.org/learning/phylogenetics/raxml/"><strong>RAxML</strong></a>, is often used to find the best tree based on the likelihood criterion. The Bayesian method finds the tree that maximizes posterior probability by sampling trees through MCMC based on the given substitution model. One of the advantages of BMCMC is that parameter variance and tree topological uncertainty, included by the posterior clade probability, can be naturally and conveniently obtained from the sampling trees in the MCMC process. Moreover, the influence of topological uncertainty on other parameter estimates is also naturally integrated into the BMCMC phylogenetic framework.</p>
<p>In a simple phylogenetic tree, data associated with the tree branches/nodes could be the branch lengths (indicating genetic or time divergence) and lineage supports such as bootstrap values estimated from bootstrapping procedure or posterior clade probability summarized from the sampled trees in the BMCMC analysis.</p>
</section>
<section id="format" class="level2">
<h2 class="anchored" data-anchor-id="format">Phylogenetic Tree Formats</h2>
<p>There are several file formats designed to store phylogenetic trees and the data associated with the nodes and branches. The three commonly used formats are Newick<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>, NEXUS <span class="citation" data-cites="maddison_nexus:_1997">(<a href="#ref-maddison_nexus:_1997" role="doc-biblioref">Maddison et al., 1997</a>)</span>, and Phylip <span class="citation" data-cites="felsenstein_phylip_1989">(<a href="#ref-felsenstein_phylip_1989" role="doc-biblioref">Felsenstein, 1989</a>)</span>. Some formats (<em>e.g.</em>, NHX) are extended from the Newick format. Newick and NEXUS formats are supported as input by most of the software in evolutionary biology, while some of the software tools output newer standard files (<em>e.g.</em>, <a href="http://beast2.org/"><strong>BEAST</strong></a> and <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a>) by introducing new rules/data blocks for storing evolutionary inferences. In the other cases (<em>e.g.</em>, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a> and <a href="http://loco.biosci.arizona.edu/r8s/"><strong>r8s</strong></a>), output log files are only recognized by their own single software.</p>
<section id="newick-tree-format" class="level3">
<h3 class="anchored" data-anchor-id="newick-tree-format">Newick tree format</h3>
<p>The Newick tree format is the standard for representing trees in computer-readable form.</p>
<p>(ref:randomTreescap) A sample tree for demonstrating Newick text to encode tree structure.</p>
<p>(ref:randomTreecap) <strong>A sample tree for demonstrating Newick text to encode tree structure.</strong> Tips were aligned to the right-hand side and branch lengths were labeled on the middle of each branch.</p>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="01_treeio_importing_tree_files/figure-html/randomTree-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>(ref:randomTreecap)</figcaption>
</figure>
</div>
</div>
</div>
<p>The rooted tree shown in Figure @ref(fig:randomTree) can be represented by the following sequence of characters as a Newick tree text.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>((t2:0.04,t1:0.34):0.89,(t5:0.37,(t4:0.03,t3:0.67):0.9):0.59); </code></pre>
</div>
</div>
<p>The tree text ends with a semicolon. Internal nodes are represented by a pair of matched parentheses. Between the parentheses are descendant nodes of that node. For instance (t2:0.04,t1:0.34) represents the parent node of t2 and t1 that are the immediate descendants. Sibling nodes are separated by a comma and tips are represented by their names. A branch length (from the parent node to child node) is represented by a real number after the child node and is preceded by a colon. Singular data (<em>e.g.</em>, bootstrap values) associated with internal nodes or branches may be encoded as node labels and represented by the simple text/numbers before the colon.</p>
<p>Newick tree format was developed by Meacham in 1984 for the Phylogeny Inference Package or <a href="https://evolution.genetics.washington.edu/phylip.html"><strong>PHYLIP</strong></a> <span class="citation" data-cites="retief_phylogenetic_2000">(<a href="#ref-retief_phylogenetic_2000" role="doc-biblioref">Retief, 2000</a>)</span> package. Newick format is now the most widely used tree format and used by <a href="https://evolution.genetics.washington.edu/phylip.html"><strong>PHYLIP</strong></a>, <a href="https://paup.phylosolutions.com/"><strong>PAUP*</strong></a> <span class="citation" data-cites="wilgenbusch_inferring_2003">(<a href="#ref-wilgenbusch_inferring_2003" role="doc-biblioref">Wilgenbusch &amp; Swofford, 2003</a>)</span>, <em>TREE-PUZZLE</em> <span class="citation" data-cites="schmidt_tree-puzzle:_2002">(<a href="#ref-schmidt_tree-puzzle:_2002" role="doc-biblioref">Schmidt et al., 2002</a>)</span>, <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a>, and many other applications. Phylip tree format contains Phylip multiple sequence alignment (MSA) with a corresponding Newick tree text that was built based on the MSA sequences in the same file.</p>
</section>
<section id="nexus-tree-format" class="level3">
<h3 class="anchored" data-anchor-id="nexus-tree-format">NEXUS tree format</h3>
<p>The NEXUS format <span class="citation" data-cites="maddison_nexus:_1997">(<a href="#ref-maddison_nexus:_1997" role="doc-biblioref">Maddison et al., 1997</a>)</span> incorporates Newick tree text with related information organized into separated units known as <strong>blocks</strong>. A NEXUS block has the following structure:</p>
<pre><code>#NEXUS
...
BEGIN characters;
...
END;</code></pre>
<p>For example, the above example tree can be saved as the following NEXUS format:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>#NEXUS
[R-package APE, Sun Jan 11 02:24:30 2026]

BEGIN TAXA;
    DIMENSIONS NTAX = 5;
    TAXLABELS
        t2
        t1
        t5
        t4
        t3
    ;
END;
BEGIN TREES;
    TRANSLATE
        1   t2,
        2   t1,
        3   t5,
        4   t4,
        5   t3
    ;
    TREE * UNTITLED = [&amp;R]
((1:0.04,2:0.34):0.89,(3:0.37,(4:0.03,5:0.67):0.9):0.59);
END;</code></pre>
</div>
</div>
<p>Comments can be placed using square brackets. Some blocks can be recognized by most of the programs including <code>TAXA</code> (contains information of taxa), <code>DATA</code> (contains data matrix, <em>e.g.</em>, sequence alignment), and <code>TREE</code> (contains a phylogenetic tree, <em>i.e.</em>, Newick tree text). Notably, blocks can be very diverse and some of them are only recognized by one particular program. For example NEXUS file exported by <a href="https://paup.phylosolutions.com/"><strong>PAUP*</strong></a> has a <em>paup</em> block that contains <em>PAUP*</em> commands, whereas <a href="http://beast.community/figtree"><strong>FigTree</strong></a> exports the NEXUS file with a <em>figtree</em> block that contains visualization settings. NEXUS organizes different types of data into different blocks, whereas programs that support reading NEXUS can parse some blocks they recognized and ignore those they could not. This is a good mechanism to allow different programs to use the same format without crashing when unsupported types of data are present. Notably, most of the programs only support parsing <code>TAXA</code>, <code>DATA</code>, and <code>TREE</code> blocks; therefore, a program/platform that could generically read all kinds of data blocks from the NEXUS would be useful for phylogenetic data integration.</p>
<p>The <code>DATA</code> block is widely used to store sequence alignment. For this purpose, the user can store tree and sequence data in Phylip format which are essentially Phylip multiple sequence alignment and Newick tree text, respectively. It is used in <a href="https://evolution.genetics.washington.edu/phylip.html"><strong>PHYLIP</strong></a>.</p>
<!-- 

~~and it's also a drawback of NEXUS for most of the software can only parse the tree strucuture without support of the associated data. NEXUS is good for storing specific information for particular program, it can be used for storing annotated trees.~~

It is very difficult to share tree annotation data among different programs unless the data is a matrix which can be stored in `DATA` block. The `DATA` block is widely used to store sequence alignment. For this purpose, user can store tree in Phylip format which was used in [**PHYLIP**](https://evolution.genetics.washington.edu/phylip.html) package and can contains Phylip multiple sequence alignment with Newick tree text.

-->
</section>
<section id="nhxtext" class="level3">
<h3 class="anchored" data-anchor-id="nhxtext">New Hampshire eXtended format</h3>
<!-- Although NEXUS supports additional information in separated blocks, they are mostly not recognized by different software and can't share information among different programs. In order to store annotated tree,-->
<p>Newick, NEXUS, and phylip are mainly designed to store phylogenetic trees and basic singular data associated with internal nodes or branches. In addition to the singular data annotation at branches and nodes (mentioned above), New Hampshire eXtended (NHX) format, which is based on Newick (also called New Hampshire), was developed to introduce tags to associate multiple data fields with the tree nodes (both internal nodes and tips). Tags are placed after branch length and must be wrapped between <code>[&amp;&amp;NHX</code> and <code>]</code> which makes it possible to be compatible with NEXUS format as it defined characters between <code>[</code> and <code>]</code> as comments. NHX is also the output format of <a href="http://pbil.univ-lyon1.fr/software/phyldog/"><strong>PHYLDOG</strong></a> <span class="citation" data-cites="boussau_genome-scale_2013">(<a href="#ref-boussau_genome-scale_2013" role="doc-biblioref">Boussau et al., 2013</a>)</span> and <a href="http://revbayes.github.io/intro.html"><strong>RevBayes</strong></a> <span class="citation" data-cites="hohna_revbayes:_2016">(<a href="#ref-hohna_revbayes:_2016" role="doc-biblioref">Höhna et al., 2016</a>)</span>. A Tree Viewer (<a href="http://phylogeny.lirmm.fr/phylo_cgi/one_task.cgi?task_type=atv"><strong>ATV</strong></a>) <span class="citation" data-cites="zmasek_atv:_2001">(<a href="#ref-zmasek_atv:_2001" role="doc-biblioref">Zmasek &amp; Eddy, 2001</a>)</span> is a java tool that supports displaying annotation data stored in NHX format, but this package is no longer maintained.</p>
<p>Here is a sample tree from NHX definition document<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>(((ADH2:0.1[&amp;&amp;NHX:S=human], ADH1:0.11[&amp;&amp;NHX:S=human]):0.05
[&amp;&amp;NHX:S=primates:D=Y:B=100],ADHY:0.1[&amp;&amp;NHX:S=nematode],
ADHX:0.12[&amp;&amp;NHX:S=insect]):0.1[&amp;&amp;NHX:S=metazoa:D=N],(ADH4:0.09
[&amp;&amp;NHX:S=yeast],ADH3:0.13[&amp;&amp;NHX:S=yeast],ADH2:0.12[&amp;&amp;NHX:S=yeast],
ADH1:0.11[&amp;&amp;NHX:S=yeast]):0.1[&amp;&amp;NHX:S=Fungi])[&amp;&amp;NHX:D=N];</code></pre>
</div>
</div>
</section>
<section id="jplace-format" class="level3">
<h3 class="anchored" data-anchor-id="jplace-format">Jplace format</h3>
<p>To store the Next Generation Sequencing (NGS) short reads mapped onto a phylogenetic tree (for metagenomic classification), Matsen <span class="citation" data-cites="matsen_format_2012">(<a href="#ref-matsen_format_2012" role="doc-biblioref">Matsen et al., 2012</a>)</span> proposed jplace format for such phylogenetic placements. Jplace format is based on JSON and contains four keys: <code>tree</code>, <code>fields</code>, <code>placements</code>, <code>metadata</code>, and <code>version</code>. The <code>tree</code> value contains tree text extended from Newick tree format by putting the edge label in brackets (if available) after branch length and putting the edge number in curly braces after the edge label. The <code>fields</code> value contains header information of placement data. The value of <code>placements</code> is a list of <code>pqueries</code>. Each <code>pquery</code> contains two keys: <code>p</code> for placements and <code>n</code> for name or <code>nm</code> for names with multiplicity. The value of <code>p</code> is a list of placement for <code>pqueries</code>.</p>
<p>Here is a jplace sample file:</p>
<pre><code>{
    "tree": "(((((((A:4{1},B:4{2}):6{3},C:5{4}):8{5},D:6{6}):
    3{7},E:21{8}):10{9},((F:4{10},G:12{11}):14{12},H:8{13}):
    13{14}):13{15},((I:5{16},J:2{17}):30{18},(K:11{19},
    L:11{20}):2{21}):17{22}):4{23},M:56{24});",
    "placements": [
    {"p":[24, -61371.300778, 0.333344, 0.000003, 0.003887], 
     "n":["AA"]
    },
    {"p":[[1, -61312.210786, 0.333335, 0.000001, 0.000003],
          [2, -61322.210823, 0.333322, 0.000003, 0.000003],
          [3, -61352.210823, 0.333322, 0.000961, 0.000003]],
     "n":["BB"]
    },
    {"p":[[8, -61312.229128, 0.200011, 0.000001, 0.000003],
          [9, -61322.229179, 0.200000, 0.000003, 0.000003],
          [10, -61342.229223, 0.199992, 0.000003, 0.000003]], 
    "n":["CC"]
    }
    ],
    "metadata": {"info": "a jplace sample file"},
    "version" : 2,
    "fields": ["edge_num", "likelihood", "like_weight_ratio", 
    "distal_length", "pendant_length"
    ]
}</code></pre>
<p>Jplace is the output format of <a href="http://matsen.fhcrc.org/pplacer/"><strong>PPLACER</strong></a> <span class="citation" data-cites="matsen_pplacer_2010">(<a href="#ref-matsen_pplacer_2010" role="doc-biblioref">Matsen et al., 2010</a>)</span> and Evolutionary Placement Algorithm (<a href="http://sco.h-its.org/exelixis/web/software/epa/index.html"><strong>EPA</strong></a>) <span class="citation" data-cites="berger_performance_2011">(<a href="#ref-berger_performance_2011" role="doc-biblioref">Berger et al., 2011</a>)</span>. But these two programs do not contain tools to visualize placement results. <a href="http://matsen.fhcrc.org/pplacer/"><strong>PPLACER</strong></a> provides <code>placeviz</code> to convert jplace file to phyloXML or Newick formats which can be visualized by <a href="https://sites.google.com/site/cmzmasek/home/software/archaeopteryx"><strong>Archaeopteryx</strong></a>.</p>
</section>
<section id="software-outputs" class="level3">
<h3 class="anchored" data-anchor-id="software-outputs">Software outputs</h3>
<p><a href="http://evomics.org/learning/phylogenetics/raxml/"><strong>RAxML</strong></a> <span class="citation" data-cites="stamatakis_raxml_2014">(<a href="#ref-stamatakis_raxml_2014" role="doc-biblioref">Stamatakis, 2014</a>)</span> can output Newick format by storing the bootstrap values as internal node labels. Another way that <a href="http://evomics.org/learning/phylogenetics/raxml/"><strong>RAxML</strong></a> supports is to place bootstrap value inside square brackets and after branch length. This could not be supported by most of the software that supports Newick format where square brackets will be ignored.</p>
<p><a href="http://beast2.org/"><strong>BEAST</strong></a> <span class="citation" data-cites="bouckaert_beast_2014">(<a href="#ref-bouckaert_beast_2014" role="doc-biblioref">Bouckaert et al., 2014</a>)</span> output is based on NEXUS, and it also introduces square brackets in the tree block to store evolutionary evidence inferred by <a href="http://beast2.org/"><strong>BEAST</strong></a>. Inside brackets, curly braces may also be incorporated if feature values have a length of more than 1 (<em>e.g.</em>, Highest Probability Density (HPD) or range of substitution rate). These brackets are placed between node and branch length (<em>i.e.</em>, after label if exists and before the colon). The bracket is not defined in Newick format and is a reserved character for NEXUS comment. So this information will be ignored for standard NEXUS parsers.</p>
<p>Here is a sample <code>TREE</code> block of the <a href="http://beast2.org/"><strong>BEAST</strong></a> output:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>TREE * TREE1 = [&amp;R] (((11[&amp;length=9.47]:9.39,14[&amp;length=6.47]:6.39)
[&amp;length=25.72]:25.44,4[&amp;length=9.14]:8.82)[&amp;length=3.01]:3.1,
(12[&amp;length=0.62]:0.57,(10[&amp;length=1.6]:1.56,(7[&amp;length=5.21]:5.19,
((((2[&amp;length=3.3]:3.26,(1[&amp;length=1.34]:1.32,(6[&amp;length=0.85]:0.83,
13[&amp;length=0.85]:0.83)[&amp;length=2.5]:2.49)[&amp;length=0.97]:0.94)
[&amp;length=0.5]:0.5,9[&amp;length=1.76]:1.76)[&amp;length=2.41]:2.36,
8[&amp;length=2.19]:2.11)[&amp;length=0.27]:0.24,(3[&amp;length=3.33]:3.31,
(15[&amp;length=5.29]:5.27,5[&amp;length=3.29]:3.27)[&amp;length=1.04]:1.04)
[&amp;length=1.98]:2.04)[&amp;length=2.83]:2.84)[&amp;length=5.39]:5.37)
[&amp;length=2.02]:2)[&amp;length=4.35]:4.36)[&amp;length=0];</code></pre>
</div>
</div>
<p><a href="http://beast2.org/"><strong>BEAST</strong></a> output can contain many different evolutionary inferences, depending on the analysis models defined in <em>BEAUTi</em> for running. For example in molecular clock analysis, it contains <code>rate</code>, <code>length</code>, <code>height</code>, <code>posterior</code> and corresponding HPD and range for uncertainty estimation. <code>Rate</code> is the estimated evolutionary rate of the branch. <code>Length</code> is the length of the branch in years. <code>Height</code> is the time from node to root, while <code>posterior</code> is the Bayesian clade credibility value. The above example is the output tree of a molecular clock analysis and should contain these inferences. To save space, only the <code>length</code> estimation was shown above. Besides, Molecular Evolutionary Genetics Analysis (<a href="https://www.megasoftware.net/"><strong>MEGA</strong></a>) <span class="citation" data-cites="kumar_mega7_2016">(<a href="#ref-kumar_mega7_2016" role="doc-biblioref">Kumar et al., 2016</a>)</span> also supports exporting trees in <a href="http://beast2.org/"><strong>BEAST</strong></a> compatible Nexus format (see <a href="#mega">session 1.3.2</a>).</p>
<p><em><a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a></em> <span class="citation" data-cites="huelsenbeck_mrbayes_2001">(<a href="#ref-huelsenbeck_mrbayes_2001" role="doc-biblioref">Huelsenbeck &amp; Ronquist, 2001</a>)</span> is a program that uses the Markov Chain Monte Carlo method to sample from the posterior probability distributions. Its output file annotates nodes and branches separately by two sets of square brackets. For example below, posterior clade probabilities for the node and branch length estimates for the branch:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>  tree con_all_compat = [&amp;U] (8[&amp;prob=1.0]:2.94e-1[&amp;length_mean=2.9e-1],
10[&amp;prob=1.0]:2.25e-1[&amp;length_mean=2.2e-1],((((1[&amp;prob=1.0]:1.43e-1
[&amp;length_mean=1.4e-1],2[&amp;prob=1.0]:1.92e-1[&amp;length_mean=1.9e-1])[&amp;prob=1.0]:
1.24e-1[&amp;length_mean=1.2e-1],9[&amp;prob=1.0]:2.27e-1[&amp;length_mean=2.2e-1])
[&amp;prob=1.0]:1.72e-1[&amp;length_mean=1.7e-1],12[&amp;prob=1.0]:5.11e-1
[&amp;length_mean=5.1e-1])[&amp;prob=1.0]:1.76e-1[&amp;length_mean=1.7e-1],
(((3[&amp;prob=1.0]:5.46e-2[&amp;length_mean=5.4e-2],(6[&amp;prob=1.0]:1.03e-2
[&amp;length_mean=1.0e-2],7[&amp;prob=1.0]:7.13e-3[&amp;length_mean=7.2e-3])[&amp;prob=1.0]:
6.93e-2[&amp;length_mean=6.9e-2])[&amp;prob=1.0]:6.03e-2[&amp;length_mean=6.0e-2],
(4[&amp;prob=1.0]:6.27e-2[&amp;length_mean=6.2e-2],5[&amp;prob=1.0]:6.31e-2
[&amp;length_mean=6.3e-2])[&amp;prob=1.0]:6.07e-2[&amp;length_mean=6.0e-2])[&amp;prob=1.0]:,
1.80e-1[&amp;length_mean=1.8e-1]11[&amp;prob=1.0]:2.37e-1[&amp;length_mean=2.3e-1])
[&amp;prob=1.0]:4.05e-1[&amp;length_mean=4.0e-1])[&amp;prob=1.0]:1.16e+000
[&amp;length_mean=1.162699558201079e+000])[&amp;prob=1.0][&amp;length_mean=0];</code></pre>
</div>
</div>
<p>To save space, most of the inferences were removed and only contains <code>prob</code> for clade probability and <code>length_mean</code> for mean value of branch length. The full version of this file also contains <code>prob_stddev</code>, <code>prob_range</code>, <code>prob(percent)</code>, <code>prob+-sd</code> for probability inferences and <code>length_median</code>, <code>length_95%_HPD</code> for every branch.</p>
<p>The <a href="http://beast2.org/"><strong>BEAST</strong></a> and <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a> outputs are expected to be parsed without inferences (dropped as comments) by software that supports NEXUS. <a href="http://beast.community/figtree"><strong>FigTree</strong></a> supports parsing <a href="http://beast2.org/"><strong>BEAST</strong></a>, and <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a> outputs with inferences that can be used to display or annotate on the tree. But from there, extracting these data for further analysis is still challenging.</p>
<p><a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a> <span class="citation" data-cites="pond_hyphy:_2005">(<a href="#ref-pond_hyphy:_2005" role="doc-biblioref">Pond et al., 2005</a>)</span> could do a number of phylogenetic analyses, including ancestral sequence reconstruction. For ancestral sequence reconstruction, these sequences and the Newick tree text are stored in NEXUS format as the major analysis output. It did not completely follow the NEXUS definition and only put the ancestral node labels in <code>TAXA</code> instead of the external node label. The <code>MATRIX</code> block contains sequence alignment of ancestral nodes which cannot be referred back to the tree stored in the <code>TREES</code> block since it does not contain node labels. Here is the sample output (to save space, only the first 72bp of alignment are shown):</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>#NEXUS

[
Generated by HYPHY 2.0020110620beta(MP) for MacOS(Universal Binary) 
    on Tue Dec 23 13:52:34 2014

]

BEGIN TAXA;
    DIMENSIONS NTAX = 13;
    TAXLABELS
        'Node1' 'Node2' 'Node3' 'Node4' 'Node5' 'Node12' 'Node13' 'Node15'
            'Node18' 'Node20' 'Node22' 'Node24' 'Node26' ;
END;

BEGIN CHARACTERS;
    DIMENSIONS NCHAR = 2148;
    FORMAT
        DATATYPE = DNA

        GAP=-
        MISSING=?
        NOLABELS
    ;

MATRIX
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATTGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
 ATGGAAGACTTTGTGCGACAGTGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATAT
END;

BEGIN TREES;
    TREE tree = (K,N,(D,(L,(J,(G,((C,(E,O)),(H,(I,(B,(A,(F,M)))))))))));
END;</code></pre>
</div>
</div>
<p>There are other applications that output rich information text that also contains phylogenetic trees with associated data. For example <a href="http://loco.biosci.arizona.edu/r8s/"><strong>r8s</strong></a> <span class="citation" data-cites="sanderson_r8s:_2003">(<a href="#ref-sanderson_r8s:_2003" role="doc-biblioref">Sanderson, 2003</a>)</span> output three trees in its log file, namely <code>TREE</code>, <code>RATE</code>, and <code>PHYLO</code> for branches scaled by time, substitution rate, and absolute substitutions, respectively.</p>
<p>Phylogenetic Analysis by Maximum Likelihood (<a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a>) <span class="citation" data-cites="yang_paml_2007">(<a href="#ref-yang_paml_2007" role="doc-biblioref">Yang, 2007</a>)</span> is a package of programs for phylogenetic analyses of DNA or protein sequences. Two main programs, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> and <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>, implement a variety of models. <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> estimates tree topology, branch lengths, and substitution parameters using a number of nucleotide substitution models available, including JC69, K80, F81, F84, HKY85, T92, TN93, and GTR. <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> estimates synonymous and non-synonymous substitution rates, likelihood ratio test of positive selection under codon substitution models <span class="citation" data-cites="goldman_codon-based_1994">(<a href="#ref-goldman_codon-based_1994" role="doc-biblioref">Goldman &amp; Yang, 1994</a>)</span>.</p>
<p><a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> outputs <em>mlb</em> file that contains input sequence (taxa) alignment and a phylogenetic tree with branch length as well as substitution model and other parameters estimated. The supplementary result file, <em>rst</em>, contains sequence alignment (with ancestral sequence if it performs reconstruction of ancestral sequences) and Naive Empirical Bayes (NBE) probabilities that each site in the alignment evolved. <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> outputs <em>mlc</em> file that contains tree structure and estimation of synonymous and non-synonymous substitution rates. <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> also outputs a supplementary result file, <em>rst</em>, that is similar to <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> except that site is defined as a codon instead of a nucleotide. Parsing these files can be tedious and would oftentimes need a number of post-processing steps and require expertise in programming (e.g., with Python<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> or Perl<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>).</p>
<p>Introducing square brackets is quite common for storing extra information, including <em>RAxML</em> to store bootstrap value, NHX format for annotation, jplace for edge label, and <a href="http://beast2.org/"><strong>BEAST</strong></a> for evolutionary estimation, <em>etc</em>. But the positions to place square brackets are not consistent in different software and the contents employ different rules for storing associated data, these properties make it difficult to parse associated data. For most of the software, they will just ignore square brackets and only parse the tree structure if the file is compatible. Some of them contain invalid characters (e.g., curly braces in <code>tree</code> field of jplace format), and even the tree structure can’t be parsed by standard parsers.</p>
<p>It is difficult to extract useful phylogeny/taxon-related information from the different analysis outputs produced by various evolutionary inference software, for displaying on the same phylogenetic tree and for further analysis. <a href="http://beast.community/figtree"><strong>FigTree</strong></a> supports <a href="http://beast2.org/"><strong>BEAST</strong></a> output, but not for most of the other software outputs that contain evolutionary inferences or associated data. For those output-rich text files (e.g., <a href="http://loco.biosci.arizona.edu/r8s/"><strong>r8s</strong></a>, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a>, <em>etc.</em>), the tree structure cannot be parsed by any tree viewing software and users need expertise to manually extract the phylogenetic tree and other useful result data from the output file. However, such manual operation is slow and error-prone.</p>
<!-- can't be supported and hard to display on the tree. A very common solution is to modified node labels by putting associated data to node labels and dispaly them as node lables. This is quite restricted and error prone.-->
<p>It was not easy to retrieve phylogenetic trees with evolutionary data from different analysis outputs of commonly used software in the field. Some of them (<em>e.g.</em>, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a> output and jplace file) without software or programming library to support parsing the file, while others (<em>e.g.</em>, <a href="http://beast2.org/"><strong>BEAST</strong></a> and <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a> output) can be parsed without evolutionary inferences as they are stored in square brackets that will be omitted as a comment by most of the software. Although <a href="http://beast.community/figtree"><strong>FigTree</strong></a> support visualizing evolutionary statistics inferred by <a href="http://beast2.org/"><strong>BEAST</strong></a> and <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a>, extracting these data for further analysis is not supported. Different software packages implement different algorithms for different analyses (<em>e.g.</em>, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a> for <em>d<sub>N</sub>/d<sub>S</sub></em>, <a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a> for ancestral sequences, and <a href="http://beast2.org/"><strong>BEAST</strong></a> for skyline analysis). Therefore, in encountering the genomic sequence data, there is a desired need to efficiently and flexibly integrate different analysis inference results for comprehensive understanding, comparison, and further analysis. This motivated us to develop the programming library to parse the phylogenetic trees and data from various sources.</p>
<!--
to complete each other or compare similar analyses from different software (*e.g.*, clade probability inferred from `BEAST` and `MrBayes`). Without programming library to parse these software outputs make it difficult to integrate evolutionary statistics for comparison and further analysis.
-->
</section>
</section>
<section id="getting-tree-data-with-treeio" class="level2">
<h2 class="anchored" data-anchor-id="getting-tree-data-with-treeio">Getting Tree Data with <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a></h2>
<p>Phylogenetic trees are commonly used to present evolutionary relationships of species. Information associated with taxon species/strains may be further analyzed in the context of the evolutionary history depicted by the phylogenetic tree. For example, the host information of the influenza virus strains in the tree could be studied to understand the host range of a virus lineage. Moreover, such meta-data (<em>e.g.</em>, isolation host, time, location, <em>etc.</em>) directly associated with taxon strains are also often subjected to further evolutionary or comparative phylogenetic models and analyses, to infer their dynamics associated with the evolutionary or transmission processes of the virus. All these meta-data or other phenotypic or experimental data are stored either as the annotation data associated with the nodes or branches and are often produced in inconsistent format by different analysis programs.</p>
<p>Getting trees into R is still limited. Newick and Nexus can be imported by several packages, including <a href="https://CRAN.R-project.org/package=ape"><strong>ape</strong></a>, <a href="https://CRAN.R-project.org/package=phylobase"><strong>phylobase</strong></a>. NeXML format can be parsed by <a href="https://CRAN.R-project.org/package=RNeXML"><strong>RNeXML</strong></a>. However, analysis results from widely used software packages in this field are not well supported. SIMMAP output can be parsed by <a href="https://CRAN.R-project.org/package=phyext2"><strong>phyext2</strong></a> and <a href="https://CRAN.R-project.org/package=phytools"><strong>phytools</strong></a>. Although <a href="http://www.christophheibl.de/Rpackages.html"><strong>PHYLOCH</strong></a> can import <a href="http://beast2.org/"><strong>BEAST</strong></a> and <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a> output, only internal node attributes were parsed and tip attributes were ignored<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>. Many other software outputs are mainly required programming expertise to import the tree with associated data. Linking external data, including experimental and clinical data, to phylogeny is another obstacle for evolution biologists.</p>
<p>To fill the gap that most of the tree formats or software outputs cannot be parsed within the same software/platform, an R package <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> was developed for parsing various tree file formats and outputs from common evolutionary analysis software. The <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package is developed with the R programming language <span class="citation" data-cites="rstats">(<a href="#ref-rstats" role="doc-biblioref">R Core Team, 2016</a>)</span>. Not only the tree structure can be parsed, but also the associated data and evolutionary inferences, including NHX annotation, clock rate inferences (from <a href="http://beast2.org/"><strong>BEAST</strong></a> or <a href="http://loco.biosci.arizona.edu/r8s/"><strong>r8s</strong></a> <span class="citation" data-cites="sanderson_r8s:_2003">(<a href="#ref-sanderson_r8s:_2003" role="doc-biblioref">Sanderson, 2003</a>)</span> programs), synonymous and non-synonymous substitutions (from <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>), and ancestral sequence construction (from <a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a>, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> or <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>), <em>etc.</em>. Currently, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> is able to read the following file formats: Newick, Nexus, New Hampshire eXtended format (NHX), jplace and Phylip as well as the data outputs from the following analysis programs: <a href="https://github.com/smirarab/ASTRAL"><strong>ASTRAL</strong></a>, <a href="http://beast2.org/"><strong>BEAST</strong></a>, <a href="http://sco.h-its.org/exelixis/web/software/epa/index.html"><strong>EPA</strong></a>, <a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a>, <a href="https://www.megasoftware.net/"><strong>MEGA</strong></a>, <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a>, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a>, <a href="http://pbil.univ-lyon1.fr/software/phyldog/"><strong>PHYLDOG</strong></a>, <a href="http://matsen.fhcrc.org/pplacer/"><strong>PPLACER</strong></a>, <a href="http://loco.biosci.arizona.edu/r8s/"><strong>r8s</strong></a>, <a href="http://evomics.org/learning/phylogenetics/raxml/"><strong>RAxML</strong></a> and <a href="http://revbayes.github.io/intro.html"><strong>RevBayes</strong></a>, <em>etc</em>. This is made possible with the several parser functions developed in <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> (Table @ref(tab:treeio-function)) <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span>.</p>
<div class="cell">
<div class="cell-output-display">
<table class="table table-striped table-hover caption-top table-sm small">
<caption>Parser functions defined in treeio</caption>
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">Parser function</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">read.astral</td>
<td style="text-align: left;">parsing output of ASTRAL</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.beast</td>
<td style="text-align: left;">parsing output of BEAST</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.codeml</td>
<td style="text-align: left;">parsing output of CodeML (rst and mlc files)</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.codeml_mlc</td>
<td style="text-align: left;">parsing mlc file (output of CodeML)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.fasta</td>
<td style="text-align: left;">parsing FASTA format sequence file</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.hyphy</td>
<td style="text-align: left;">parsing output of HYPHY</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.hyphy.seq</td>
<td style="text-align: left;">parsing ancestral sequences from HYPHY output</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.iqtree</td>
<td style="text-align: left;">parsing IQ-Tree Newick string, with the ability to parse SH-aLRT and UFBoot support values</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.jplace</td>
<td style="text-align: left;">parsing jplace file including the output of EPA and pplacer</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.jtree</td>
<td style="text-align: left;">parsing [jtree](#write-jtree) format</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.mega</td>
<td style="text-align: left;">parsing MEGA Nexus output</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.mega_tabular</td>
<td style="text-align: left;">parsing MEGA tabular output</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.mrbayes</td>
<td style="text-align: left;">parsing output of MrBayes</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.newick</td>
<td style="text-align: left;">parsing Newick string, with the ability to parse node label as support values</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.nexus</td>
<td style="text-align: left;">parsing standard NEXUS file (re-exported from ape)</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.nhx</td>
<td style="text-align: left;">parsing NHX file including the output of PHYLDOG and RevBayes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.paml_rst</td>
<td style="text-align: left;">parsing rst file (output of BaseML or CodeML)</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.phylip</td>
<td style="text-align: left;">parsing phylip file (phylip alignment + Newick string)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.phylip.seq</td>
<td style="text-align: left;">parsing multiple sequence alignment from phylip file</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.phylip.tree</td>
<td style="text-align: left;">parsing newick string from phylip file</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.phyloxml</td>
<td style="text-align: left;">parsing phyloXML file</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.r8s</td>
<td style="text-align: left;">parsing output of r8s</td>
</tr>
<tr class="odd">
<td style="text-align: left;">read.raxml</td>
<td style="text-align: left;">parsing output of RAxML</td>
</tr>
<tr class="even">
<td style="text-align: left;">read.tree</td>
<td style="text-align: left;">parsing newick string (re-exported from ape)</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package defines base classes and functions for phylogenetic tree input and output. It is an infrastructure that enables evolutionary evidence inferred by commonly used software packages to be used in <code>R</code>. For instance, <em>d<sub>N</sub>/d<sub>S</sub></em> values or ancestral sequences inferred by <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> <span class="citation" data-cites="yang_paml_2007">(<a href="#ref-yang_paml_2007" role="doc-biblioref">Yang, 2007</a>)</span>, clade support values (posterior) inferred by <a href="http://beast2.org/"><strong>BEAST</strong></a> <span class="citation" data-cites="bouckaert_beast_2014">(<a href="#ref-bouckaert_beast_2014" role="doc-biblioref">Bouckaert et al., 2014</a>)</span> and short read placement by <a href="http://sco.h-its.org/exelixis/web/software/epa/index.html"><strong>EPA</strong></a> <span class="citation" data-cites="berger_performance_2011">(<a href="#ref-berger_performance_2011" role="doc-biblioref">Berger et al., 2011</a>)</span> and <a href="http://matsen.fhcrc.org/pplacer/"><strong>PPLACER</strong></a> <span class="citation" data-cites="matsen_pplacer_2010">(<a href="#ref-matsen_pplacer_2010" role="doc-biblioref">Matsen et al., 2010</a>)</span>. These pieces of evolutionary evidence can be further analyzed in <code>R</code> and used to annotate a phylogenetic tree using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span>. The growth of analysis tools and models introduces a challenge to integrate different varieties of data and analysis results from different sources for an integral analysis on the same phylogenetic tree background. The <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> provides a <code>merge_tree</code> function to allow <a href="#merge-tree">combining tree data</a> obtained from different sources. In addition, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> also enables <a href="#link-external-data">external data</a> to be linked to a phylogenetic tree structure.</p>
<p>After parsing, storage of the tree structure with associated data is made through an S4 class, <code>treedata</code>, defined in the <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> package. These parsed data are mapped to the tree branches and nodes inside <code>treedata</code> object, so that they can be efficiently used to visually annotate the tree using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span> and <a href="http://bioconductor.org/packages/ggtreeExtra"><strong>ggtreeExtra</strong></a> <span class="citation" data-cites="ggtreeExtra_2021">(<a href="#ref-ggtreeExtra_2021" role="doc-biblioref">Xu et al., 2021</a>)</span>. A programmable platform for phylogenetic data parsing, integration, and annotations as such makes us more easily to identify the evolutionary dynamics and correlation patterns (Figure @ref(fig:treeioDiagram)) <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span>.</p>
<p>(ref:treeioDiagramscap) Overview of the <em>treeio</em> package and its relations with <em>tidytree</em> and <em>ggtree</em>.</p>
<p>(ref:treeioDiagramcap) <strong>Overview of the <em>treeio</em> package and its relations with <em>tidytree</em> and <em>ggtree</em></strong>. <em>Treeio</em> supports parsing a tree with data from a number of file formats and software outputs. A <em>treedata</em> object stores a phylogenetic tree with node/branch-associated data. <em>Treeio</em> provides several functions to manipulate a tree with data. Users can convert the <em>treedata</em> object into a tidy data frame (each row represents a node in the tree and each column represents a variable) and process the tree with data using the tidy interface implemented in <em>tidytree</em>. The tree can be extracted from the <em>treedata</em> object and exported to a Newick and NEXUS file or can be exported with associated data into a single file (either in the BEAST NEXUS or jtree format). Associated data stored in the <em>treedata</em> object can be used to annotate the tree using <em>ggtree</em>. In addition, <em>ggtree</em> supports a number of tree objects, including <em>phyloseq</em> for microbiome data and <em>obkData</em> for outbreak data. The <em>phylo</em>, multiPhylo (<em>ape</em> package), <em>phylo4</em>, <em>phylo4d</em> (<em>phylobase</em> package), <em>phylog</em> (<em>ade4</em> package), <em>phyloseq</em> (<em>phyloseq</em> package), and <em>obkData</em> (<em>OutbreakTools</em> package) are tree objects defined by the R community to store tree with or without domain-specific data. All these tree objects as well as hierarchical clustering results (<em>e.g.</em>, <em>hclust</em> and <em>dendrogram</em> objects) are supported by <em>ggtree</em>.</p>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/treeio-diagram.png" class="img-fluid figure-img" style="width:95.0%"></p>
<figcaption>(ref:treeioDiagramcap)</figcaption>
</figure>
</div>
</div>
</div>
<section id="overview-of-treeio" class="level3">
<h3 class="anchored" data-anchor-id="overview-of-treeio">Overview of <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a></h3>
<p>The <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> defined <code>S4</code> classes for storing phylogenetic trees with diverse types of associated data or covariates from different sources including analysis outputs from different software packages. It also defined corresponding parser functions for parsing phylogenetic trees with annotation data and stored them as data objects in R for further manipulation or analysis (see Table @ref(tab:treeio-function)). Several accessor functions were defined to facilitate accessing the tree annotation data, including <code>get.fields</code> for obtaining annotation features available in the tree object, <code>get.placements</code> for obtaining the phylogenetic placement results (<em>i.e.</em>, the output of <a href="http://matsen.fhcrc.org/pplacer/"><strong>PPLACER</strong></a>, <a href="http://sco.h-its.org/exelixis/web/software/epa/index.html"><strong>EPA</strong></a>, <em>etc.</em>), <code>get.subs</code> for obtaining the genetic substitutions from parent node to child node, and <code>get.tipseq</code> for getting the tip sequences.</p>
<p>The <code>S3</code> class, <code>phylo</code>, which was defined in <a href="https://CRAN.R-project.org/package=ape"><strong>ape</strong></a> <span class="citation" data-cites="paradis_ape_2004">(<a href="#ref-paradis_ape_2004" role="doc-biblioref">Paradis et al., 2004</a>)</span> package, is widely used in <code>R</code> community and many packages. As <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> uses <code>S4</code> class, to enable those available R packages to analyze the tree imported by <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a>, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> provides <code>as.phylo</code> function to convert <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a>-generated tree object to <code>phylo</code> object that only contains tree structure without annotation data. In the other way, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> also provides <code>as.treedata</code> function to convert <code>phylo</code> object with evolutionary analysis result (<em>e.g.</em>, bootstrap values calculated by <a href="https://CRAN.R-project.org/package=ape"><strong>ape</strong></a> or ancestral states inferred by <a href="https://CRAN.R-project.org/package=phangorn"><strong>phangorn</strong></a> <span class="citation" data-cites="schliep_phangorn_2011">(<a href="#ref-schliep_phangorn_2011" role="doc-biblioref">Schliep, 2011</a>)</span> <em>etc.</em>) to be stored as a <code>treedata</code> <code>S4</code> object, making it easy to map the data to the tree structure and to be visualized using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span>.</p>
<!--
(*e.g.*, `phangorn`  for ancestral state reconstruction and `ape` for divergence time estimation) that implemented evolutionary inferences are supporting `phylo` object. 
-->
<p>To allow integration of different kinds of data in a phylogenetic tree, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> provides <code>merge_tree</code> function (details in <a href="#merge-tree">section 2.2.1</a>) for combining evolutionary statistics/evidence imported from different sources including those common tree files and outputs from analysis programs (Table @ref(tab:treeio-function)). There is other information, such as sampling location, taxonomy information, experimental result, and evolutionary traits, <em>etc.</em> that are stored in separate files with the user-defined format. In <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a>, we could read in these data from the users’ files using standard R <em>IO</em> functions, and attach them to the tree object by the <a href="#link-external-data">full_join</a> methods defined in <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> and <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> packages (see also the <a href="#attach-operator"><code>%&lt;+%</code> operator</a> defined in <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a>). After attaching, the data will become the attributes associated with nodes or branches, which can be compared with other data incorporated or can be visually displayed on the tree <span class="citation" data-cites="yu_two_2018">(<a href="#ref-yu_two_2018" role="doc-biblioref">Yu et al., 2018</a>)</span>.</p>
<p>To facilitate storing complex data associated with the phylogenetic tree, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> implemented <code>write.beast</code> and <code>write.jtree</code> functions to export a <code>treedata</code> object into a single file (see <a href="#chapter3">Chapter 3</a>).</p>
<!-- A full list of functions that defined in *treeio* can be found in Table \@ref(tab:treeio). -->
</section>
<section id="function-demonstration" class="level3">
<h3 class="anchored" data-anchor-id="function-demonstration">Function demonstration</h3>
<section id="parsing-beast-output" class="level4">
<h4 class="anchored" data-anchor-id="parsing-beast-output">Parsing BEAST output</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/BEAST"</span>, <span class="st">"beast_mcc.tree"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>beast <span class="ot">&lt;-</span> <span class="fu">read.beast</span>(file)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>beast</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/BEAST/beast_mcc.tree'.

...@ phylo:

Phylogenetic tree with 15 tips and 14 internal nodes.

Tip labels:
  A_1995, B_1996, C_1995, D_1987, E_1996, F_1997, ...

Rooted; includes branch length(s).

with the following features available:
  'height', 'height_0.95_HPD', 'height_median', 'height_range', 'length',
'length_0.95_HPD', 'length_median', 'length_range', 'posterior', 'rate',
'rate_0.95_HPD', 'rate_median', 'rate_range'.</code></pre>
</div>
</div>
<p>Since <em><code>%</code></em> is not a valid character in <em><code>names</code></em>, all the feature names that contain <em><code>x%</code></em> will convert to <em><code>0.x</code></em>. For example, <em><code>length_95%_HPD</code></em> will be changed to <em><code>length_0.95_HPD</code></em>.</p>
<p>Not only tree structure but also all the features inferred by <a href="http://beast2.org/"><strong>BEAST</strong></a> will be stored in the <code>S4</code> object. These features can be used for tree annotation (Figure @ref(fig:beast)).</p>
</section>
<section id="mega" class="level4">
<h4 class="anchored" data-anchor-id="mega">Parsing MEGA output</h4>
<p>Molecular Evolutionary Genetics Analysis (<a href="https://www.megasoftware.net/"><strong>MEGA</strong></a>) software <span class="citation" data-cites="kumar_mega7_2016">(<a href="#ref-kumar_mega7_2016" role="doc-biblioref">Kumar et al., 2016</a>)</span> supports exporting trees in three distinct formats: Newick, tabular, and Nexus. The Newick file can be parsed using the <code>read.tree</code> or <code>read.newick</code> functions. MEGA Nexus file is similar to BEAST Nexus and <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> provides <code>read.mega</code> function to parse the tree.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/MEGA7"</span>, <span class="st">"mtCDNA_timetree.nex"</span>, </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">package =</span> <span class="st">"treeio"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">read.mega</span>(file)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/MEGA7/mtCDNA_timetree.nex'.

...@ phylo:

Phylogenetic tree with 7 tips and 6 internal nodes.

Tip labels:
  homo_sapiens, chimpanzee, bonobo, gorilla, orangutan, sumatran, ...

Rooted; includes branch length(s).

with the following features available:
  'branch_length', 'data_coverage', 'rate', 'reltime', 'reltime_0.95_CI',
'reltime_stderr'.</code></pre>
</div>
</div>
<p>The tabular output contains tree and associated information (divergence time in this example) in a tabular flat text file. The <code>read.mega_tabular</code> function can parse the tree with data simultaneously.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/MEGA7"</span>, <span class="st">"mtCDNA_timetree_tabular.txt"</span>, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">package =</span> <span class="st">"treeio"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">read.mega_tabular</span>(file) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/MEGA7/mtCDNA_timetree_tabular.txt'.

...@ phylo:

Phylogenetic tree with 7 tips and 6 internal nodes.

Tip labels:
  chimpanzee, orangutan, bonobo, homo sapiens, gorilla, sumatran, ...
Node labels:
  , , , demoLabel2, , 

Rooted; no branch length.

with the following features available:
  'RelTime', 'CI_Lower', 'CI_Upper', 'Rate', 'Data Coverage'.</code></pre>
</div>
</div>
</section>
<section id="parsing-mrbayes-output" class="level4">
<h4 class="anchored" data-anchor-id="parsing-mrbayes-output">Parsing MrBayes output</h4>
<p>Although the Nexus file generated by <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a> is different from the output of <a href="http://beast2.org/"><strong>BEAST</strong></a>, they are similar. The <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package provides the <code>read.mrbayes()</code> which internally calls <code>read.beast()</code> to parse <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a> outputs.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/MrBayes"</span>, <span class="st">"Gq_nxs.tre"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">read.mrbayes</span>(file)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/MrBayes/Gq_nxs.tre'.

...@ phylo:

Phylogenetic tree with 12 tips and 10 internal nodes.

Tip labels:
  B_h, B_s, G_d, G_k, G_q, G_s, ...

Unrooted; includes branch length(s).

with the following features available:
  'length_0.95HPD', 'length_mean', 'length_median', 'prob', 'prob_range',
'prob_stddev', 'prob_percent', 'prob+-sd'.</code></pre>
</div>
</div>
</section>
<section id="parsing-paml-output" class="level4">
<h4 class="anchored" data-anchor-id="parsing-paml-output">Parsing PAML output</h4>
<p>Phylogenetic Analysis by Maximum Likelihood (<a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a>) is a package of tools for phylogenetic analyses of DNA and protein sequences using maximum likelihood. Tree search algorithms are implemented in <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> and <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>. The <code>read.paml_rst()</code> function provided in <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> can parse <em>rst</em> file from <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> and <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>. The only difference is the space in the sequences. For <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a>, every ten bases are separated by one space, while for <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>, every three bases (triplet) are separated by one space.</p>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>brstfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/PAML_Baseml"</span>, <span class="st">"rst"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>brst <span class="ot">&lt;-</span> <span class="fu">read.paml_rst</span>(brstfile)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>brst</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/PAML_Baseml/rst'.

...@ phylo:

Phylogenetic tree with 15 tips and 13 internal nodes.

Tip labels:
  A, B, C, D, E, F, ...
Node labels:
  16, 17, 18, 19, 20, 21, ...

Unrooted; includes branch length(s).

with the following features available:
  'subs', 'AA_subs'.</code></pre>
</div>
</div>
<p>Similarly, we can parse the <em>rst</em> file from <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>crstfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/PAML_Codeml"</span>, <span class="st">"rst"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="do">## type can be one of "Marginal" or "Joint"</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>crst <span class="ot">&lt;-</span> <span class="fu">read.paml_rst</span>(crstfile, <span class="at">type =</span> <span class="st">"Joint"</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>crst</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/PAML_Codeml/rst'.

...@ phylo:

Phylogenetic tree with 15 tips and 13 internal nodes.

Tip labels:
  A, B, C, D, E, F, ...
Node labels:
  16, 17, 18, 19, 20, 21, ...

Unrooted; includes branch length(s).

with the following features available:
  'subs', 'AA_subs'.</code></pre>
</div>
</div>
<p>Ancestral sequences inferred by <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> or <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> via marginal or joint ML reconstruction methods will be stored in the S4 object and mapped to tree nodes. <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> will automatically determine the substitutions between the sequences at both ends of each branch. The amino acid substitution will also be determined by translating nucleotide sequences to amino acid sequences. These computed substitutions will also be stored in the S4 object for efficient tree annotation later (Figure @ref(fig:codeml)).</p>
<p><a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> infers selection pressure and estimated <em>d<sub>N</sub>/d<sub>S</sub></em>, <em>d<sub>N</sub></em> and <em>d<sub>S</sub></em>. These pieces of information are stored in output file <em>mlc</em>, which can be parsed by the <code>read.codeml_mlc()</code> function.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>mlcfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/PAML_Codeml"</span>, <span class="st">"mlc"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>mlc <span class="ot">&lt;-</span> <span class="fu">read.codeml_mlc</span>(mlcfile)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>mlc</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/PAML_Codeml/mlc'.

...@ phylo:

Phylogenetic tree with 15 tips and 13 internal nodes.

Tip labels:
  A, B, C, D, E, F, ...
Node labels:
  16, 17, 18, 19, 20, 21, ...

Unrooted; includes branch length(s).

with the following features available:
  't', 'N', 'S', 'dN_vs_dS', 'dN', 'dS', 'N_x_dN', 'S_x_dS'.</code></pre>
</div>
</div>
<p>The <em>rst</em> and <em>mlc</em> files can be parsed separately as demonstrated previously, they can also be parsed together using the <code>read.codeml()</code> function.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="do">## tree can be one of "rst" or "mlc" to specify</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="do">## using tree from which file as base tree in the object</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>ml <span class="ot">&lt;-</span> <span class="fu">read.codeml</span>(crstfile, mlcfile, <span class="at">tree =</span> <span class="st">"mlc"</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>ml</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/PAML_Codeml/rst',
    '/home/runner/work/_temp/Library/treeio/extdata/PAML_Codeml/mlc'.

...@ phylo:

Phylogenetic tree with 15 tips and 13 internal nodes.

Tip labels:
  A, B, C, D, E, F, ...
Node labels:
  16, 17, 18, 19, 20, 21, ...

Unrooted; includes branch length(s).

with the following features available:
  'subs', 'AA_subs', 't', 'N', 'S', 'dN_vs_dS', 'dN', 'dS', 'N_x_dN', 'S_x_dS'.</code></pre>
</div>
</div>
<p>All the features in both <em>rst</em> and <em>mlc</em> files are imported into a single S4 object and hence are available for further annotation and visualization. For example, we can annotate and display both <em>d<sub>N</sub>/d<sub>S</sub></em> (from <em>mlc</em> file) and amino acid substitutions (derived from <em>rst</em> file) on the same phylogenetic tree <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span>.</p>
</section>
<section id="parsing-hyphy-output" class="level4">
<h4 class="anchored" data-anchor-id="parsing-hyphy-output">Parsing HyPhy output</h4>
<p>Hypothesis testing using Phylogenies (<a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a>) is a software package for analyzing genetic sequences. Ancestral sequences inferred by <a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a> are stored in the Nexus output file, which contains the tree topology and ancestral sequences. To parse this data file, users can use the <code>read.hyphy.seq()</code> function.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>ancseq <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/HYPHY"</span>, <span class="st">"ancseq.nex"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">read.hyphy.seq</span>(ancseq)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>13 DNA sequences in binary format stored in a list.

All sequences of same length: 2148 

Labels:
Node1
Node2
Node3
Node4
Node5
Node12
...

Base composition:
    a     c     g     t 
0.335 0.208 0.237 0.220 
(Total: 27.92 kb)</code></pre>
</div>
</div>
<p>To map the sequences on the tree, users should also provide an internal-node-labeled tree. If users want to determine substitution, they need to also provide tip sequences. In this case, substitutions will be determined automatically, just as we parse the output of <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>nwk <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/HYPHY"</span>, <span class="st">"labelledtree.tree"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>tipfas <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"pa.fas"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>hy <span class="ot">&lt;-</span> <span class="fu">read.hyphy</span>(nwk, ancseq, tipfas)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>hy</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/HYPHY/labelledtree.tree'.

...@ phylo:

Phylogenetic tree with 15 tips and 13 internal nodes.

Tip labels:
  K, N, D, L, J, G, ...
Node labels:
  Node1, Node2, Node3, Node4, Node5, Node12, ...

Unrooted; includes branch length(s).

with the following features available:
  'subs', 'AA_subs'.</code></pre>
</div>
</div>
</section>
<section id="parsing-r8s-output" class="level4">
<h4 class="anchored" data-anchor-id="parsing-r8s-output">Parsing r8s output</h4>
<p>The <a href="http://loco.biosci.arizona.edu/r8s/"><strong>r8s</strong></a> package uses parametric, semi-parametric, and non-parametric methods to relax the molecular clock to allow better estimations of divergence times and evolution rates <span class="citation" data-cites="sanderson_r8s:_2003">(<a href="#ref-sanderson_r8s:_2003" role="doc-biblioref">Sanderson, 2003</a>)</span>. It outputs three trees in a log file, namely, <em>TREE</em>, <em>RATO</em>, and <em>PHYLO</em> for time tree, rate tree, and absolute substitution tree, respectively.</p>
<p>The time tree is scaled by divergence time, the rate tree is scaled by substitution rate and the absolute substitution tree is scaled by the absolute number of substitutions. After parsing the file, all these three trees are stored in a <em>multiPhylo</em> object (Figure @ref(fig:multiPhylo)).</p>
<div class="cell" data-layout-align="center" width="60">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>r8s <span class="ot">&lt;-</span> <span class="fu">read.r8s</span>(<span class="fu">system.file</span>(<span class="st">"extdata/r8s"</span>, <span class="st">"H3_r8s_output.log"</span>, <span class="at">package=</span><span class="st">"treeio"</span>))</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>r8s</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>3 phylogenetic trees</code></pre>
</div>
</div>
</section>
<section id="parsing-output-of-raxml-bootstrap-analysis" class="level4">
<h4 class="anchored" data-anchor-id="parsing-output-of-raxml-bootstrap-analysis">Parsing output of RAxML bootstrap analysis</h4>
<p><a href="http://evomics.org/learning/phylogenetics/raxml/"><strong>RAxML</strong></a> bootstrapping analysis outputs a Newick tree text that is not standard, as it stores bootstrap values inside square brackets after branch lengths. This file usually cannot be parsed by a traditional Newick parser, such as <code>ape::read.tree()</code>. The function <code>read.raxml()</code> can read such files and store the bootstrap as an additional feature, which can be used to display on the tree or used to color tree branches, <em>etc.</em></p>
<div class="cell" data-layout-align="center" width="60">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>raxml_file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/RAxML"</span>, </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"RAxML_bipartitionsBranchLabels.H3"</span>, </span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>raxml <span class="ot">&lt;-</span> <span class="fu">read.raxml</span>(raxml_file)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>raxml</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/RAxML/RAxML_bipartitionsBranchLabels.H3'.

...@ phylo:

Phylogenetic tree with 64 tips and 62 internal nodes.

Tip labels:
  A_Hokkaido_M1_2014_H3N2_2014, A_Czech_Republic_1_2014_H3N2_2014,
FJ532080_A_California_09_2008_H3N2_2008,
EU199359_A_Pennsylvania_05_2007_H3N2_2007,
EU857080_A_Hong_Kong_CUHK69904_2006_H3N2_2006,
EU857082_A_Hong_Kong_CUHK7047_2005_H3N2_2005, ...

Unrooted; includes branch length(s).

with the following features available:
  'bootstrap'.</code></pre>
</div>
</div>
</section>
<section id="parsing-nhx-tree" class="level4">
<h4 class="anchored" data-anchor-id="parsing-nhx-tree">Parsing NHX tree</h4>
<p>NHX (New Hampshire eXtended) format is an extension of Newick by introducing NHX tags. NHX is commonly used in phylogenetics software, including <a href="http://pbil.univ-lyon1.fr/software/phyldog/"><strong>PHYLDOG</strong></a> <span class="citation" data-cites="boussau_genome-scale_2013">(<a href="#ref-boussau_genome-scale_2013" role="doc-biblioref">Boussau et al., 2013</a>)</span>, <a href="http://revbayes.github.io/intro.html"><strong>RevBayes</strong></a> <span class="citation" data-cites="hohna_probabilistic_2014">(<a href="#ref-hohna_probabilistic_2014" role="doc-biblioref">Höhna et al., 2014</a>)</span>, for storing statistical inferences. The following codes imported an NHX tree with associated data inferred by <a href="http://pbil.univ-lyon1.fr/software/phyldog/"><strong>PHYLDOG</strong></a> (Figure @ref(fig:beastFigtree)A).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>nhxfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/NHX"</span>, <span class="st">"phyldog.nhx"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>nhx <span class="ot">&lt;-</span> <span class="fu">read.nhx</span>(nhxfile)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>nhx</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/NHX/phyldog.nhx'.

...@ phylo:

Phylogenetic tree with 16 tips and 15 internal nodes.

Tip labels:
  Prayidae_D27SS7@2825365, Kephyes_ovata@2606431,
Chuniphyes_multidentata@1277217, Apolemia_sp_@1353964,
Bargmannia_amoena@263997, Bargmannia_elongata@946788, ...

Rooted; includes branch length(s).

with the following features available:
  'V1', 'Ev', 'ND', 'S'.</code></pre>
</div>
</div>
</section>
<section id="parsing-phylip-tree" class="level4">
<h4 class="anchored" data-anchor-id="parsing-phylip-tree">Parsing Phylip tree</h4>
<p>Phylip format contains multiple sequence alignment of taxa in Phylip sequence format with corresponding Newick tree text that was built from taxon sequences. Multiple sequence alignment can be sorted based on the tree structure and displayed at the right-hand side of the tree using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> through the <a href="#msaplot"><code>msaplot()</code> function</a> or in combining with the <a href="http://bioconductor.org/packages/ggmsa"><strong>ggmsa</strong></a> package (see also Basic Protocol 5 of <span class="citation" data-cites="yu_cp_2020">(<a href="#ref-yu_cp_2020" role="doc-biblioref">Yu, 2020</a>)</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>phyfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"sample.phy"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>phylip <span class="ot">&lt;-</span> <span class="fu">read.phylip</span>(phyfile)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>phylip</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/sample.phy'.

...@ phylo:

Phylogenetic tree with 15 tips and 13 internal nodes.

Tip labels:
  K, N, D, L, J, G, ...

Unrooted; no branch length.</code></pre>
</div>
</div>
</section>
<section id="parsing-epa-and-pplacer-output" class="level4">
<h4 class="anchored" data-anchor-id="parsing-epa-and-pplacer-output">Parsing EPA and pplacer output</h4>
<p><a href="http://sco.h-its.org/exelixis/web/software/epa/index.html"><strong>EPA</strong></a> <span class="citation" data-cites="berger_performance_2011">(<a href="#ref-berger_performance_2011" role="doc-biblioref">Berger et al., 2011</a>)</span> and <a href="http://matsen.fhcrc.org/pplacer/"><strong>PPLACER</strong></a> <span class="citation" data-cites="matsen_pplacer_2010">(<a href="#ref-matsen_pplacer_2010" role="doc-biblioref">Matsen et al., 2010</a>)</span> have a common output file format, <code>jplace</code>, which can be parsed by the <code>read.jplace()</code> function.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>jpf <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/EPA.jplace"</span>,  <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>jp <span class="ot">&lt;-</span> <span class="fu">read.jplace</span>(jpf)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(jp)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/EPA.jplace'.

...@ phylo:

Phylogenetic tree with 493 tips and 492 internal nodes.

Tip labels:
  CIR000447A, CIR000479, CIR000078, CIR000083, CIR000070, CIR000060, ...

Rooted; includes branch length(s).

with the following features available:
  'nplace'.</code></pre>
</div>
</div>
<p>The number of evolutionary placement on each branch will be calculated and stored as the <em>nplace</em> feature, which can be mapped to line size and/or color using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span>.</p>
</section>
<section id="jtree" class="level4">
<h4 class="anchored" data-anchor-id="jtree">Parsing jtree format</h4>
<p>The <em>jtree</em> is a JSON-based format that was defined in this <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> to support tree data interchange (see <a href="#write-jtree">session 3.3</a>). Phylogenetic tree with associated data can be exported to a single <em>jtree</em> file using the <code>write.jtree()</code> function. The <em>jtree</em> can be easily parsed using any JSON parser. The <em>jtree</em> format contains three keys: tree, data, and meta-data. The tree value contains tree text extended from Newick tree format by putting the edge number in curly braces after branch length. The data value contains node/branch-specific data, while the meta-data value contains additional meta information.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>jtree_file <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">'.jtree'</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.jtree</span>(beast, <span class="at">file =</span> jtree_file)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="fu">read.jtree</span>(<span class="at">file =</span> jtree_file)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/tmp/RtmpdQ6kCo/file245dafb228a.jtree'.

...@ phylo:

Phylogenetic tree with 15 tips and 14 internal nodes.

Tip labels:
  K_2013, N_2010, D_1987, L_1980, J_1983, G_1992, ...

Rooted; includes branch length(s).

with the following features available:
  'height', 'height_0.95_HPD', 'height_range', 'length', 'length_0.95_HPD',
'length_median', 'length_range', 'rate', 'rate_0.95_HPD', 'rate_median',
'rate_range', 'height_median', 'posterior'.</code></pre>
</div>
</div>
</section>
</section>
<section id="as-treedata" class="level3">
<h3 class="anchored" data-anchor-id="as-treedata">Converting other tree-like objects to <code>phylo</code> or <code>treedata</code> objects</h3>
<p>To extend the application scopes of <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a>, <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> and <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a>, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> provides several <code>as.phylo</code> and <code>as.treedata</code> methods to convert other tree-like objects, such as <code>phylo4d</code> and <code>pml</code>, to <code>phylo</code> or <code>treedata</code> object. So that users can easily <a href="02_tidytree.html#data-integration">map associated data to the tree structure</a>, <a href="#chapter3">export a tree with/without data to a single file</a>, <a href="#chapter2">manipulate</a> and <a href="#chapter4">visualize a tree</a> with/without data. These convert functions (Table @ref(tab:treeio-treedata-object)) create the possibility of using <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> to process tree using tidy interface and <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> to visualize tree using the grammar of graphic syntax.</p>
<div class="cell">
<div class="cell-output-display">
<table class="table table-striped table-hover caption-top table-sm small">
<caption>Conversion of tree-like object to phylo or treedata object</caption>
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">Convert function</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Supported object</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td rowspan="5" style="text-align: left; vertical-align: top !important;">as.phylo</td>
<td style="text-align: left;">ggtree</td>
<td style="text-align: left;">convert ggtree object to phylo object</td>
</tr>
<tr class="even">
<td style="text-align: left;">igraph</td>
<td style="text-align: left;">convert igraph object (only tree graph supported) to phylo object</td>
</tr>
<tr class="odd">
<td style="text-align: left;">phylo4</td>
<td style="text-align: left;">convert phylo4 object to phylo object</td>
</tr>
<tr class="even">
<td style="text-align: left;">pvclust</td>
<td style="text-align: left;">convert pvclust object to phylo object</td>
</tr>
<tr class="odd">
<td style="text-align: left;">treedata</td>
<td style="text-align: left;">convert treedata object to phylo object</td>
</tr>
<tr class="even">
<td rowspan="5" style="text-align: left; vertical-align: top !important;">as.treedata</td>
<td style="text-align: left;">ggtree</td>
<td style="text-align: left;">convert ggtree object to treedata object</td>
</tr>
<tr class="odd">
<td style="text-align: left;">phylo4</td>
<td style="text-align: left;">convert phylo4 object to treedata object</td>
</tr>
<tr class="even">
<td style="text-align: left;">phylo4d</td>
<td style="text-align: left;">convert phylo4d object to treedata object</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pml</td>
<td style="text-align: left;">convert pml object to treedata object</td>
</tr>
<tr class="even">
<td style="text-align: left;">pvclust</td>
<td style="text-align: left;">convert pvclust object to treedata object</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Here, we used <code>pml</code> object which was defined in the <a href="https://CRAN.R-project.org/package=phangorn"><strong>phangorn</strong></a> package, as an example. The <code>pml()</code> function computes the likelihood of a phylogenetic tree given a sequence alignment and a model and the <code>optim.pml()</code> function optimizes different model parameters. The output is a <code>pml</code> object, and it can be converted to a <code>treedata</code> object using <code>as.treedata</code> provided by <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span>. The amino acid substitution (ancestral sequence estimated by <code>pml</code>) that stored in the <code>treedata</code> object can be visualized using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> as demonstrated in Figure @ref(fig:pmlTree).</p>
<p>(ref:pmlTreescap) Converting <code>pml</code> object to <code>treedata</code> object.</p>
<p>(ref:pmlTreecap) <strong>Converting <code>pml</code> object to <code>treedata</code> object.</strong> This allows using <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> to process the tree data as well as using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> and <a href="http://bioconductor.org/packages/ggtreeExtra"><strong>ggtreeExtra</strong></a> to visualize the tree with associated data.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phangorn)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>treefile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"pa.nwk"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>tre <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(treefile)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>tipseqfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"pa.fas"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>tipseq <span class="ot">&lt;-</span> <span class="fu">read.phyDat</span>(tipseqfile,<span class="at">format=</span><span class="st">"fasta"</span>)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">pml</span>(tre, tipseq, <span class="at">k=</span><span class="dv">4</span>)</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">optim.pml</span>(fit, <span class="at">optNni=</span><span class="cn">FALSE</span>, <span class="at">optBf=</span>T, <span class="at">optQ=</span>T,</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">optInv=</span>T, <span class="at">optGamma=</span>T, <span class="at">optEdge=</span><span class="cn">TRUE</span>,</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">optRooted=</span><span class="cn">FALSE</span>, <span class="at">model =</span> <span class="st">"GTR"</span>,</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>                 <span class="at">control =</span> <span class="fu">pml.control</span>(<span class="at">trace =</span><span class="dv">0</span>))</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>pmltree <span class="ot">&lt;-</span> <span class="fu">as.treedata</span>(fit)</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(pmltree) <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x=</span>branch, <span class="at">label=</span>AA_subs, <span class="at">vjust=</span><span class="sc">-</span>.<span class="dv">5</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 1 row containing missing values or values outside the scale range
(`geom_text()`).</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="01_treeio_importing_tree_files/figure-html/pmlTree-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption>(ref:pmlTreecap)</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="get-treedata-data" class="level3">
<h3 class="anchored" data-anchor-id="get-treedata-data">Getting information from <code>treedata</code> object</h3>
<p>After the tree was imported, users may want to extract information stored in the <code>treedata</code> object. <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> provides several accessor methods to extract tree structure, features/attributes that stored in the object, and their corresponding values.</p>
<p>The <code>get.tree()</code> or <code>as.phylo()</code> methods can convert the <code>treedata</code> object to a <code>phylo</code> object which is the fundamental tree object in the R community and many packages work with <code>phylo</code> object.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>beast_file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"examples/MCC_FluA_H3.tree"</span>, <span class="at">package=</span><span class="st">"ggtree"</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>beast_tree <span class="ot">&lt;-</span> <span class="fu">read.beast</span>(beast_file)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="co"># or get.tree</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="fu">as.phylo</span>(beast_tree)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>beast_file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"examples/MCC_FluA_H3.tree"</span>, <span class="at">package=</span><span class="st">"ggtree"</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>beast_tree <span class="ot">&lt;-</span> <span class="fu">read.beast</span>(beast_file)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="co"># or get.tree</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">as.phylo</span>(beast_tree), <span class="at">printlen=</span><span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 76 tips and 75 internal nodes.

Tip labels:
  A/Hokkaido/30-1-a/2013, A/New_York/334/2004, A/New_York/463/2005, ...

Rooted; includes branch length(s).</code></pre>
</div>
</div>
<p>The <code>get.fields</code> method returns a vector of features/attributes stored in the object and associated with the phylogeny.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get.fields</span>(beast_tree)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "height"          "height_0.95_HPD" "height_median"   "height_range"   
 [5] "length"          "length_0.95_HPD" "length_median"   "length_range"   
 [9] "posterior"       "rate"            "rate_0.95_HPD"   "rate_median"    
[13] "rate_range"     </code></pre>
</div>
</div>
<p>The <code>get.data</code> method returns a tibble of all the associated data.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get.data</span>(beast_tree)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 151 × 14
   height height_0.95_HPD height_median height_range length length_0.95_HPD
    &lt;dbl&gt; &lt;list&gt;                  &lt;dbl&gt; &lt;list&gt;        &lt;dbl&gt; &lt;list&gt;         
 1   19   &lt;dbl [2]&gt;                19   &lt;dbl [2]&gt;     2.34  &lt;dbl [2]&gt;      
 2   17   &lt;dbl [2]&gt;                17   &lt;dbl [2]&gt;     1.18  &lt;dbl [2]&gt;      
 3   14   &lt;dbl [2]&gt;                14   &lt;dbl [2]&gt;     0.966 &lt;dbl [2]&gt;      
 4   12   &lt;dbl [2]&gt;                12   &lt;dbl [2]&gt;     1.87  &lt;dbl [2]&gt;      
 5    9   &lt;dbl [2]&gt;                 9   &lt;dbl [2]&gt;     2.93  &lt;dbl [2]&gt;      
 6   10   &lt;dbl [2]&gt;                10   &lt;dbl [2]&gt;     0.827 &lt;dbl [2]&gt;      
 7   10   &lt;dbl [2]&gt;                10   &lt;dbl [2]&gt;     0.834 &lt;dbl [2]&gt;      
 8   10.8 &lt;dbl [2]&gt;                10.8 &lt;dbl [2]&gt;     0.233 &lt;dbl [2]&gt;      
 9    9   &lt;dbl [2]&gt;                 9   &lt;dbl [2]&gt;     1.28  &lt;dbl [2]&gt;      
10    9   &lt;dbl [2]&gt;                 9   &lt;dbl [2]&gt;     0.414 &lt;dbl [2]&gt;      
# ℹ 141 more rows
# ℹ 8 more variables: length_median &lt;dbl&gt;, length_range &lt;list&gt;,
#   posterior &lt;dbl&gt;, rate &lt;dbl&gt;, rate_0.95_HPD &lt;list&gt;, rate_median &lt;dbl&gt;,
#   rate_range &lt;list&gt;, node &lt;int&gt;</code></pre>
</div>
</div>
<p>If users are only interested in a subset of the features/attributes returned by <code>get.fields</code>, they can extract the information from the output of <code>get.data</code> or directly subset the data by <code>[</code> or <code>[[</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>beast_tree[, <span class="fu">c</span>(<span class="st">"node"</span>, <span class="st">"height"</span>)]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 151 × 2
    node height
   &lt;int&gt;  &lt;dbl&gt;
 1    10   19  
 2     9   17  
 3    36   14  
 4    31   12  
 5    29    9  
 6    28   10  
 7    39   10  
 8    90   10.8
 9    16    9  
10     2    9  
# ℹ 141 more rows</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(beast_tree[[<span class="st">"height_median"</span>]])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>height_median1 height_median2 height_median3 height_median4 height_median5 
            19             17             14             12              9 
height_median6 
            10 </code></pre>
</div>
</div>
</section>
</section>
<section id="summary1" class="level2">
<h2 class="anchored" data-anchor-id="summary1">Summary</h2>
<p>Software tools for inferring molecular evolution (<em>e.g.</em>, ancestral states, molecular dating and selection pressure, <em>etc.</em>) are proliferating, but there is no single data format that is used by all different programs and capable to store different types of phylogenetic data. Most of the software packages have their unique output formats and these formats are not compatible with each other. Parsing software outputs is challenging, which restricts the joint analysis using different tools. The <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> provides a set of functions (Table @ref(tab:treeio-function)) for parsing various types of phylogenetic data files and a set of converters (Table @ref(tab:treeio-treedata-object)) to convert tree-like objects to phylo or treedata objects. These phylogenetic data can be integrated that allow further exploration and comparison. To date, most software tools in the field of molecular evolution are isolated and often not fully compatible with each other’s input and output files. These software tools are designed to do their analysis and the outputs are often not readable in other software. No tools have been designed to unify the inference data from different analysis programs. Efficient incorporation of data from different inference methods can enhance the comparison and understanding of the study target, which may help discover new systematic patterns and generate new hypotheses.</p>
<p>As phylogenetic trees are growing in their application to identify patterns in an evolutionary context, more different disciplines are employing phylogenetic trees in their research. For example, spatial ecologists may map the geographical positions of the organisms to their phylogenetic trees to understand the biogeography of the species <span class="citation" data-cites="schon_age_2015">(<a href="#ref-schon_age_2015" role="doc-biblioref">Schön et al., 2015</a>)</span>; disease epidemiologists may incorporate the pathogen sampling time and locations into the phylogenetic analysis to infer the disease transmission dynamics in spatiotemporal space <span class="citation" data-cites="he_emergence_2013">(<a href="#ref-he_emergence_2013" role="doc-biblioref">He et al., 2013</a>)</span>; microbiologists may determine the pathogenicity of different pathogen strains and map them into their phylogenetic trees to identify the genetic determinants of the pathogenicity <span class="citation" data-cites="bosi_comparative_2016">(<a href="#ref-bosi_comparative_2016" role="doc-biblioref">Bosi et al., 2016</a>)</span>; genomic scientists may use the phylogenetic trees to help taxonomically classify their metagenomic sequence data <span class="citation" data-cites="gupta_using_2015">(<a href="#ref-gupta_using_2015" role="doc-biblioref">Gupta &amp; Sharma, 2015</a>)</span>. A robust tool such as <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> to import and map different types of data into the phylogenetic tree is important to facilitate these phylogenetics-related research, or ‘phylodynamics’. Such tools could also help integrate different meta-data (time, geography, genotype, epidemiological information) and analysis results (selective pressure, evolutionary rates) at the highest level and provide a comprehensive understanding of the study organisms. In the field of influenza research, there have been such attempts of studying the phylodynamics of the influenza virus by mapping different meta-data and analysis results on the same phylogenetic tree and evolutionary timescale <span class="citation" data-cites="lam_dissemination_2015">(<a href="#ref-lam_dissemination_2015" role="doc-biblioref">Lam et al., 2015</a>)</span>.</p>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-arenas_trends_2015" class="csl-entry" role="listitem">
Arenas, M. (2015). Trends in substitution models of molecular evolution. <em>Frontiers in Genetics</em>, <em>6</em>. <a href="https://doi.org/10.3389/fgene.2015.00319">https://doi.org/10.3389/fgene.2015.00319</a>
</div>
<div id="ref-berger_performance_2011" class="csl-entry" role="listitem">
Berger, S. A., Krompass, D., &amp; Stamatakis, A. (2011). Performance, <span>Accuracy</span>, and <span>Web</span> <span>Server</span> for <span>Evolutionary</span> <span>Placement</span> of <span>Short</span> <span>Sequence</span> <span>Reads</span> under <span>Maximum</span> <span>Likelihood</span>. <em>Systematic Biology</em>, 291–302. <a href="https://doi.org/10.1093/sysbio/syr010">https://doi.org/10.1093/sysbio/syr010</a>
</div>
<div id="ref-bosi_comparative_2016" class="csl-entry" role="listitem">
Bosi, E., Monk, J. M., Aziz, R. K., Fondi, M., Nizet, V., &amp; Palsson, B. Ø. (2016). Comparative genome-scale modelling of <span>Staphylococcus</span> aureus strains identifies strain-specific metabolic capabilities linked to pathogenicity. <em>Proceedings of the National Academy of Sciences of the United States of America</em>, <em>113</em>(26), E3801–E3809. <a href="https://doi.org/10.1073/pnas.1523199113">https://doi.org/10.1073/pnas.1523199113</a>
</div>
<div id="ref-bouckaert_beast_2014" class="csl-entry" role="listitem">
Bouckaert, R., Heled, J., Kühnert, D., Vaughan, T., Wu, C.-H., Xie, D., Suchard, M. A., Rambaut, A., &amp; Drummond, A. J. (2014). <span>BEAST</span> 2: <span>A</span> <span>Software</span> <span>Platform</span> for <span>Bayesian</span> <span>Evolutionary</span> <span>Analysis</span>. <em>PLoS Comput Biol</em>, <em>10</em>(4), e1003537. <a href="https://doi.org/10.1371/journal.pcbi.1003537">https://doi.org/10.1371/journal.pcbi.1003537</a>
</div>
<div id="ref-boussau_genome-scale_2013" class="csl-entry" role="listitem">
Boussau, B., Szöllősi, G. J., Duret, L., Gouy, M., Tannier, E., &amp; Daubin, V. (2013). Genome-scale coestimation of species and gene trees. <em>Genome Research</em>, <em>23</em>(2), 323–330. <a href="https://doi.org/10.1101/gr.141978.112">https://doi.org/10.1101/gr.141978.112</a>
</div>
<div id="ref-felsenstein_cases_1978" class="csl-entry" role="listitem">
Felsenstein, J. (1978). Cases in which <span>Parsimony</span> or <span>Compatibility</span> <span>Methods</span> will be <span>Positively</span> <span>Misleading</span>. <em>Systematic Biology</em>, <em>27</em>(4), 401–410. <a href="https://doi.org/10.1093/sysbio/27.4.401">https://doi.org/10.1093/sysbio/27.4.401</a>
</div>
<div id="ref-felsenstein_evolutionary_1981" class="csl-entry" role="listitem">
Felsenstein, J. (1981). <a href="https://www.ncbi.nlm.nih.gov/pubmed/7288891">Evolutionary trees from <span>DNA</span> sequences: A maximum likelihood approach</a>. <em>Journal of Molecular Evolution</em>, <em>17</em>(6), 368–376.
</div>
<div id="ref-felsenstein_phylip_1989" class="csl-entry" role="listitem">
Felsenstein, J. (1989). <span>PHYLIP</span> - <span>Phylogeny</span> <span>Inference</span> <span>Package</span> (<span>Version</span> 3.2). <em>Cladistics</em>, <em>5</em>, 164–166.
</div>
<div id="ref-fitch_toward_1971" class="csl-entry" role="listitem">
Fitch, W. M. (1971). Toward <span>Defining</span> the <span>Course</span> of <span>Evolution</span>: <span>Minimum</span> <span>Change</span> for a <span>Specific</span> <span>Tree</span> <span>Topology</span>. <em>Systematic Zoology</em>, <em>20</em>(4), 406–416. <a href="https://doi.org/10.2307/2412116">https://doi.org/10.2307/2412116</a>
</div>
<div id="ref-goldman_codon-based_1994" class="csl-entry" role="listitem">
Goldman, N., &amp; Yang, Z. (1994). <a href="https://www.ncbi.nlm.nih.gov/pubmed/7968486">A codon-based model of nucleotide substitution for protein-coding <span>DNA</span> sequences</a>. <em>Molecular Biology and Evolution</em>, <em>11</em>(5), 725–736.
</div>
<div id="ref-gupta_using_2015" class="csl-entry" role="listitem">
Gupta, A., &amp; Sharma, V. K. (2015). Using the taxon-specific genes for the taxonomic classification of bacterial genomes. <em>BMC Genomics</em>, <em>16</em>(1). <a href="https://doi.org/10.1186/s12864-015-1542-0">https://doi.org/10.1186/s12864-015-1542-0</a>
</div>
<div id="ref-he_emergence_2013" class="csl-entry" role="listitem">
He, Y.-Q., Chen, L., Xu, W.-B., Yang, H., Wang, H.-Z., Zong, W.-P., Xian, H.-X., Chen, H.-L., Yao, X.-J., Hu, Z.-L., Luo, M., Zhang, H.-L., Ma, H.-W., Cheng, J.-Q., Feng, Q.-J., &amp; Zhao, D.-J. (2013). Emergence, <span>Circulation</span>, and <span>Spatiotemporal</span> <span>Phylogenetic</span> <span>Analysis</span> of <span>Coxsackievirus</span> <span>A</span>6- and <span>Coxsackievirus</span> <span>A</span>10-<span>Associated</span> <span>Hand</span>, <span>Foot</span>, and <span>Mouth</span> <span>Disease</span> <span>Infections</span> from 2008 to 2012 in <span>Shenzhen</span>, <span>China</span>. <em>Journal of Clinical Microbiology</em>, <em>51</em>(11), 3560–3566. <a href="https://doi.org/10.1128/JCM.01231-13">https://doi.org/10.1128/JCM.01231-13</a>
</div>
<div id="ref-hohna_probabilistic_2014" class="csl-entry" role="listitem">
Höhna, S., Heath, T. A., Boussau, B., Landis, M. J., Ronquist, F., &amp; Huelsenbeck, J. P. (2014). Probabilistic graphical model representation in phylogenetics. <em>Systematic Biology</em>, <em>63</em>(5), 753–771. <a href="https://doi.org/10.1093/sysbio/syu039">https://doi.org/10.1093/sysbio/syu039</a>
</div>
<div id="ref-hohna_revbayes:_2016" class="csl-entry" role="listitem">
Höhna, S., Landis, M. J., Heath, T. A., Boussau, B., Lartillot, N., Moore, B. R., Huelsenbeck, J. P., &amp; Ronquist, F. (2016). <span>RevBayes</span>: <span>Bayesian</span> <span>Phylogenetic</span> <span>Inference</span> <span>Using</span> <span>Graphical</span> <span>Models</span> and an <span>Interactive</span> <span>Model</span>-<span>Specification</span> <span>Language</span>. <em>Systematic Biology</em>, <em>65</em>(4), 726–736. <a href="https://doi.org/10.1093/sysbio/syw021">https://doi.org/10.1093/sysbio/syw021</a>
</div>
<div id="ref-huelsenbeck_mrbayes_2001" class="csl-entry" role="listitem">
Huelsenbeck, J. P., &amp; Ronquist, F. (2001). <a href="https://www.ncbi.nlm.nih.gov/pubmed/11524383"><span>MRBAYES</span>: <span>Bayesian</span> inference of phylogenetic trees</a>. <em>Bioinformatics (Oxford, England)</em>, <em>17</em>(8), 754–755.
</div>
<div id="ref-kumar_mega7_2016" class="csl-entry" role="listitem">
Kumar, S., Stecher, G., &amp; Tamura, K. (2016). <span>MEGA</span>7: Molecular evolutionary genetics analysis version 7.0 for bigger datasets. <em>Molecular Biology and Evolution</em>, <em>33</em>(7), 1870–1874. <a href="https://doi.org/10.1093/molbev/msw054">https://doi.org/10.1093/molbev/msw054</a>
</div>
<div id="ref-lam_dissemination_2015" class="csl-entry" role="listitem">
Lam, T. T.-Y., Zhou, B., Wang, J., Chai, Y., Shen, Y., Chen, X., Ma, C., Hong, W., Chen, Y., Zhang, Y., Duan, L., Chen, P., Jiang, J., Zhang, Y., Li, L., Poon, L. L. M., Webby, R. J., Smith, D. K., Leung, G. M., … Zhu, H. (2015). Dissemination, divergence and establishment of <span>H</span>7N9 influenza viruses in <span>China</span>. <em>Nature</em>, <em>522</em>(7554), 102–105. <a href="https://doi.org/10.1038/nature14348">https://doi.org/10.1038/nature14348</a>
</div>
<div id="ref-lemmon_importance_2004" class="csl-entry" role="listitem">
Lemmon, A. R., &amp; Moriarty, E. C. (2004). The importance of proper model assumption in bayesian phylogenetics. <em>Systematic Biology</em>, <em>53</em>(2), 265–277. <a href="https://doi.org/10.1080/10635150490423520">https://doi.org/10.1080/10635150490423520</a>
</div>
<div id="ref-maddison_nexus:_1997" class="csl-entry" role="listitem">
Maddison, D. R., Swofford, D. L., Maddison, W. P., &amp; Cannatella, D. (1997). Nexus: <span>An</span> <span>Extensible</span> <span>File</span> <span>Format</span> for <span>Systematic</span> <span>Information</span>. <em>Systematic Biology</em>, <em>46</em>(4), 590–621. <a href="https://doi.org/10.1093/sysbio/46.4.590">https://doi.org/10.1093/sysbio/46.4.590</a>
</div>
<div id="ref-matsen_format_2012" class="csl-entry" role="listitem">
Matsen, F. A., Hoffman, N. G., Gallagher, A., &amp; Stamatakis, A. (2012). A <span>Format</span> for <span>Phylogenetic</span> <span>Placements</span>. <em>PLOS ONE</em>, <em>7</em>(2), e31009. <a href="https://doi.org/10.1371/journal.pone.0031009">https://doi.org/10.1371/journal.pone.0031009</a>
</div>
<div id="ref-matsen_pplacer_2010" class="csl-entry" role="listitem">
Matsen, F. A., Kodner, R. B., &amp; Armbrust, E. V. (2010). Pplacer: Linear time maximum-likelihood and bayesian phylogenetic placement of sequences onto a fixed reference tree. <em><span>BMC</span> Bioinformatics</em>, <em>11</em>(1), 538. <a href="https://doi.org/10.1186/1471-2105-11-538">https://doi.org/10.1186/1471-2105-11-538</a>
</div>
<div id="ref-paradis_ape_2004" class="csl-entry" role="listitem">
Paradis, E., Claude, J., &amp; Strimmer, K. (2004). <span>APE</span>: <span>Analyses</span> of <span>Phylogenetics</span> and <span>Evolution</span> in <span>R</span> language. <em>Bioinformatics</em>, <em>20</em>(2), 289–290. <a href="https://doi.org/10.1093/bioinformatics/btg412">https://doi.org/10.1093/bioinformatics/btg412</a>
</div>
<div id="ref-pond_hyphy:_2005" class="csl-entry" role="listitem">
Pond, S. L. K., Frost, S. D. W., &amp; Muse, S. V. (2005). <span>HyPhy</span>: Hypothesis testing using phylogenies. <em>Bioinformatics (Oxford, England)</em>, <em>21</em>(5), 676–679. <a href="https://doi.org/10.1093/bioinformatics/bti079">https://doi.org/10.1093/bioinformatics/bti079</a>
</div>
<div id="ref-rstats" class="csl-entry" role="listitem">
R Core Team. (2016). <em>R: A language and environment for statistical computing</em>. R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>
</div>
<div id="ref-rannala_probability_1996" class="csl-entry" role="listitem">
Rannala, B., &amp; Yang, Z. (1996). <a href="https://www.ncbi.nlm.nih.gov/pubmed/8703097">Probability distribution of molecular evolutionary trees: A new method of phylogenetic inference</a>. <em>Journal of Molecular Evolution</em>, <em>43</em>(3), 304–311.
</div>
<div id="ref-retief_phylogenetic_2000" class="csl-entry" role="listitem">
Retief, J. D. (2000). <a href="https://www.ncbi.nlm.nih.gov/pubmed/10547839">Phylogenetic analysis using <span>PHYLIP</span></a>. <em>Methods in Molecular Biology (Clifton, N.J.)</em>, <em>132</em>, 243–258.
</div>
<div id="ref-sanderson_r8s:_2003" class="csl-entry" role="listitem">
Sanderson, M. J. (2003). r8s: Inferring absolute rates of molecular evolution and divergence times in the absence of a molecular clock. <em>Bioinformatics</em>, <em>19</em>(2), 301–302. <a href="https://doi.org/10.1093/bioinformatics/19.2.301">https://doi.org/10.1093/bioinformatics/19.2.301</a>
</div>
<div id="ref-schliep_phangorn_2011" class="csl-entry" role="listitem">
Schliep, K. P. (2011). Phangorn: Phylogenetic analysis in <span>R</span>. <em>Bioinformatics</em>, <em>27</em>(4), 592–593. <a href="https://doi.org/10.1093/bioinformatics/btq706">https://doi.org/10.1093/bioinformatics/btq706</a>
</div>
<div id="ref-schmidt_tree-puzzle:_2002" class="csl-entry" role="listitem">
Schmidt, H. A., Strimmer, K., Vingron, M., &amp; Haeseler, A. von. (2002). <a href="https://www.ncbi.nlm.nih.gov/pubmed/11934758"><span>TREE</span>-<span>PUZZLE</span>: Maximum likelihood phylogenetic analysis using quartets and parallel computing</a>. <em>Bioinformatics (Oxford, England)</em>, <em>18</em>(3), 502–504.
</div>
<div id="ref-schon_age_2015" class="csl-entry" role="listitem">
Schön, I., Shearn, R., Martens, K., Koenders, A., &amp; Halse, S. (2015). Age and origin of <span>Australian</span> <span>Bennelongia</span> (<span>Crustacea</span>, <span>Ostracoda</span>). <em>Hydrobiologia</em>, <em>750</em>(1), 125–146. <a href="https://doi.org/10.1007/s10750-014-2159-z">https://doi.org/10.1007/s10750-014-2159-z</a>
</div>
<div id="ref-shoemaker_evidence_1989" class="csl-entry" role="listitem">
Shoemaker, J. S., &amp; Fitch, W. M. (1989). <a href="https://www.ncbi.nlm.nih.gov/pubmed/2622335">Evidence from nuclear sequences that invariable sites should be considered when sequence divergence is calculated</a>. <em>Molecular Biology and Evolution</em>, <em>6</em>(3), 270–289.
</div>
<div id="ref-stamatakis_raxml_2014" class="csl-entry" role="listitem">
Stamatakis, A. (2014). <span>RAxML</span> version 8: A tool for phylogenetic analysis and post-analysis of large phylogenies. <em>Bioinformatics</em>, <em>30</em>(9), 1312–1313. <a href="https://doi.org/10.1093/bioinformatics/btu033">https://doi.org/10.1093/bioinformatics/btu033</a>
</div>
<div id="ref-wang_treeio_2020" class="csl-entry" role="listitem">
Wang, L.-G., Lam, T. T.-Y., Xu, S., Dai, Z., Zhou, L., Feng, T., Guo, P., Dunn, C. W., Jones, B. R., Bradley, T., Zhu, H., Guan, Y., Jiang, Y., &amp; Yu, G. (2020). Treeio: An r package for phylogenetic tree input and output with richly annotated and associated data. <em>Molecular Biology and Evolution</em>, <em>37</em>(2), 599–603. <a href="https://doi.org/10.1093/molbev/msz240">https://doi.org/10.1093/molbev/msz240</a>
</div>
<div id="ref-wilgenbusch_inferring_2003" class="csl-entry" role="listitem">
Wilgenbusch, J. C., &amp; Swofford, D. (2003). Inferring evolutionary trees with <span>PAUP</span>*. <em>Current Protocols in Bioinformatics</em>, <em>Chapter 6</em>, Unit 6.4. <a href="https://doi.org/10.1002/0471250953.bi0604s00">https://doi.org/10.1002/0471250953.bi0604s00</a>
</div>
<div id="ref-ggtreeExtra_2021" class="csl-entry" role="listitem">
Xu, S., Dai, Z., Guo, P., Fu, X., Liu, S., Zhou, L., Tang, W., Feng, T., Chen, M., Zhan, L., Wu, T., Hu, E., Jiang, Y., Bo, X., &amp; Yu, G. (2021). <span class="nocase">ggtreeExtra: Compact Visualization of Richly Annotated Phylogenetic Data</span>. <em>Molecular Biology and Evolution</em>, <em>38</em>(9), 4039–4042. <a href="https://doi.org/10.1093/molbev/msab166">https://doi.org/10.1093/molbev/msab166</a>
</div>
<div id="ref-yang_maximum_1994" class="csl-entry" role="listitem">
Yang, Z. (1994). <a href="https://www.ncbi.nlm.nih.gov/pubmed/7932792">Maximum likelihood phylogenetic estimation from <span>DNA</span> sequences with variable rates over sites: Approximate methods</a>. <em>Journal of Molecular Evolution</em>, <em>39</em>(3), 306–314.
</div>
<div id="ref-yang_paml_2007" class="csl-entry" role="listitem">
Yang, Z. (2007). <span>PAML</span> 4: <span>Phylogenetic</span> <span>Analysis</span> by <span>Maximum</span> <span>Likelihood</span>. <em>Molecular Biology and Evolution</em>, <em>24</em>(8), 1586–1591. <a href="https://doi.org/10.1093/molbev/msm088">https://doi.org/10.1093/molbev/msm088</a>
</div>
<div id="ref-yu_cp_2020" class="csl-entry" role="listitem">
Yu, G. (2020). Using ggtree to visualize data on tree-like structures. <em>Current Protocols in Bioinformatics</em>, <em>69</em>(1), e96. <a href="https://doi.org/10.1002/cpbi.96">https://doi.org/10.1002/cpbi.96</a>
</div>
<div id="ref-yu_two_2018" class="csl-entry" role="listitem">
Yu, G., Lam, T. T.-Y., Zhu, H., &amp; Guan, Y. (2018). Two methods for mapping and visualizing associated data on phylogeny using ggtree. <em>Molecular Biology and Evolution</em>, <em>35</em>(12), 3041–3043. <a href="https://doi.org/10.1093/molbev/msy194">https://doi.org/10.1093/molbev/msy194</a>
</div>
<div id="ref-yu_ggtree:_2017" class="csl-entry" role="listitem">
Yu, G., Smith, D. K., Zhu, H., Guan, Y., &amp; Lam, T. T.-Y. (2017). Ggtree: An r package for visualization and annotation of phylogenetic trees with their covariates and other associated data. <em>Methods in Ecology and Evolution</em>, <em>8</em>(1), 28–36. <a href="https://doi.org/10.1111/2041-210X.12628">https://doi.org/10.1111/2041-210X.12628</a>
</div>
<div id="ref-zmasek_atv:_2001" class="csl-entry" role="listitem">
Zmasek, C. M., &amp; Eddy, S. R. (2001). <span>ATV</span>: Display and manipulation of annotated phylogenetic trees. <em>Bioinformatics</em>, <em>17</em>(4), 383–384. <a href="https://doi.org/10.1093/bioinformatics/17.4.383">https://doi.org/10.1093/bioinformatics/17.4.383</a>
</div>
</div>
</section>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>http://evolution.genetics.washington.edu/phylip/newick_doc.html<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>http://www.genetics.wustl.edu/eddy/forester/NHX.html<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p><a href="http://biopython.org/wiki/PAML" class="uri">http://biopython.org/wiki/PAML</a><a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p><a href="http://bioperl.org/howtos/PAML_HOWTO.html" class="uri">http://bioperl.org/howtos/PAML_HOWTO.html</a><a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p><a href="https://github.com/ropensci/software-review/issues/179#issuecomment-369164110" class="uri">https://github.com/ropensci/software-review/issues/179#issuecomment-369164110</a><a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./00_author.html" class="pagination-link" aria-label="About the Author">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">About the Author</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./02_tidytree.html" class="pagination-link" aria-label="Manipulating Tree with Data">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Manipulating Tree with Data</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/YuLab-SMU/treedata-book/edit/main/01_treeio_importing_tree.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/blob/main/01_treeio_importing_tree.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div></div></footer></body></html>