<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Master ggtree package suite to handle tree with data.">

<title>2&nbsp; Manipulating Tree with Data – Data Integration, Manipulation and Visualization of Phylogenetic Trees</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./03_treeio_exporting_tree.html" rel="next">
<link href="./01_treeio_importing_tree.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-0dd2bd5de344125cf763a379ddc3eb04.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="css/style.css">
</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./02_tidytree.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Manipulating Tree with Data</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Data Integration, Manipulation and Visualization of Phylogenetic Trees</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/YuLab-SMU/treedata-book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Data Integration, Manipulation and Visualization of Phylogenetic Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./00_author.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About the Author</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_treeio_importing_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part I: Tree data input, output, and manipulation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_tidytree.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Manipulating Tree with Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_treeio_exporting_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Exporting tree with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_ggtree_visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part II: Tree data visualization and annotation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_ggtree_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Phylogenetic Tree Annotation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_ggtree_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Visual Exploration of Phylogenetic Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_ggtree_tree_with_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Plotting tree with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_ggtree_with_silhouette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Annotating Tree with Silhouette Images and Sub-plots</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_ggtree_tree_objects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">ggtree for Other Tree-like Objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_ggtree_exts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part III: ggtree extensions</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11_ggtree_exts-others.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Other ggtree Extensions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12_ggtree_utilities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part IV: Miscellaneous topics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13_ggtree_gallery.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Gallery of Reproducible Examples</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendix</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-faq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(APPENDIX) Appendix</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Related Tools</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-fig-tab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Figures and Tables</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-tidytree" id="toc-sec-tidytree" class="nav-link active" data-scroll-target="#sec-tidytree"><span class="header-section-number">2.1</span> Manipulating Tree Data Using Tidy Interface</a>
  <ul class="collapse">
  <li><a href="#the-phylo-object" id="toc-the-phylo-object" class="nav-link" data-scroll-target="#the-phylo-object"><span class="header-section-number">2.1.1</span> The <code>phylo</code> object</a></li>
  <li><a href="#the-treedata-object" id="toc-the-treedata-object" class="nav-link" data-scroll-target="#the-treedata-object"><span class="header-section-number">2.1.2</span> The <code>treedata</code> object</a></li>
  <li><a href="#sec-accesor-tidytree" id="toc-sec-accesor-tidytree" class="nav-link" data-scroll-target="#sec-accesor-tidytree"><span class="header-section-number">2.1.3</span> Access related nodes</a></li>
  </ul></li>
  <li><a href="#data-integration" id="toc-data-integration" class="nav-link" data-scroll-target="#data-integration"><span class="header-section-number">2.2</span> Data Integration</a>
  <ul class="collapse">
  <li><a href="#sec-merge-tree" id="toc-sec-merge-tree" class="nav-link" data-scroll-target="#sec-merge-tree"><span class="header-section-number">2.2.1</span> Combining tree data</a></li>
  <li><a href="#sec-link-external-data" id="toc-sec-link-external-data" class="nav-link" data-scroll-target="#sec-link-external-data"><span class="header-section-number">2.2.2</span> Linking external data to phylogeny</a></li>
  <li><a href="#grouping-taxa" id="toc-grouping-taxa" class="nav-link" data-scroll-target="#grouping-taxa"><span class="header-section-number">2.2.3</span> Grouping taxa</a></li>
  </ul></li>
  <li><a href="#sec-reroot-treeio" id="toc-sec-reroot-treeio" class="nav-link" data-scroll-target="#sec-reroot-treeio"><span class="header-section-number">2.3</span> Rerooting tree</a></li>
  <li><a href="#sec-rescale-treeio" id="toc-sec-rescale-treeio" class="nav-link" data-scroll-target="#sec-rescale-treeio"><span class="header-section-number">2.4</span> Rescaling Tree Branches</a></li>
  <li><a href="#subsetting-tree-with-data" id="toc-subsetting-tree-with-data" class="nav-link" data-scroll-target="#subsetting-tree-with-data"><span class="header-section-number">2.5</span> Subsetting Tree with Data</a>
  <ul class="collapse">
  <li><a href="#sec-remove-tip" id="toc-sec-remove-tip" class="nav-link" data-scroll-target="#sec-remove-tip"><span class="header-section-number">2.5.1</span> Removing tips in a phylogenetic tree</a></li>
  <li><a href="#sec-subset-tip" id="toc-sec-subset-tip" class="nav-link" data-scroll-target="#sec-subset-tip"><span class="header-section-number">2.5.2</span> Subsetting tree by tip label</a></li>
  <li><a href="#sec-subset-node" id="toc-sec-subset-node" class="nav-link" data-scroll-target="#sec-subset-node"><span class="header-section-number">2.5.3</span> Subsetting tree by internal node number</a></li>
  </ul></li>
  <li><a href="#sec-ggtree-fortify" id="toc-sec-ggtree-fortify" class="nav-link" data-scroll-target="#sec-ggtree-fortify"><span class="header-section-number">2.6</span> Manipulating Tree Data for Visualization</a></li>
  <li><a href="#sec-summary2" id="toc-sec-summary2" class="nav-link" data-scroll-target="#sec-summary2"><span class="header-section-number">2.7</span> Summary</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/YuLab-SMU/treedata-book/edit/main/02_tidytree.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/blob/main/02_tidytree.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-chapter2" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Manipulating Tree with Data</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div style="page-break-after: always;"></div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(knitr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">message=</span><span class="cn">FALSE</span>, <span class="at">warning=</span><span class="cn">FALSE</span>, <span class="at">eval=</span><span class="cn">TRUE</span>, <span class="at">echo=</span><span class="cn">TRUE</span>, <span class="at">cache=</span><span class="cn">TRUE</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(yulab.utils)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"software-link.R"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ape"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"tidytree"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(treeio)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">show_data_for_treedata =</span> <span class="cn">FALSE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="sec-tidytree" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="sec-tidytree"><span class="header-section-number">2.1</span> Manipulating Tree Data Using Tidy Interface</h2>
<p>All the tree data parsed/merged by <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> can be converted to a tidy data frame using the <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> package. The <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> package provides tidy interfaces to manipulate trees with associated data. For instance, external data can be linked to phylogeny or evolutionary data obtained from different sources can be merged using tidyverse verbs. After the tree data was processed, it can be converted back to a <code>treedata</code> object and exported to <a href="#chapter3">a single tree file</a>, further analyzed in R or visualized using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span> and <a href="http://bioconductor.org/packages/ggtreeExtra"><strong>ggtreeExtra</strong></a> <span class="citation" data-cites="ggtreeExtra_2021">(<a href="#ref-ggtreeExtra_2021" role="doc-biblioref">Xu et al., 2021</a>)</span>.</p>
<section id="the-phylo-object" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="the-phylo-object"><span class="header-section-number">2.1.1</span> The <code>phylo</code> object</h3>
<p>The <code>phylo</code> class defined in the <a href="https://CRAN.R-project.org/package=ape"><strong>ape</strong></a> package <span class="citation" data-cites="paradis_ape_2004">(<a href="#ref-paradis_ape_2004" role="doc-biblioref">Paradis et al., 2004</a>)</span> is fundamental for phylogenetic analysis in R. Most of the R packages in <a href="https://CRAN.R-project.org/view=Phylogenetics">this field</a> rely extensively on the <code>phylo</code> object. The <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> package provides <code>as_tibble</code> method to convert the <code>phylo</code> object to a tidy data frame, a <code>tbl_tree</code> object.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2017</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="dv">4</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>tree</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 4 tips and 3 internal nodes.

Tip labels:
  t4, t1, t3, t2

Rooted; includes branch length(s).</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(tree)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>x</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_tree abstraction: 7 × 4
# which can be converted to treedata or phylo 
# via as.treedata or as.phylo
  parent  node branch.length label
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;
1      5     1       0.435   t4   
2      7     2       0.674   t1   
3      7     3       0.00202 t3   
4      6     4       0.0251  t2   
5      5     5      NA       &lt;NA&gt; 
6      5     6       0.472   &lt;NA&gt; 
7      6     7       0.274   &lt;NA&gt; </code></pre>
</div>
</div>
<p>The <code>tbl_tree</code> object can be converted back to a <code>phylo</code> object using the <code>as.phylo()</code> method.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.phylo</span>(x)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 4 tips and 3 internal nodes.

Tip labels:
  t4, t1, t3, t2

Rooted; includes branch length(s).</code></pre>
</div>
</div>
<p>Using <code>tbl_tree</code> object makes tree and data manipulation more effective and easier (see also the example in <a href="#bind-tip">FAQ</a>). For example, we can link evolutionary trait to phylogeny using the <a href="https://CRAN.R-project.org/package=dplyr"><strong>dplyr</strong></a> verbs <code>full_join()</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">'t'</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>),</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">trait =</span> <span class="fu">rnorm</span>(<span class="dv">4</span>))</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">full_join</span>(x, d, <span class="at">by =</span> <span class="st">'label'</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>y</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_tree abstraction: 7 × 5
# which can be converted to treedata or phylo 
# via as.treedata or as.phylo
  parent  node branch.length label  trait
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt;
1      5     1       0.435   t4     0.943
2      7     2       0.674   t1    -0.171
3      7     3       0.00202 t3     0.570
4      6     4       0.0251  t2    -0.283
5      5     5      NA       &lt;NA&gt;  NA    
6      5     6       0.472   &lt;NA&gt;  NA    
7      6     7       0.274   &lt;NA&gt;  NA    </code></pre>
</div>
</div>
</section>
<section id="the-treedata-object" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="the-treedata-object"><span class="header-section-number">2.1.2</span> The <code>treedata</code> object</h3>
<p>The <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> package defines <code>treedata</code> class to store a phylogenetic tree with associated data. After mapping external data to the tree structure, the <code>tbl_tree</code> object can be converted to a <code>treedata</code> object.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.treedata</span>(y)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 4 tips and 3 internal nodes.

Tip labels:
  t4, t1, t3, t2

Rooted; includes branch length(s).

with the following features available:
  'trait'.</code></pre>
</div>
</div>
<p>The <code>treedata</code> class is used in the <a href="https://bioconductor.org/packages/treeio/">treeio</a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> to store evolutionary evidence inferred by commonly used software (<a href="http://beast2.org/"><strong>BEAST</strong></a>, <a href="http://sco.h-its.org/exelixis/web/software/epa/index.html"><strong>EPA</strong></a>, <a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a>, <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a>, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a>, <a href="http://pbil.univ-lyon1.fr/software/phyldog/"><strong>PHYLDOG</strong></a>, <a href="http://matsen.fhcrc.org/pplacer/"><strong>PPLACER</strong></a>, <a href="http://loco.biosci.arizona.edu/r8s/"><strong>r8s</strong></a>, <a href="http://evomics.org/learning/phylogenetics/raxml/"><strong>RAxML</strong></a>, and <a href="http://revbayes.github.io/intro.html"><strong>RevBayes</strong></a>, etc.) (see details in <a href="01_treeio_importing_tree.html#chapter1">Chapter 1</a>).</p>
<p>The <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> package also provides the <code>as_tibble()</code> method to convert a <code>treedata</code> object to a tidy data frame. The phylogenetic tree structure and the evolutionary inferences were stored in the <code>tbl_tree</code> object, making it consistent and easier for manipulating evolutionary statistics inferred by different software as well as linking external data to the same tree structure.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>y <span class="sc">%&gt;%</span> as.treedata <span class="sc">%&gt;%</span> as_tibble</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_tree abstraction: 7 × 5
# which can be converted to treedata or phylo 
# via as.treedata or as.phylo
  parent  node branch.length label  trait
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt;
1      5     1       0.435   t4     0.943
2      7     2       0.674   t1    -0.171
3      7     3       0.00202 t3     0.570
4      6     4       0.0251  t2    -0.283
5      5     5      NA       &lt;NA&gt;  NA    
6      5     6       0.472   &lt;NA&gt;  NA    
7      6     7       0.274   &lt;NA&gt;  NA    </code></pre>
</div>
</div>
</section>
<section id="sec-accesor-tidytree" class="level3" data-number="2.1.3">
<h3 data-number="2.1.3" class="anchored" data-anchor-id="sec-accesor-tidytree"><span class="header-section-number">2.1.3</span> Access related nodes</h3>
<p>The <a href="https://CRAN.R-project.org/package=dplyr"><strong>dplyr</strong></a> verbs can be applied to <code>tbl_tree</code> directly to manipulate tree data. In addition, <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> provides several verbs to filter related nodes, including <code>child()</code>, <code>parent()</code>, <code>offspring()</code>, <code>ancestor()</code>, <code>sibling()</code> and <code>MRCA()</code>.</p>
<p>These verbs accept a <code>tbl_tree</code> object and a selected node which can be node number or label.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">child</span>(y, <span class="dv">5</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  parent  node branch.length label  trait
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt;
1      5     1         0.435 t4     0.943
2      5     6         0.472 &lt;NA&gt;  NA    </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">parent</span>(y, <span class="dv">2</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  parent  node branch.length label trait
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt;
1      6     7         0.274 &lt;NA&gt;     NA</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">offspring</span>(y, <span class="dv">5</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  parent  node branch.length label  trait
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt;
1      5     1       0.435   t4     0.943
2      7     2       0.674   t1    -0.171
3      7     3       0.00202 t3     0.570
4      6     4       0.0251  t2    -0.283
5      5     6       0.472   &lt;NA&gt;  NA    
6      6     7       0.274   &lt;NA&gt;  NA    </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ancestor</span>(y, <span class="dv">2</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 5
  parent  node branch.length label trait
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt;
1      5     5        NA     &lt;NA&gt;     NA
2      5     6         0.472 &lt;NA&gt;     NA
3      6     7         0.274 &lt;NA&gt;     NA</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sibling</span>(y, <span class="dv">2</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  parent  node branch.length label trait
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt;
1      7     3       0.00202 t3    0.570</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MRCA</span>(y, <span class="dv">2</span>, <span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  parent  node branch.length label trait
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt;
1      6     7         0.274 &lt;NA&gt;     NA</code></pre>
</div>
</div>
<p>All these methods are also implemented in <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> for working with <code>phylo</code> and <code>treedata</code> objects. You can try accessing related nodes using the tree object. For instance, the following command will output child nodes of the selected internal node <code>5</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">child</span>(tree, <span class="dv">5</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1 6</code></pre>
</div>
</div>
<p>Beware that the methods for tree objects output relevant node numbers, while the methods for <code>tbl_tree</code> object output a <code>tibble</code> object that contains related information.</p>
</section>
</section>
<section id="data-integration" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="data-integration"><span class="header-section-number">2.2</span> Data Integration</h2>
<section id="sec-merge-tree" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="sec-merge-tree"><span class="header-section-number">2.2.1</span> Combining tree data</h3>
<p>The <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> serves as an infrastructure that enables various types of phylogenetic data inferred from common analysis programs to be imported and used in R. For instance, <em>d<sub>N</sub>/d<sub>S</sub></em> or ancestral sequences estimated by <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>, and clade support values (posterior) inferred by <a href="http://beast2.org/"><strong>BEAST</strong></a>/<a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a>. In addition, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> supports linking external data to phylogeny. It brings these external phylogenetic data (either from software output or external sources) to the R community and makes it available for further analysis in R. Furthermore, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> can combine multiple phylogenetic trees into one with their node/branch-specific attribute data. Essentially, as a result, one such attribute (<em>e.g.</em>, substitution rate) can be mapped to another attribute (<em>e.g.</em>, <em>d<sub>N</sub>/d<sub>S</sub></em>) of the same node/branch for comparison and further computations <span class="citation" data-cites="yu_ggtree:_2017 yu_two_2018">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>; <a href="#ref-yu_two_2018" role="doc-biblioref">Yu et al., 2018</a>)</span>.</p>
<p>A previously published dataset, seventy-six H3 hemagglutinin gene sequences of a lineage containing swine and human influenza A viruses <span class="citation" data-cites="liang_expansion_2014">(<a href="#ref-liang_expansion_2014" role="doc-biblioref">Liang et al., 2014</a>)</span>, was used here to demonstrate the utilities of comparing evolutionary statistics inferred by different software. The dataset was re-analyzed by <a href="http://beast2.org/"><strong>BEAST</strong></a> for timescale estimation and <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> for synonymous and non-synonymous substitution estimation. In this example, we first parsed the outputs from <a href="http://beast2.org/"><strong>BEAST</strong></a> using the <code>read.beast()</code> function and from <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> using the <code>read.codeml()</code> function into two <code>treedata</code> objects. Then these two objects containing separate sets of node/branch-specific data were merged via the <code>merge_tree()</code> function.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>beast_file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"examples/MCC_FluA_H3.tree"</span>, <span class="at">package=</span><span class="st">"ggtree"</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>rst_file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"examples/rst"</span>, <span class="at">package=</span><span class="st">"ggtree"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>mlc_file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"examples/mlc"</span>, <span class="at">package=</span><span class="st">"ggtree"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>beast_tree <span class="ot">&lt;-</span> <span class="fu">read.beast</span>(beast_file)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>codeml_tree <span class="ot">&lt;-</span> <span class="fu">read.codeml</span>(rst_file, mlc_file)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>merged_tree <span class="ot">&lt;-</span> <span class="fu">merge_tree</span>(beast_tree, codeml_tree)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>merged_tree</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/ggtree/examples/MCC_FluA_H3.tree',
    '/home/runner/work/_temp/Library/ggtree/examples/rst',
    '/home/runner/work/_temp/Library/ggtree/examples/mlc'.

...@ phylo:

Phylogenetic tree with 76 tips and 75 internal nodes.

Tip labels:
  A/Hokkaido/30-1-a/2013, A/New_York/334/2004, A/New_York/463/2005,
A/New_York/452/1999, A/New_York/238/2005, A/New_York/523/1998, ...

Rooted; includes branch length(s).

with the following features available:
  'height', 'height_0.95_HPD', 'height_median', 'height_range', 'length',
'length_0.95_HPD', 'length_median', 'length_range', 'posterior', 'rate',
'rate_0.95_HPD', 'rate_median', 'rate_range', 'subs', 'AA_subs', 't', 'N',
'S', 'dN_vs_dS', 'dN', 'dS', 'N_x_dN', 'S_x_dS'.</code></pre>
</div>
</div>
<p>After merging the <code>beast_tree</code> and <code>codeml_tree</code> objects, all node/branch-specific data imported from <a href="http://beast2.org/"><strong>BEAST</strong></a> and <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> output files are all available in the <code>merged_tree</code> object. The tree object was converted to a tidy data frame using the <a href="https://cran.r-project.org/package=tidytree">tidytree</a> package and visualized as hexbin scatterplots of <em>d<sub>N</sub>/d<sub>S</sub></em>, <em>d<sub>N</sub></em>, and <em>d<sub>S</sub></em> inferred by <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> vs.&nbsp;<em>rate</em> (substitution rate in a unit of substitutions/site/year) inferred by <a href="http://beast2.org/"><strong>BEAST</strong></a> on the same branches.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> merged_tree <span class="sc">%&gt;%</span> </span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(dN_vs_dS, dN, dS, rate) <span class="sc">%&gt;%</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset</span>(dN_vs_dS <span class="sc">&gt;=</span><span class="dv">0</span> <span class="sc">&amp;</span> dN_vs_dS <span class="sc">&lt;=</span> <span class="fl">1.5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">gather</span>(type, value, dN_vs_dS<span class="sc">:</span>dS)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>type[df<span class="sc">$</span>type <span class="sc">==</span> <span class="st">'dN_vs_dS'</span>] <span class="ot">&lt;-</span> <span class="st">'dN/dS'</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>type <span class="ot">&lt;-</span> <span class="fu">factor</span>(df<span class="sc">$</span>type, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"dN/dS"</span>, <span class="st">"dN"</span>, <span class="st">"dS"</span>))</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(rate, value)) <span class="sc">+</span> <span class="fu">geom_hex</span>() <span class="sc">+</span> </span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>type, <span class="at">scale=</span><span class="st">'free_y'</span>) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="02_tidytree_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Correlation of <em>d<sub>N</sub>/d<sub>S</sub></em>, <em>d<sub>N</sub></em>, and <em>d<sub>S</sub></em> vs.&nbsp;substitution rate.</strong> After merging the <em>BEAST</em> and <em>CodeML</em> outputs, the branch-specific estimates (substitution rate, <em>d<sub>N</sub>/d<sub>S</sub></em> , <em>d<sub>N</sub></em>, and <em>d<sub>S</sub></em>) from the two analysis programs are compared on the same branch basis. The associations of <em>d<sub>N</sub>/d<sub>S</sub></em>, <em>d<sub>N</sub></em>, and <em>d<sub>S</sub></em> vs.&nbsp;<em>rate</em> are visualized in hexbin scatter plots.</figcaption>
</figure>
</div>
</div>
</div>
<p>The output is illustrated in Figure <span class="quarto-unresolved-ref">?fig-correlations</span>. We can then test the association of these node/branch-specific data using Pearson correlation, which in this case showed that <em>d<sub>N</sub></em> and <em>d<sub>S</sub></em>, but not <em>d<sub>N</sub>/d<sub>S</sub></em> are significantly (<em>p</em>-values) associated with <em>rate</em>.</p>
<p>Using the <code>merge_tree()</code> function, we are able to compare analysis results using an identical model from different software packages or different models using different or identical software. It also allows users to integrate different analysis findings from different software packages. Merging tree data is not restricted to software findings, associating external data to analysis findings is also granted. The <code>merge_tree()</code> function is chainable and allows several tree objects to be merged into one.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>phylo <span class="ot">&lt;-</span> <span class="fu">as.phylo</span>(beast_tree)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="fu">Nnode2</span>(phylo)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">node =</span> <span class="dv">1</span><span class="sc">:</span>N, <span class="at">fake_trait =</span> <span class="fu">rnorm</span>(N), <span class="at">another_trait =</span> <span class="fu">runif</span>(N))</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>fake_tree <span class="ot">&lt;-</span> <span class="fu">treedata</span>(<span class="at">phylo =</span> phylo, <span class="at">data =</span> d)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>triple_tree <span class="ot">&lt;-</span> <span class="fu">merge_tree</span>(merged_tree, fake_tree)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>triple_tree</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/ggtree/examples/MCC_FluA_H3.tree',
    '/home/runner/work/_temp/Library/ggtree/examples/rst',
    '/home/runner/work/_temp/Library/ggtree/examples/mlc'.

...@ phylo:

Phylogenetic tree with 76 tips and 75 internal nodes.

Tip labels:
  A/Hokkaido/30-1-a/2013, A/New_York/334/2004, A/New_York/463/2005,
A/New_York/452/1999, A/New_York/238/2005, A/New_York/523/1998, ...

Rooted; includes branch length(s).

with the following features available:
  'height', 'height_0.95_HPD', 'height_median', 'height_range', 'length',
'length_0.95_HPD', 'length_median', 'length_range', 'posterior', 'rate',
'rate_0.95_HPD', 'rate_median', 'rate_range', 'subs', 'AA_subs', 't', 'N',
'S', 'dN_vs_dS', 'dN', 'dS', 'N_x_dN', 'S_x_dS', 'fake_trait', 'another_trait'.</code></pre>
</div>
</div>
<p>The <code>triple_tree</code> object shown above contains analysis results obtained from <a href="http://beast2.org/"><strong>BEAST</strong></a> and <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>, and evolutionary traits from external sources. All these pieces of information can be used to annotate the tree using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span> and <a href="http://bioconductor.org/packages/ggtreeExtra"><strong>ggtreeExtra</strong></a> <span class="citation" data-cites="ggtreeExtra_2021">(<a href="#ref-ggtreeExtra_2021" role="doc-biblioref">Xu et al., 2021</a>)</span>.</p>
</section>
<section id="sec-link-external-data" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="sec-link-external-data"><span class="header-section-number">2.2.2</span> Linking external data to phylogeny</h3>
<p>In addition to analysis findings that are associated with the tree as demonstrated above, there is a wide range of heterogeneous data, including phenotypic data, experimental data, and clinical data, <em>etc.</em>, that need to be integrated and linked to phylogeny. For example, in the study of viral evolution, tree nodes may be associated with epidemiological information, such as location, age, and subtype. Functional annotations may need to be mapped onto gene trees for comparative genomics studies. To facilitate data integration, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> provides <code>full_join()</code> methods to link external data to phylogeny and store it in either <code>phylo</code> or <code>treedata</code> object. Beware that linking external data to a <code>phylo</code> object will produce a <code>treedata</code> object to store the input <code>phylo</code> with associated data. The <code>full_join</code> methods can also be used at tidy data frame level (<em>i.e.</em>, <code>tbl_tree</code> object described previously) and at <code>ggtree</code> level (described in <a href="#attach-operator">Chapter 7</a>) <span class="citation" data-cites="yu_two_2018">(<a href="#ref-yu_two_2018" role="doc-biblioref">Yu et al., 2018</a>)</span>.</p>
<p>The following example calculated bootstrap values and merged those values with the tree (a <code>phylo</code> object) by matching their node numbers.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(woodmouse)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist.dna</span>(woodmouse)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>tr <span class="ot">&lt;-</span> <span class="fu">nj</span>(d)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>bp <span class="ot">&lt;-</span> <span class="fu">boot.phylo</span>(tr, woodmouse, <span class="cf">function</span>(x) <span class="fu">nj</span>(<span class="fu">dist.dna</span>(x)))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Running bootstraps:       100 / 100
Calculating bootstrap values... done.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>bp2 <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">node=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">Nnode</span>(tr) <span class="sc">+</span> <span class="fu">Ntip</span>(tr), <span class="at">bootstrap =</span> bp)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">full_join</span>(tr, bp2, <span class="at">by=</span><span class="st">"node"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 15 tips and 13 internal nodes.

Tip labels:
  No305, No304, No306, No0906S, No0908S, No0909S, ...

Unrooted; includes branch length(s).

with the following features available:
  '', 'bootstrap'.</code></pre>
</div>
</div>
<p>Another example demonstrates merging evolutionary traits with the tree (a <code>treedata</code> object) by matching their tip labels.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/BEAST"</span>, <span class="st">"beast_mcc.tree"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>beast <span class="ot">&lt;-</span> <span class="fu">read.beast</span>(file)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">label =</span> <span class="fu">as.phylo</span>(beast)<span class="sc">$</span>tip.label, <span class="at">trait =</span> <span class="fu">rnorm</span>(<span class="fu">Ntip</span>(beast)))</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="fu">full_join</span>(beast, x, <span class="at">by=</span><span class="st">"label"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object that stored information
of
    '/home/runner/work/_temp/Library/treeio/extdata/BEAST/beast_mcc.tree'.

...@ phylo:

Phylogenetic tree with 15 tips and 14 internal nodes.

Tip labels:
  A_1995, B_1996, C_1995, D_1987, E_1996, F_1997, ...

Rooted; includes branch length(s).

with the following features available:
  'height', 'height_0.95_HPD', 'height_median', 'height_range', 'length',
'length_0.95_HPD', 'length_median', 'length_range', 'posterior', 'rate',
'rate_0.95_HPD', 'rate_median', 'rate_range', 'trait'.</code></pre>
</div>
</div>
<p>Manipulating tree objects is frustrated with the fragmented functions available for working with <code>phylo</code> objects, not to mention linking external data to the phylogeny structure. With the <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span>, it is easy to combine tree data from various sources. In addition, with the <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> package, manipulating trees is easier using the <a href="https://www.jstatsoft.org/article/view/v059i10">tidy data principles</a> and consistent with tools already in wide use, including <a href="https://CRAN.R-project.org/package=dplyr"><strong>dplyr</strong></a>, <a href="https://CRAN.R-project.org/package=tidyr"><strong>tidyr</strong></a>, <a href="https://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a>, and <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span>.</p>
</section>
<section id="grouping-taxa" class="level3" data-number="2.2.3">
<h3 data-number="2.2.3" class="anchored" data-anchor-id="grouping-taxa"><span class="header-section-number">2.2.3</span> Grouping taxa</h3>
<p>The <code>groupOTU()</code> and <code>groupClade()</code> methods are designed for adding taxa grouping information to the input tree object. The methods were implemented in <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a>, <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a>, and <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> respectively to support adding grouping information for the <code>tbl_tree</code>, <code>phylo</code> and <code>treedata</code>, and <code>ggtree</code> objects. This grouping information can be used directly in tree visualization (<em>e.g.</em>, <a href="#group-taxa-vis">coloring a tree based on grouping information</a>) with <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> (Figure <span class="quarto-unresolved-ref">?fig-groupOTU</span>).</p>
<section id="groupclade" class="level4" data-number="2.2.3.1">
<h4 data-number="2.2.3.1" class="anchored" data-anchor-id="groupclade"><span class="header-section-number">2.2.3.1</span> groupClade</h4>
<p>The <code>groupClade()</code> method accepts an internal node or a vector of internal nodes to add grouping information of selected clade/clades.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>nwk <span class="ot">&lt;-</span> <span class="st">'(((((((A:4,B:4):6,C:5):8,D:6):3,E:21):10,((F:4,G:12):14,H:8):13):</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="st">        13,((I:5,J:2):30,(K:11,L:11):2):17):4,M:56);'</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(<span class="at">text=</span>nwk)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="fu">groupClade</span>(<span class="fu">as_tibble</span>(tree), <span class="fu">c</span>(<span class="dv">17</span>, <span class="dv">21</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_tree abstraction: 25 × 5
# which can be converted to treedata or phylo 
# via as.treedata or as.phylo
   parent  node branch.length label group
    &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;fct&gt;
 1     20     1             4 A     1    
 2     20     2             4 B     1    
 3     19     3             5 C     1    
 4     18     4             6 D     1    
 5     17     5            21 E     1    
 6     22     6             4 F     2    
 7     22     7            12 G     2    
 8     21     8             8 H     2    
 9     24     9             5 I     0    
10     24    10             2 J     0    
# ℹ 15 more rows</code></pre>
</div>
</div>
</section>
<section id="groupotu" class="level4" data-number="2.2.3.2">
<h4 data-number="2.2.3.2" class="anchored" data-anchor-id="groupotu"><span class="header-section-number">2.2.3.2</span> groupOTU</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2017</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>tr <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="dv">4</span>)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(tr)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="do">## the input nodes can be node ID or label</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="fu">groupOTU</span>(x, <span class="fu">c</span>(<span class="st">'t1'</span>, <span class="st">'t4'</span>), <span class="at">group_name =</span> <span class="st">"fake_group"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_tree abstraction: 7 × 5
# which can be converted to treedata or phylo 
# via as.treedata or as.phylo
  parent  node branch.length label fake_group
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;fct&gt;     
1      5     1       0.435   t4    1         
2      7     2       0.674   t1    1         
3      7     3       0.00202 t3    0         
4      6     4       0.0251  t2    0         
5      5     5      NA       &lt;NA&gt;  1         
6      5     6       0.472   &lt;NA&gt;  1         
7      6     7       0.274   &lt;NA&gt;  1         </code></pre>
</div>
</div>
<p>Both <code>groupClade()</code> and <code>groupOTU()</code> work with the <code>tbl_tree</code>, <code>phylo</code> and <code>treedata</code>, and <code>ggtree</code> objects. Here is an example of using <code>groupOTU()</code> with a <code>phylo</code> tree object.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">groupOTU</span>(tr, <span class="fu">c</span>(<span class="st">'t2'</span>, <span class="st">'t4'</span>), <span class="at">group_name =</span> <span class="st">"fake_group"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  as_tibble</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_tree abstraction: 7 × 5
# which can be converted to treedata or phylo 
# via as.treedata or as.phylo
  parent  node branch.length label fake_group
   &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;fct&gt;     
1      5     1       0.435   t4    1         
2      7     2       0.674   t1    0         
3      7     3       0.00202 t3    0         
4      6     4       0.0251  t2    1         
5      5     5      NA       &lt;NA&gt;  1         
6      5     6       0.472   &lt;NA&gt;  1         
7      6     7       0.274   &lt;NA&gt;  0         </code></pre>
</div>
</div>
<p>Another example of working with the <code>ggtree</code> object can be found in <a href="#group-taxa-vis">session 6.4</a>.</p>
<p>The <code>groupOTU</code> will trace back from input nodes to most recent common ancestor. In this example, nodes 1, 4, 5 and 6 are grouping together (<code>4 (t2) -&gt; 6 -&gt; 5</code> and <code>1 (t4) -&gt; 5</code>).</p>
<p>Related operational taxonomic units (OTUs) are grouping and they are not necessarily within a clade. They can be monophyletic (clade), polyphyletic or paraphyletic.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>cls <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">c1=</span><span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>, <span class="st">"E"</span>),</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">c2=</span><span class="fu">c</span>(<span class="st">"F"</span>, <span class="st">"G"</span>, <span class="st">"H"</span>),</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">c3=</span><span class="fu">c</span>(<span class="st">"L"</span>, <span class="st">"K"</span>, <span class="st">"I"</span>, <span class="st">"J"</span>),</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">c4=</span><span class="st">"M"</span>)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(tree) <span class="sc">%&gt;%</span> <span class="fu">groupOTU</span>(cls)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_tree abstraction: 25 × 5
# which can be converted to treedata or phylo 
# via as.treedata or as.phylo
   parent  node branch.length label group
    &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt; &lt;fct&gt;
 1     20     1             4 A     c1   
 2     20     2             4 B     c1   
 3     19     3             5 C     c1   
 4     18     4             6 D     c1   
 5     17     5            21 E     c1   
 6     22     6             4 F     c2   
 7     22     7            12 G     c2   
 8     21     8             8 H     c2   
 9     24     9             5 I     c3   
10     24    10             2 J     c3   
# ℹ 15 more rows</code></pre>
</div>
</div>
<p>If there are conflicts when tracing back to the most recent common ancestor, users can set <code>overlap</code> parameter to “origin” (the first one counts), “overwrite” (default, the last one counts), or “abandon” (un-selected for grouping)<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p>
</section>
</section>
</section>
<section id="sec-reroot-treeio" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="sec-reroot-treeio"><span class="header-section-number">2.3</span> Rerooting tree</h2>
<p>A phylogenetic tree can be rerooted with a specified <code>outgroup</code>. The <a href="https://CRAN.R-project.org/package=ape"><strong>ape</strong></a> package implements a <code>root()</code> method to reroot a tree stored in a <code>phylo</code> object, while the <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package provides the <code>root()</code> method for <code>treedata</code> object. This method is designed to re-root a phylogenetic tree with associated data concerning the specified <code>outgroup</code> or at the specified <code>node</code> based on the <code>root()</code> implemented in the <a href="https://CRAN.R-project.org/package=ape"><strong>ape</strong></a> package.</p>
<p>We first linked external data to a tree using <code>left_join()</code> and stored all the information in a <code>treedata</code> object, <code>trda</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(treeio)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidytree)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TDbook)</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="co"># load `tree_boots`, `df_tip_data`, and `df_inode_data` from 'TDbook'</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>trda <span class="ot">&lt;-</span> tree_boots <span class="sc">%&gt;%</span> </span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">left_join</span>(df_tip_data, <span class="at">by=</span><span class="fu">c</span>(<span class="st">"label"</span> <span class="ot">=</span> <span class="st">"Newick_label"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">left_join</span>(df_inode_data, <span class="at">by=</span><span class="fu">c</span>(<span class="st">"label"</span> <span class="ot">=</span> <span class="st">"newick_label"</span>))</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>trda</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 7 tips and 6 internal nodes.

Tip labels:
  Rangifer_tarandus, Cervus_elaphus, Bos_taurus, Ovis_orientalis,
Suricata_suricatta, Cystophora_cristata, ...
Node labels:
  Mammalia, Artiodactyla, Cervidae, Bovidae, Carnivora, Caniformia

Rooted; includes branch length(s).

with the following features available:
  '', 'vernacularName', 'imageURL', 'imageLicense', 'imageAuthor', 'infoURL',
'mass_in_kg', 'trophic_habit', 'ncbi_taxid', 'rank', 'vernacularName.y',
'infoURL.y', 'rank.y', 'bootstrap', 'posterior'.</code></pre>
</div>
</div>
<p>Then we can reroot the tree with the associated data mapping to the branches and nodes correctly as demonstrated in Figure <span class="quarto-unresolved-ref">?fig-reroot</span>. The figure was visualized using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> (see also Chapters <a href="#chapter4">4</a> and <a href="#chapter5">5</a>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># reroot</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>trda2 <span class="ot">&lt;-</span> <span class="fu">root</span>(trda, <span class="at">outgroup =</span> <span class="st">"Suricata_suricatta"</span>, <span class="at">edgelabel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The original tree</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> trda <span class="sc">%&gt;%</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggtree</span>() <span class="sc">+</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_nodelab</span>(</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">x =</span> branch,</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">label =</span> bootstrap</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">nudge_y =</span> <span class="fl">0.36</span></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>      <span class="fu">xlim</span>(<span class="sc">-</span>.<span class="dv">1</span>, <span class="fl">4.5</span>) <span class="sc">+</span></span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_tippoint</span>(</span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a>          <span class="at">shape =</span> trophic_habit, </span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>          <span class="at">color =</span> trophic_habit, </span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>          <span class="at">size =</span> mass_in_kg</span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_size_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_tiplab</span>(</span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">offset =</span> .<span class="dv">14</span>, </span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_nodelab</span>(</span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a>          <span class="at">label =</span> vernacularName.y, </span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a>          <span class="at">fill =</span> posterior</span>
<span id="cb51-29"><a href="#cb51-29" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb51-30"><a href="#cb51-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">geom =</span> <span class="st">"label"</span></span>
<span id="cb51-31"><a href="#cb51-31" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span> </span>
<span id="cb51-32"><a href="#cb51-32" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"YlGnBu"</span>)) <span class="sc">+</span></span>
<span id="cb51-33"><a href="#cb51-33" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)  </span>
<span id="cb51-34"><a href="#cb51-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-35"><a href="#cb51-35" aria-hidden="true" tabindex="-1"></a><span class="co"># after reroot</span></span>
<span id="cb51-36"><a href="#cb51-36" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> trda2 <span class="sc">%&gt;%</span></span>
<span id="cb51-37"><a href="#cb51-37" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggtree</span>() <span class="sc">+</span></span>
<span id="cb51-38"><a href="#cb51-38" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_nodelab</span>(</span>
<span id="cb51-39"><a href="#cb51-39" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb51-40"><a href="#cb51-40" aria-hidden="true" tabindex="-1"></a>          <span class="at">x =</span> branch,</span>
<span id="cb51-41"><a href="#cb51-41" aria-hidden="true" tabindex="-1"></a>          <span class="at">label =</span> bootstrap</span>
<span id="cb51-42"><a href="#cb51-42" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb51-43"><a href="#cb51-43" aria-hidden="true" tabindex="-1"></a>        <span class="at">nudge_y =</span> <span class="fl">0.36</span></span>
<span id="cb51-44"><a href="#cb51-44" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb51-45"><a href="#cb51-45" aria-hidden="true" tabindex="-1"></a>      <span class="fu">xlim</span>(<span class="sc">-</span>.<span class="dv">1</span>, <span class="fl">5.5</span>) <span class="sc">+</span></span>
<span id="cb51-46"><a href="#cb51-46" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_tippoint</span>(</span>
<span id="cb51-47"><a href="#cb51-47" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb51-48"><a href="#cb51-48" aria-hidden="true" tabindex="-1"></a>          <span class="at">shape =</span> trophic_habit,</span>
<span id="cb51-49"><a href="#cb51-49" aria-hidden="true" tabindex="-1"></a>          <span class="at">color =</span> trophic_habit,</span>
<span id="cb51-50"><a href="#cb51-50" aria-hidden="true" tabindex="-1"></a>          <span class="at">size =</span> mass_in_kg</span>
<span id="cb51-51"><a href="#cb51-51" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb51-52"><a href="#cb51-52" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb51-53"><a href="#cb51-53" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_size_continuous</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb51-54"><a href="#cb51-54" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_tiplab</span>(</span>
<span id="cb51-55"><a href="#cb51-55" aria-hidden="true" tabindex="-1"></a>        <span class="at">offset =</span> .<span class="dv">14</span>,</span>
<span id="cb51-56"><a href="#cb51-56" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb51-57"><a href="#cb51-57" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_nodelab</span>(</span>
<span id="cb51-58"><a href="#cb51-58" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb51-59"><a href="#cb51-59" aria-hidden="true" tabindex="-1"></a>          <span class="at">label =</span> vernacularName.y,</span>
<span id="cb51-60"><a href="#cb51-60" aria-hidden="true" tabindex="-1"></a>          <span class="at">fill =</span> posterior</span>
<span id="cb51-61"><a href="#cb51-61" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb51-62"><a href="#cb51-62" aria-hidden="true" tabindex="-1"></a>        <span class="at">geom =</span> <span class="st">"label"</span></span>
<span id="cb51-63"><a href="#cb51-63" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb51-64"><a href="#cb51-64" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_fill_gradientn</span>(<span class="at">colors =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"YlGnBu"</span>)) <span class="sc">+</span></span>
<span id="cb51-65"><a href="#cb51-65" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span>
<span id="cb51-66"><a href="#cb51-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-67"><a href="#cb51-67" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_list</span>(p1, p2, <span class="at">tag_levels=</span><span class="st">'A'</span>, <span class="at">ncol=</span><span class="dv">2</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="02_tidytree_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="1344"></p>
<figcaption><strong>Reroot a phylogenetic tree with associated data.</strong> Original tree (A) and re-rooted tree (B) with associated data mapped to the branches or nodes of the tree correctly. (A) and (B) present before and after rooting on the branch leading to the tip node ‘Suricata_suricatta’, respectively.</figcaption>
</figure>
</div>
</div>
</div>
<p>The <code>outgroup</code> parameter represents the specific new <code>outgroup</code>, it can be a node label (character) or node number. If it is a “single one” value, meaning using the node below this tip as the new root, if it has multiple values, meaning the most recent common of the values will be used as the new root. Note that, if the node labels should be treated as edge labels, the <code>edgelabel</code> should be set to <code>TRUE</code> to return the correct relationship between the <code>node</code> and <code>associated data</code>. For more details about re-root, including precautions and pitfalls, please refer to the review article <span class="citation" data-cites="reroot_review">(<a href="#ref-reroot_review" role="doc-biblioref">Czech et al., 2017</a>)</span>.</p>
</section>
<section id="sec-rescale-treeio" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="sec-rescale-treeio"><span class="header-section-number">2.4</span> Rescaling Tree Branches</h2>
<p>Phylogenetic data can be merged for joint analysis (Figure <span class="quarto-unresolved-ref">?fig-correlations</span>). They can be displayed on the same tree structure as a more complex annotation to help visually inspection of their evolutionary patterns. All the numerical data stored in a <code>treedata</code> object can be used to re-scale tree branches. For example, CodeML infers d<sub>N</sub>/d<sub>S</sub>, d<sub>N</sub>, and d<sub>S</sub>, all these statistics can be used as branch lengths (Figure <span class="quarto-unresolved-ref">?fig-rescale</span>). All these values can also be used to color the tree (session <a href="04_ggtree_visualization.html#color-tree">4.3.4</a>) and can be projected to a vertical dimension to create a two-dimensional tree or phenogram (session <a href="#layouts-of-phylogenetic-tree">4.2.2</a> and Figures <span class="quarto-unresolved-ref">?fig-2d</span> and <span class="quarto-unresolved-ref">?fig-continuousColor</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(merged_tree) <span class="sc">+</span> <span class="fu">theme_tree2</span>()</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(<span class="fu">rescale_tree</span>(merged_tree, <span class="st">'dN'</span>)) <span class="sc">+</span> <span class="fu">theme_tree2</span>()</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(<span class="fu">rescale_tree</span>(merged_tree, <span class="st">'rate'</span>)) <span class="sc">+</span> <span class="fu">theme_tree2</span>()</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_list</span>(p1, p2, p3, <span class="at">ncol=</span><span class="dv">3</span>, <span class="at">tag_levels=</span><span class="st">'A'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="02_tidytree_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Rescaling tree branches.</strong> The tree with branches scaled in time (year from the root) (A). The tree was rescaled using <em>d<sub>N</sub></em> as branch lengths (B). The tree was rescaled using substitution rates (C).</figcaption>
</figure>
</div>
</div>
</div>
<p>Modifying branch lengths in the tree object in addtion to using the <code>rescale_tree()</code> function, users can directly specify a variable as branch length in <code>ggtree()</code> as demonstrated in <a href="04_ggtree_visualization.html#rescale-tree">session 4.3.6</a>.</p>
</section>
<section id="subsetting-tree-with-data" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="subsetting-tree-with-data"><span class="header-section-number">2.5</span> Subsetting Tree with Data</h2>
<section id="sec-remove-tip" class="level3" data-number="2.5.1">
<h3 data-number="2.5.1" class="anchored" data-anchor-id="sec-remove-tip"><span class="header-section-number">2.5.1</span> Removing tips in a phylogenetic tree</h3>
<p>Sometimes we want to remove selected tips from a phylogenetic tree. This is due to several reasons, including low sequence quality, errors in sequence assembly, an alignment error in part of the sequence, an error in phylogenetic inference, <em>etc</em>.</p>
<p>Let’s say that we want to remove three tips (colored red) from the tree (Figure <span class="quarto-unresolved-ref">?fig-removeTipA</span>), the <code>drop.tip()</code> method removes specified tips and updates the tree (Figure <span class="quarto-unresolved-ref">?fig-removeTipB</span>). All associated data will be maintained in the updated tree.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/NHX"</span>, <span class="st">"phyldog.nhx"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>nhx <span class="ot">&lt;-</span> <span class="fu">read.nhx</span>(f)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>to_drop <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Physonect_sp_@2066767"</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Lychnagalma_utricularia@2253871"</span>,</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Kephyes_ovata@2606431"</span>)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(nhx) <span class="sc">+</span> <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> label <span class="sc">%in%</span> to_drop)) <span class="sc">+</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>)) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="fl">0.8</span>)</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>nhx_reduced <span class="ot">&lt;-</span> <span class="fu">drop.tip</span>(nhx, to_drop)</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(nhx_reduced) <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="fl">0.8</span>)  </span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_list</span>(p1, p2, <span class="at">ncol=</span><span class="dv">2</span>, <span class="at">tag_levels =</span> <span class="st">"A"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="02_tidytree_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Removing tips from a tree.</strong> Original tree with three tips (colored red) to remove (A). The updated tree removed selected tips (B).</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-subset-tip" class="level3" data-number="2.5.2">
<h3 data-number="2.5.2" class="anchored" data-anchor-id="sec-subset-tip"><span class="header-section-number">2.5.2</span> Subsetting tree by tip label</h3>
<p>Sometimes a tree can be large and difficult to look at only the portions of interest. The <code>tree_subset()</code> function was created in the <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> to extract a subset of the tree portion while still maintaining the structure of the tree portion. The <code>beast_tree</code> in Figure <span class="quarto-unresolved-ref">?fig-subsetTipA</span> is slightly crowded. Obviously, we can make the figure taller to allow more space for the labels (similar to using the “Expansion” slider in <code>FigTree</code>) or we can make the text smaller. However, these solutions are not always applicable when you have a lot of tips (<em>e.g.</em>, hundreds or thousands of tips). In particular, when you are only interested in the portion of the tree around a particular tip, you certainly don’t want to explore a large tree to find the certain species you are interested in.</p>
<p>Let’s say you are interested in tip <em>A/Swine/HK/168/2012</em> from the tree (Figure <span class="quarto-unresolved-ref">?fig-subsetTipA</span>) and you want to look at the immediate relatives of this tip.</p>
<p>The <code>tree_subset()</code> function allows you to look at the portions of the tree that are of interest. By default, the <code>tree_subset()</code> function will internally call the <a href="#groupotu"><code>groupOTU()</code></a> to assign the group specified tip from the rest of the other tips (Figure <span class="quarto-unresolved-ref">?fig-subsetTipB</span>). Additionally, the branch lengths and related associated data are maintained after subsetting (Figure <span class="quarto-unresolved-ref">?fig-subsetTipC</span>). The root of the tree is always anchored at zero for the subset tree by default and all the distances are relative to this root. If you want all the distances to be relative to the original root, you can specify the root position (by <code>root.position</code> parameter) to the root edge of the subset tree, which is the sum of branch lengths from the original root to the root of the subset tree (Figures <span class="quarto-unresolved-ref">?fig-subsetTipD</span> and E).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>beast_file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"examples/MCC_FluA_H3.tree"</span>, <span class="at">package=</span><span class="st">"ggtree"</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>beast_tree <span class="ot">&lt;-</span> <span class="fu">read.beast</span>(beast_file)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">=</span> <span class="fu">ggtree</span>(beast_tree) <span class="sc">+</span> </span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="at">offset=</span>.<span class="dv">05</span>) <span class="sc">+</span>  <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">40</span>) <span class="sc">+</span> <span class="fu">theme_tree2</span>()</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>tree2 <span class="ot">=</span> <span class="fu">tree_subset</span>(beast_tree, <span class="st">"A/Swine/HK/168/2012"</span>, <span class="at">levels_back=</span><span class="dv">4</span>)  </span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree2, <span class="fu">aes</span>(<span class="at">color=</span>group)) <span class="sc">+</span></span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>), <span class="at">guide =</span> <span class="st">'none'</span>) <span class="sc">+</span></span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="at">offset=</span>.<span class="dv">2</span>) <span class="sc">+</span>  <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="fl">4.5</span>) <span class="sc">+</span> <span class="fu">theme_tree2</span>() </span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> p2 <span class="sc">+</span>   </span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">fill =</span> rate), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_continuous</span>(<span class="at">low =</span> <span class="st">'blue'</span>, <span class="at">high =</span> <span class="st">'red'</span>) <span class="sc">+</span></span>
<span id="cb54-15"><a href="#cb54-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">5</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">'right'</span>)</span>
<span id="cb54-16"><a href="#cb54-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-17"><a href="#cb54-17" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree2, <span class="fu">aes</span>(<span class="at">color=</span>group), </span>
<span id="cb54-18"><a href="#cb54-18" aria-hidden="true" tabindex="-1"></a>          <span class="at">root.position =</span> <span class="fu">as.phylo</span>(tree2)<span class="sc">$</span>root.edge) <span class="sc">+</span></span>
<span id="cb54-19"><a href="#cb54-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>() <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">18</span>, <span class="dv">24</span>) <span class="sc">+</span> </span>
<span id="cb54-20"><a href="#cb54-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>), <span class="at">guide =</span> <span class="st">'none'</span>) <span class="sc">+</span></span>
<span id="cb54-21"><a href="#cb54-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tree2</span>()</span>
<span id="cb54-22"><a href="#cb54-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-23"><a href="#cb54-23" aria-hidden="true" tabindex="-1"></a>p5 <span class="ot">&lt;-</span> p4 <span class="sc">+</span> </span>
<span id="cb54-24"><a href="#cb54-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rootedge</span>() <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">50</span>) </span>
<span id="cb54-25"><a href="#cb54-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-26"><a href="#cb54-26" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_list</span>(p1, p2, p3, p4, p5, </span>
<span id="cb54-27"><a href="#cb54-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">design=</span><span class="st">"AABBCC</span><span class="sc">\n</span><span class="st">AADDEE"</span>, <span class="at">tag_levels=</span><span class="st">'A'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="02_tidytree_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Subsetting tree for a specific tip.</strong> Original tree (A). Subset tree (B). Subset tree with data (C). Visualize the subset tree relative to the original position, without rootedge (D) and with rootedge (E).</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-subset-node" class="level3" data-number="2.5.3">
<h3 data-number="2.5.3" class="anchored" data-anchor-id="sec-subset-node"><span class="header-section-number">2.5.3</span> Subsetting tree by internal node number</h3>
<p>If you are interested in a certain clade, you can specify the input node as an internal node number. The <code>tree_subset()</code> function will take the clade as a whole and also trace it back to particular levels to look at the immediate relatives of the clade (Figures <span class="quarto-unresolved-ref">?fig-subsetNodeA</span> and B). We can use the <code>tree_subset()</code> function to zoom in selected portions and plot a whole tree with the portion of it, which is similar to the <code>ape::zoom()</code> function to explore a very large tree (Figures <span class="quarto-unresolved-ref">?fig-subsetNodeC</span> and D). Users can also use <code>viewClade()</code> function to restrict tree visualization at specific clade as demonstrated in <a href="06_ggtree_manipulation.html#viewing-selected-clade">session 6.1</a>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>clade <span class="ot">&lt;-</span> <span class="fu">tree_subset</span>(beast_tree, <span class="at">node=</span><span class="dv">121</span>, <span class="at">levels_back=</span><span class="dv">0</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>clade2 <span class="ot">&lt;-</span> <span class="fu">tree_subset</span>(beast_tree, <span class="at">node=</span><span class="dv">121</span>, <span class="at">levels_back=</span><span class="dv">2</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(clade) <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">5</span>)</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(clade2, <span class="fu">aes</span>(<span class="at">color=</span>group)) <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> </span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">9</span>) <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>))</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidytree)</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(treeio)</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(chiroptera)</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"Plecotus"</span>, chiroptera<span class="sc">$</span>tip.label)</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a>chiroptera <span class="ot">&lt;-</span> <span class="fu">groupOTU</span>(chiroptera, nodes)</span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a>clade <span class="ot">&lt;-</span> <span class="fu">MRCA</span>(chiroptera, nodes)</span>
<span id="cb55-18"><a href="#cb55-18" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">tree_subset</span>(chiroptera, clade, <span class="at">levels_back =</span> <span class="dv">0</span>)</span>
<span id="cb55-19"><a href="#cb55-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-20"><a href="#cb55-20" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(chiroptera, <span class="fu">aes</span>(<span class="at">colour =</span> group)) <span class="sc">+</span> </span>
<span id="cb55-21"><a href="#cb55-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>)) <span class="sc">+</span></span>
<span id="cb55-22"><a href="#cb55-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb55-23"><a href="#cb55-23" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(x) <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">6</span>)</span>
<span id="cb55-24"><a href="#cb55-24" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_list</span>(p1, p2, p3, p4, </span>
<span id="cb55-25"><a href="#cb55-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol=</span><span class="dv">2</span>, <span class="at">tag_levels =</span> <span class="st">'A'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="02_tidytree_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Subsetting tree for the specific clade.</strong> Extracting a clade (A). Extracting a clade and tracing it back to look at its immediate relatives (B). Viewing a very large tree (C) and a selected portion of it (D).</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="sec-ggtree-fortify" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="sec-ggtree-fortify"><span class="header-section-number">2.6</span> Manipulating Tree Data for Visualization</h2>
<p>Tree visualization is supported by <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span>. Although <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> implemented several methods for <a href="#chapter6">visual exploration of trees with data</a>, you may want to do something that is not supported directly. In this case, you need to manipulate tree data with node coordination positions that are used for visualization. This is quite easy with <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a>. Users can use the <code>fortify()</code> method which internally calls <code>tidytree::as_tibble()</code> to convert the tree to a tidy data frame and add columns of coordination positions (<em>i.e.</em>, x, y, branch, and angle) that are used to plot the tree. You can also access the data via <code>ggtree(tree)$data</code>.</p>
<p>Here is an example to plot two trees face-to-face that is similar to a graph produced by the <code>ape::cophyloplot()</code> function (Figure <span class="quarto-unresolved-ref">?fig-cophylo</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1024</span>)</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="dv">30</span>)</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="dv">30</span>)</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(x, <span class="at">layout=</span><span class="st">'roundrect'</span>) <span class="sc">+</span> </span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hilight</span>(</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">subset =</span> node <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">38</span>, <span class="dv">48</span>, <span class="dv">58</span>, <span class="dv">36</span>),</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>                     <span class="at">node =</span> node,</span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>                     <span class="at">fill =</span> <span class="fu">as.factor</span>(node)</span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a>                     )</span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a>     ) <span class="sc">+</span></span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"clades for tree in left"</span> )</span>
<span id="cb56-15"><a href="#cb56-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-16"><a href="#cb56-16" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(y)</span>
<span id="cb56-17"><a href="#cb56-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-18"><a href="#cb56-18" aria-hidden="true" tabindex="-1"></a>d1 <span class="ot">&lt;-</span> p1<span class="sc">$</span>data</span>
<span id="cb56-19"><a href="#cb56-19" aria-hidden="true" tabindex="-1"></a>d2 <span class="ot">&lt;-</span> p2<span class="sc">$</span>data</span>
<span id="cb56-20"><a href="#cb56-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-21"><a href="#cb56-21" aria-hidden="true" tabindex="-1"></a><span class="do">## reverse x-axis and </span></span>
<span id="cb56-22"><a href="#cb56-22" aria-hidden="true" tabindex="-1"></a><span class="do">## set offset to make the tree on the right-hand side of the first tree</span></span>
<span id="cb56-23"><a href="#cb56-23" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span>x <span class="ot">&lt;-</span> <span class="fu">max</span>(d2<span class="sc">$</span>x) <span class="sc">-</span> d2<span class="sc">$</span>x <span class="sc">+</span> <span class="fu">max</span>(d1<span class="sc">$</span>x) <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb56-24"><a href="#cb56-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-25"><a href="#cb56-25" aria-hidden="true" tabindex="-1"></a>pp <span class="ot">&lt;-</span> p1 <span class="sc">+</span> <span class="fu">geom_tree</span>(<span class="at">data=</span>d2, <span class="at">layout=</span><span class="st">'ellipse'</span>) <span class="sc">+</span>      </span>
<span id="cb56-26"><a href="#cb56-26" aria-hidden="true" tabindex="-1"></a>  ggnewscale<span class="sc">::</span><span class="fu">new_scale_fill</span>() <span class="sc">+</span></span>
<span id="cb56-27"><a href="#cb56-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hilight</span>(</span>
<span id="cb56-28"><a href="#cb56-28" aria-hidden="true" tabindex="-1"></a>         <span class="at">data =</span> d2, </span>
<span id="cb56-29"><a href="#cb56-29" aria-hidden="true" tabindex="-1"></a>         <span class="at">mapping =</span> <span class="fu">aes</span>( </span>
<span id="cb56-30"><a href="#cb56-30" aria-hidden="true" tabindex="-1"></a>            <span class="at">subset =</span> node <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">38</span>, <span class="dv">48</span>, <span class="dv">58</span>),</span>
<span id="cb56-31"><a href="#cb56-31" aria-hidden="true" tabindex="-1"></a>            <span class="at">node=</span>node,</span>
<span id="cb56-32"><a href="#cb56-32" aria-hidden="true" tabindex="-1"></a>            <span class="at">fill=</span><span class="fu">as.factor</span>(node))</span>
<span id="cb56-33"><a href="#cb56-33" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb56-34"><a href="#cb56-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"clades for tree in right"</span> ) </span>
<span id="cb56-35"><a href="#cb56-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-36"><a href="#cb56-36" aria-hidden="true" tabindex="-1"></a>dd <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(d1, d2) <span class="sc">%&gt;%</span> </span>
<span id="cb56-37"><a href="#cb56-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(label))</span>
<span id="cb56-38"><a href="#cb56-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-39"><a href="#cb56-39" aria-hidden="true" tabindex="-1"></a>pp <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="fu">aes</span>(x, y, <span class="at">group=</span>label), <span class="at">data=</span>dd, <span class="at">color=</span><span class="st">'grey'</span>) <span class="sc">+</span></span>
<span id="cb56-40"><a href="#cb56-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>(<span class="at">geom =</span> <span class="st">'shadowtext'</span>, <span class="at">bg.colour =</span> <span class="fu">alpha</span>(<span class="st">'firebrick'</span>, .<span class="dv">5</span>)) <span class="sc">+</span></span>
<span id="cb56-41"><a href="#cb56-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>(<span class="at">data =</span> d2, <span class="at">hjust=</span><span class="dv">1</span>, <span class="at">geom =</span> <span class="st">'shadowtext'</span>, </span>
<span id="cb56-42"><a href="#cb56-42" aria-hidden="true" tabindex="-1"></a>                <span class="at">bg.colour =</span> <span class="fu">alpha</span>(<span class="st">'firebrick'</span>, .<span class="dv">5</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="02_tidytree_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Plot two phylogenetic trees face-to-face.</strong> Plotting a tree using <code>ggtree()</code> (left-hand side) and subsequently adding another layer of a tree by <code>geom_tree()</code> (right-hand side). The relative positions of the plotted trees can be manually adjusted and adding layers to each of the trees (<em>e.g.</em>, tip labels and highlighting clades) is independent.</figcaption>
</figure>
</div>
</div>
</div>
<p>It is quite easy to plot multiple trees and connect taxa in one figure; for instance, plotting trees constructed from all internal gene segments of influenza virus and connecting equivalent strains across the trees <span class="citation" data-cites="venkatesh_avian_2018">(<a href="#ref-venkatesh_avian_2018" role="doc-biblioref">Venkatesh et al., 2018</a>)</span>. Figure <span class="quarto-unresolved-ref">?fig-cophylo3</span> demonstrates the usage of plotting multiple trees by combining multiple layers of <code>geom_tree()</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="dv">30</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>d2 <span class="ot">&lt;-</span> <span class="fu">fortify</span>(y)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>d3 <span class="ot">&lt;-</span> <span class="fu">fortify</span>(z)</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span>x <span class="ot">&lt;-</span> d2<span class="sc">$</span>x <span class="sc">+</span> <span class="fu">max</span>(d1<span class="sc">$</span>x) <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>d3<span class="sc">$</span>x <span class="ot">&lt;-</span> d3<span class="sc">$</span>x <span class="sc">+</span> <span class="fu">max</span>(d2<span class="sc">$</span>x) <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>dd <span class="ot">=</span> <span class="fu">bind_rows</span>(d1, d2, d3) <span class="sc">%&gt;%</span> </span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(label))</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> <span class="fu">geom_tree</span>(<span class="at">data =</span> d2) <span class="sc">+</span> <span class="fu">geom_tree</span>(<span class="at">data =</span> d3) <span class="sc">+</span> <span class="fu">geom_tiplab</span>(<span class="at">data=</span>d3) <span class="sc">+</span> </span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(x, y, <span class="at">group=</span>label, <span class="at">color=</span>node <span class="sc">&lt;</span> <span class="dv">15</span>), <span class="at">data=</span>dd, <span class="at">alpha=</span>.<span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="02_tidytree_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Plot multiple phylogenetic trees side-by-side.</strong> Plotting a tree using <code>ggtree()</code> and subsequently adding multiple layers of trees by <code>geom_tree()</code>.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-summary2" class="level2" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="sec-summary2"><span class="header-section-number">2.7</span> Summary</h2>
<p>The <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package allows us to import diverse phylogeny associated data into R. However, a phylogenetic tree is stored in a way to facilitate computational processing which is not human friendly and needs the expertise to manipulate and explore tree data. The <a href="https://CRAN.R-project.org/package=tidytree"><strong>tidytree</strong></a> package provides a tidy interface for exploring tree data, while <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> provides a set of utilities to visualize and explore tree data using the grammar of graphics. This full suite of packages makes it easy for ordinary users to interact with tree data and allows us to integrate phylogeny associated data from different sources (<em>e.g.</em>, experimental results or analysis findings), which creates the possibilities of integrative and comparative study. Moreover, this package suite brings phylogenetic analysis into the tidyverse and certainly takes us to the next level of processing phylogenetic data.</p>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-reroot_review" class="csl-entry" role="listitem">
Czech, L., Huerta-Cepas, J., &amp; Stamatakis, A. (2017). <span class="nocase">A Critical Review on the Use of Support Values in Tree Viewers and Bioinformatics Toolkits</span>. <em>Molecular Biology and Evolution</em>, <em>34</em>(6), 1535–1542. <a href="https://doi.org/10.1093/molbev/msx055">https://doi.org/10.1093/molbev/msx055</a>
</div>
<div id="ref-liang_expansion_2014" class="csl-entry" role="listitem">
Liang, H., Lam, T. T.-Y., Fan, X., Chen, X., Zeng, Y., Zhou, J., Duan, L., Tse, M., Chan, C.-H., Li, L., Leung, T.-Y., Yip, C.-H., Cheung, C.-L., Zhou, B., Smith, D. K., Poon, L. L.-M., Peiris, M., Guan, Y., &amp; Zhu, H. (2014). Expansion of genotypic diversity and establishment of 2009 <span>H</span>1N1 pandemic-origin internal genes in pigs in <span>China</span>. <em>Journal of Virology</em>, JVI.01327–14. <a href="https://doi.org/10.1128/JVI.01327-14">https://doi.org/10.1128/JVI.01327-14</a>
</div>
<div id="ref-paradis_ape_2004" class="csl-entry" role="listitem">
Paradis, E., Claude, J., &amp; Strimmer, K. (2004). <span>APE</span>: <span>Analyses</span> of <span>Phylogenetics</span> and <span>Evolution</span> in <span>R</span> language. <em>Bioinformatics</em>, <em>20</em>(2), 289–290. <a href="https://doi.org/10.1093/bioinformatics/btg412">https://doi.org/10.1093/bioinformatics/btg412</a>
</div>
<div id="ref-venkatesh_avian_2018" class="csl-entry" role="listitem">
Venkatesh, D., Poen, M. J., Bestebroer, T. M., Scheuer, R. D., Vuong, O., Chkhaidze, M., Machablishvili, A., Mamuchadze, J., Ninua, L., Fedorova, N. B., Halpin, R. A., Lin, X., Ransier, A., Stockwell, T. B., Wentworth, D. E., Kriti, D., Dutta, J., Bakel, H. van, Puranik, A., … Lewis, N. S. (2018). Avian influenza viruses in wild birds: Virus evolution in a multihost ecosystem. <em>Journal of Virology</em>, <em>92</em>(15). <a href="https://doi.org/10.1128/JVI.00433-18">https://doi.org/10.1128/JVI.00433-18</a>
</div>
<div id="ref-wang_treeio_2020" class="csl-entry" role="listitem">
Wang, L.-G., Lam, T. T.-Y., Xu, S., Dai, Z., Zhou, L., Feng, T., Guo, P., Dunn, C. W., Jones, B. R., Bradley, T., Zhu, H., Guan, Y., Jiang, Y., &amp; Yu, G. (2020). Treeio: An r package for phylogenetic tree input and output with richly annotated and associated data. <em>Molecular Biology and Evolution</em>, <em>37</em>(2), 599–603. <a href="https://doi.org/10.1093/molbev/msz240">https://doi.org/10.1093/molbev/msz240</a>
</div>
<div id="ref-ggtreeExtra_2021" class="csl-entry" role="listitem">
Xu, S., Dai, Z., Guo, P., Fu, X., Liu, S., Zhou, L., Tang, W., Feng, T., Chen, M., Zhan, L., Wu, T., Hu, E., Jiang, Y., Bo, X., &amp; Yu, G. (2021). <span class="nocase">ggtreeExtra: Compact Visualization of Richly Annotated Phylogenetic Data</span>. <em>Molecular Biology and Evolution</em>, <em>38</em>(9), 4039–4042. <a href="https://doi.org/10.1093/molbev/msab166">https://doi.org/10.1093/molbev/msab166</a>
</div>
<div id="ref-yu_two_2018" class="csl-entry" role="listitem">
Yu, G., Lam, T. T.-Y., Zhu, H., &amp; Guan, Y. (2018). Two methods for mapping and visualizing associated data on phylogeny using ggtree. <em>Molecular Biology and Evolution</em>, <em>35</em>(12), 3041–3043. <a href="https://doi.org/10.1093/molbev/msy194">https://doi.org/10.1093/molbev/msy194</a>
</div>
<div id="ref-yu_ggtree:_2017" class="csl-entry" role="listitem">
Yu, G., Smith, D. K., Zhu, H., Guan, Y., &amp; Lam, T. T.-Y. (2017). Ggtree: An r package for visualization and annotation of phylogenetic trees with their covariates and other associated data. <em>Methods in Ecology and Evolution</em>, <em>8</em>(1), 28–36. <a href="https://doi.org/10.1111/2041-210X.12628">https://doi.org/10.1111/2041-210X.12628</a>
</div>
</div>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p><a href="https://groups.google.com/forum/#!msg/bioc-ggtree/Q4LnwoTf1DM/uqYdYB_VBAAJ" class="uri">https://groups.google.com/forum/#!msg/bioc-ggtree/Q4LnwoTf1DM/uqYdYB_VBAAJ</a><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./01_treeio_importing_tree.html" class="pagination-link" aria-label="(PART\*) Part I: Tree data input, output, and manipulation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">(PART*) Part I: Tree data input, output, and manipulation</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./03_treeio_exporting_tree.html" class="pagination-link" aria-label="Exporting tree with data">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Exporting tree with data</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/YuLab-SMU/treedata-book/edit/main/02_tidytree.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/blob/main/02_tidytree.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div></div></footer></body></html>