<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Master ggtree package suite to handle tree with data.">

<title>4&nbsp; Phylogenetic Tree Annotation – Data Integration, Manipulation and Visualization of Phylogenetic Trees</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./06_ggtree_manipulation.html" rel="next">
<link href="./04_ggtree_visualization.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-0dd2bd5de344125cf763a379ddc3eb04.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">


<link rel="stylesheet" href="css/style.css">
</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./05_ggtree_annotation.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Phylogenetic Tree Annotation</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Data Integration, Manipulation and Visualization of Phylogenetic Trees</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/YuLab-SMU/treedata-book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Data Integration, Manipulation and Visualization of Phylogenetic Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./00_author.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About the Author</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_treeio_importing_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part I: Tree data input, output, and manipulation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_tidytree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Manipulating Tree with Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_treeio_exporting_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Exporting tree with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_ggtree_visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part II: Tree data visualization and annotation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_ggtree_annotation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Phylogenetic Tree Annotation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_ggtree_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Visual Exploration of Phylogenetic Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_ggtree_tree_with_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Plotting tree with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_ggtree_with_silhouette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Annotating Tree with Silhouette Images and Sub-plots</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_ggtree_tree_objects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">ggtree for Other Tree-like Objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_ggtree_exts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part III: ggtree extensions</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11_ggtree_exts-others.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Other ggtree Extensions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12_ggtree_utilities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part IV: Miscellaneous topics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13_ggtree_gallery.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Gallery of Reproducible Examples</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendix</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-faq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(APPENDIX) Appendix</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Related Tools</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-fig-tab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Figures and Tables</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#visualizing-and-annotating-tree-using-grammar-of-graphics" id="toc-visualizing-and-annotating-tree-using-grammar-of-graphics" class="nav-link active" data-scroll-target="#visualizing-and-annotating-tree-using-grammar-of-graphics"><span class="header-section-number">4.1</span> Visualizing and Annotating Tree Using Grammar of Graphics</a></li>
  <li><a href="#layers-for-tree-annotation" id="toc-layers-for-tree-annotation" class="nav-link" data-scroll-target="#layers-for-tree-annotation"><span class="header-section-number">4.2</span> Layers for Tree Annotation</a>
  <ul class="collapse">
  <li><a href="#colored-strips" id="toc-colored-strips" class="nav-link" data-scroll-target="#colored-strips"><span class="header-section-number">4.2.1</span> Colored strips</a></li>
  <li><a href="#highlight-clades" id="toc-highlight-clades" class="nav-link" data-scroll-target="#highlight-clades"><span class="header-section-number">4.2.2</span> Highlight clades</a></li>
  <li><a href="#taxa-connection" id="toc-taxa-connection" class="nav-link" data-scroll-target="#taxa-connection"><span class="header-section-number">4.2.3</span> Taxa connection</a></li>
  <li><a href="#uncertainty-of-evolutionary-inference" id="toc-uncertainty-of-evolutionary-inference" class="nav-link" data-scroll-target="#uncertainty-of-evolutionary-inference"><span class="header-section-number">4.2.4</span> Uncertainty of evolutionary inference</a></li>
  </ul></li>
  <li><a href="#tree-annotation-with-output-from-evolution-software" id="toc-tree-annotation-with-output-from-evolution-software" class="nav-link" data-scroll-target="#tree-annotation-with-output-from-evolution-software"><span class="header-section-number">4.3</span> Tree Annotation with Output from Evolution Software</a>
  <ul class="collapse">
  <li><a href="#tree-annotation-using-data-from-evolutionary-analysis-software" id="toc-tree-annotation-using-data-from-evolutionary-analysis-software" class="nav-link" data-scroll-target="#tree-annotation-using-data-from-evolutionary-analysis-software"><span class="header-section-number">4.3.1</span> Tree annotation using data from evolutionary analysis software</a></li>
  </ul></li>
  <li><a href="#sec-summary5" id="toc-sec-summary5" class="nav-link" data-scroll-target="#sec-summary5"><span class="header-section-number">4.4</span> Summary</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/YuLab-SMU/treedata-book/edit/main/05_ggtree_annotation.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/blob/main/05_ggtree_annotation.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-chapter5" class="quarto-section-identifier"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Phylogenetic Tree Annotation</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div style="page-break-after: always;"></div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"kableExtra"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ape"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"cowplot"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"treeio"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>treeio v1.35.0 Learn more at https://yulab-smu.top/contribution-tree-data/

Please cite:

LG Wang, TTY Lam, S Xu, Z Dai, L Zhou, T Feng, P Guo, CW Dunn, BR
Jones, T Bradley, H Zhu, Y Guan, Y Jiang, G Yu. treeio: an R package
for phylogenetic tree input and output with richly annotated and
associated data. Molecular Biology and Evolution. 2020, 37(2):599-603.
doi: 10.1093/molbev/msz240</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggtree"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>ggtree v4.1.1.003 Learn more at https://yulab-smu.top/contribution-tree-data/

Please cite:

Shuangbin Xu, Lin Li, Xiao Luo, Meijun Chen, Wenli Tang, Li Zhan, Zehan
Dai, Tommy T. Lam, Yi Guan, Guangchuang Yu. Ggtree: A serialized data
object for visualization of a phylogenetic tree and annotation data.
iMeta 2022, 1(4):e56. doi:10.1002/imt2.56</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ggtree'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:ape':

    rotate</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"yulab.utils"</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"software-link.R"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="visualizing-and-annotating-tree-using-grammar-of-graphics" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="visualizing-and-annotating-tree-using-grammar-of-graphics"><span class="header-section-number">4.1</span> Visualizing and Annotating Tree Using Grammar of Graphics</h2>
<p>The <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span> is designed for a more general-purpose or a specific type of tree visualization and annotation. It supports the grammar of graphics implemented in <a href="https://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a> and users can freely visualize/annotate a tree by combining several annotation layers.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>treetext <span class="ot">&lt;-</span> <span class="st">"(((ADH2:0.1[&amp;&amp;NHX:S=human], ADH1:0.11[&amp;&amp;NHX:S=human]):</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="st">  0.05 [&amp;&amp;NHX:S=primates:D=Y:B=100],ADHY:</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="st">  0.1[&amp;&amp;NHX:S=nematode],ADHX:0.12 [&amp;&amp;NHX:S=insect]):</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="st">  0.1[&amp;&amp;NHX:S=metazoa:D=N],(ADH4:0.09[&amp;&amp;NHX:S=yeast],</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="st">  ADH3:0.13[&amp;&amp;NHX:S=yeast], ADH2:0.12[&amp;&amp;NHX:S=yeast],</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="st">  ADH1:0.11[&amp;&amp;NHX:S=yeast]):0.1[&amp;&amp;NHX:S=Fungi])[&amp;&amp;NHX:D=N];"</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">read.nhx</span>(<span class="fu">textConnection</span>(treetext))</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">ggtree</span>(tree) <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="fu">aes</span>(<span class="at">x=</span>branch, <span class="at">label=</span>S), <span class="at">fill=</span><span class="st">'lightgreen'</span>) <span class="sc">+</span> </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="fu">aes</span>(<span class="at">label=</span>D), <span class="at">fill=</span><span class="st">'steelblue'</span>) <span class="sc">+</span> </span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>B), <span class="at">hjust=</span><span class="sc">-</span>.<span class="dv">5</span>)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">xlim</span>(<span class="cn">NA</span>, <span class="fl">0.28</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>treetext <span class="ot">=</span> <span class="st">"(((ADH2:0.1[&amp;&amp;NHX:S=human], ADH1:0.11[&amp;&amp;NHX:S=human]):</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="st">0.05 [&amp;&amp;NHX:S=primates:D=Y:B=100],ADHY:</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="st">0.1[&amp;&amp;NHX:S=nematode],ADHX:0.12 [&amp;&amp;NHX:S=insect]):</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="st">0.1[&amp;&amp;NHX:S=metazoa:D=N],(ADH4:0.09[&amp;&amp;NHX:S=yeast],</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="st">ADH3:0.13[&amp;&amp;NHX:S=yeast], ADH2:0.12[&amp;&amp;NHX:S=yeast],</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="st">ADH1:0.11[&amp;&amp;NHX:S=yeast]):0.1[&amp;&amp;NHX:S=Fungi])[&amp;&amp;NHX:D=N];"</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">read.nhx</span>(<span class="fu">textConnection</span>(treetext))</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree) <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> </span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="fu">aes</span>(<span class="at">x=</span>branch, <span class="at">label=</span>S), <span class="at">fill=</span><span class="st">'lightgreen'</span>) <span class="sc">+</span> </span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="fu">aes</span>(<span class="at">label=</span>D), <span class="at">fill=</span><span class="st">'steelblue'</span>) <span class="sc">+</span> </span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>B), <span class="at">hjust=</span><span class="sc">-</span>.<span class="dv">5</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption><strong>Annotating tree using the grammar of graphics.</strong> The NHX tree was annotated using the grammar of graphic syntax by combining different layers using the <code>+</code> operator. Species information was labeled in the middle of the branches. Duplication events were shown on the most recent common ancestor and clade bootstrap values were displayed near to it.</figcaption>
</figure>
</div>
</div>
</div>
<p>Here, as an example, we visualized the tree with several layers to display annotation stored in NHX tags, including a layer of <code>geom_tiplab()</code> to display tip labels (gene name in this case), a layer using <code>geom_label()</code> to show species information (<code>S</code> tag) colored by light green, a layer of duplication event information (<code>D</code> tag) colored by steelblue and another layer using <code>geom_text()</code> to show bootstrap value (<code>B</code> tag).</p>
<p>Layers defined in <a href="https://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a> can be applied to <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> directly as demonstrated in Figure <span class="quarto-unresolved-ref">?fig-ggtreeNHX</span> of using <code>geom_label()</code> and <code>geom_text()</code>. But <a href="https://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a> does not provide graphic layers that are specifically designed for phylogenetic tree annotation. For instance, layers for tip labels, tree branch scale legend, highlight, or labeling clade are all unavailable. To make tree annotation more flexible, several layers have been implemented in <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> (Table <span class="quarto-unresolved-ref">?tbl-geoms</span>), enabling different ways of annotation on various parts/components of a phylogenetic tree.</p>
<div class="cell">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small">
<caption>Geom layers defined in ggtree.</caption>
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">Layer</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">geom_balance</td>
<td style="text-align: left;">Highlights the two direct descendant clades of an internal node</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_cladelab</td>
<td style="text-align: left;">Annotates a clade with bar and text label (or image)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_facet</td>
<td style="text-align: left;">Plots associated data in a specific panel (facet) and aligns the plot with the tree</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_hilight</td>
<td style="text-align: left;">Highlights selected clade with rectangular or round shape</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_inset</td>
<td style="text-align: left;">Adds insets (subplots) to tree nodes</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_label2</td>
<td style="text-align: left;">The modified version of geom_label, with subset aesthetic supported</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_nodepoint</td>
<td style="text-align: left;">Annotates internal nodes with symbolic points</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_point2</td>
<td style="text-align: left;">The modified version of geom_point, with subset aesthetic supported</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_range</td>
<td style="text-align: left;">Bar layer to present uncertainty of evolutionary inference</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_rootpoint</td>
<td style="text-align: left;">Annotates root node with symbolic point</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_rootedge</td>
<td style="text-align: left;">Adds root-edge to a tree</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_segment2</td>
<td style="text-align: left;">The modified version of geom_segment, with subset aesthetic supported</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_strip</td>
<td style="text-align: left;">Annotates associated taxa with bar and (optional) text label</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_taxalink</td>
<td style="text-align: left;">Links related taxa</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_text2</td>
<td style="text-align: left;">The modified version of geom_text, with subset aesthetic supported</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_tiplab</td>
<td style="text-align: left;">The layer of tip labels</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_tippoint</td>
<td style="text-align: left;">Annotates external nodes with symbolic points</td>
</tr>
<tr class="even">
<td style="text-align: left;">geom_tree</td>
<td style="text-align: left;">Tree structure layer, with multiple layouts supported</td>
</tr>
<tr class="odd">
<td style="text-align: left;">geom_treescale</td>
<td style="text-align: left;">Tree branch scale legend</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="layers-for-tree-annotation" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="layers-for-tree-annotation"><span class="header-section-number">4.2</span> Layers for Tree Annotation</h2>
<section id="colored-strips" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="colored-strips"><span class="header-section-number">4.2.1</span> Colored strips</h3>
<p>The <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> <span class="citation" data-cites="yu_ggtree:_2017">(<a href="#ref-yu_ggtree:_2017" role="doc-biblioref">Yu et al., 2017</a>)</span> implements <code>geom_cladelab()</code> layer to annotate a selected clade with a bar indicating the clade with a corresponding label.</p>
<p>The <code>geom_cladelab()</code> layer accepts a selected internal node number and labels the corresponding clade automatically (Figure <span class="quarto-unresolved-ref">?fig-cladelabelA</span>). To get the internal node number, please refer to <a href="#accesor-tidytree">Chapter 2</a>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2015-12-21</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="dv">30</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="cn">NA</span>, <span class="dv">8</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">45</span>, <span class="at">label=</span><span class="st">"test label"</span>) <span class="sc">+</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">34</span>, <span class="at">label=</span><span class="st">"another clade"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Users can set the parameter, <code>align = TRUE</code>, to align the clade label, <code>offset</code>, to adjust the position and color to set the color of the bar and label text, <em>etc.</em> (Figure <span class="quarto-unresolved-ref">?fig-cladelabelB</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">45</span>, <span class="at">label=</span><span class="st">"test label"</span>, <span class="at">align=</span><span class="cn">TRUE</span>,  </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">offset =</span> .<span class="dv">2</span>, <span class="at">textcolor=</span><span class="st">'red'</span>, <span class="at">barcolor=</span><span class="st">'red'</span>) <span class="sc">+</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">34</span>, <span class="at">label=</span><span class="st">"another clade"</span>, <span class="at">align=</span><span class="cn">TRUE</span>, </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">offset =</span> .<span class="dv">2</span>, <span class="at">textcolor=</span><span class="st">'blue'</span>, <span class="at">barcolor=</span><span class="st">'blue'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Users can change the <code>angle</code> of the clade label text and relative position from text to bar via the parameter <code>offset.text</code>. The size of the bar and text can be changed via the parameters <code>barsize</code> and <code>fontsize</code>, respectively (Figure <span class="quarto-unresolved-ref">?fig-cladelabelC</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">45</span>, <span class="at">label=</span><span class="st">"test label"</span>, <span class="at">align=</span><span class="cn">TRUE</span>, <span class="at">angle=</span><span class="dv">270</span>, </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">hjust=</span><span class="st">'center'</span>, <span class="at">offset.text=</span>.<span class="dv">5</span>, <span class="at">barsize=</span><span class="fl">1.5</span>, <span class="at">fontsize=</span><span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">34</span>, <span class="at">label=</span><span class="st">"another clade"</span>, <span class="at">align=</span><span class="cn">TRUE</span>, <span class="at">angle=</span><span class="dv">45</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.
ℹ The deprecated feature was likely used in the ggtree package.
  Please report the issue at &lt;https://github.com/YuLab-SMU/ggtree/issues&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Users can also use <code>geom_label()</code> to label the text and can set the background color by <code>fill</code> parameter (Figure <span class="quarto-unresolved-ref">?fig-cladelabelD</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">34</span>, <span class="at">label=</span><span class="st">"another clade"</span>, <span class="at">align=</span><span class="cn">TRUE</span>, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">geom=</span><span class="st">'label'</span>, <span class="at">fill=</span><span class="st">'lightblue'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in (function (mapping = NULL, data = NULL, ..., stat = "identity", :
Ignoring unknown parameters: `label.size`</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Warning in (function (mapping = NULL, data = NULL, ..., stat = "identity", :
Ignoring unknown parameters: `label.size`</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Labeling clades.</strong> Default (A); aligning and coloring clade bar and text (B); changing size and angle (C) and using <code>geom_label()</code> with background color in the text (D).</figcaption>
</figure>
</div>
</div>
</div>
<p>In addition, <code>geom_cladelab()</code> allows users to use the image or phylopic to annotate the clades, and supports using aesthetic mapping to automatically annotate the clade with bar and text label or image (<em>e.g.</em>, mapping variable to color the clade labels) (Figure <span class="quarto-unresolved-ref">?fig-cladelabaes</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">node =</span> <span class="fu">c</span>(<span class="dv">45</span>, <span class="dv">34</span>), </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"test label"</span>, <span class="st">"another clade"</span>))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The node and label is required when geom="text" </span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="do">## or geom="label" or geom="shadowtext".</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_cladelab</span>(<span class="at">data =</span> dat, </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">node =</span> node, <span class="at">label =</span> name, <span class="at">color =</span> name), </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">fontsize =</span> <span class="dv">3</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>dt <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">node =</span> <span class="fu">c</span>(<span class="dv">45</span>, <span class="dv">34</span>), </span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>                 <span class="at">image =</span> <span class="fu">c</span>(<span class="st">"7fb9bea8-e758-4986-afb2-95a2c3bf983d"</span>, </span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"0174801d-15a6-4668-bfe0-4c421fbe51e8"</span>), </span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>                 <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"specie A"</span>, <span class="st">"specie B"</span>))</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="co"># when geom="phylopic" or geom="image", the image of aes is required.</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_cladelab</span>(<span class="at">data =</span> dt, </span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>                <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">node =</span> node, <span class="at">label =</span> name, <span class="at">image =</span> image), </span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>                <span class="at">geom =</span> <span class="st">"phylopic"</span>, <span class="at">imagecolor =</span> <span class="st">"black"</span>, </span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>                <span class="at">offset=</span><span class="dv">1</span>, <span class="at">offset.text=</span><span class="fl">0.5</span>)</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="co"># The color or size of image also can be mapped.</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_cladelab</span>(<span class="at">data =</span> dt, </span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>              <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">node =</span> node, <span class="at">label =</span> name, </span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>                          <span class="at">image =</span> image, <span class="at">color =</span> name), </span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>              <span class="at">geom =</span> <span class="st">"phylopic"</span>, <span class="at">offset =</span> <span class="dv">1</span>, <span class="at">offset.text=</span><span class="fl">0.5</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Labeling clades using aesthetic mapping.</strong> The geom_cladelab() layer allows users to use aesthetic mapping to annotate the clades (A); it supports using images or phylopic to annotate clades (B); mapping variable to change color or size of the text or image is also supported (C).</figcaption>
</figure>
</div>
</div>
</div>
<p>The <code>geom_cladelab()</code> layer also supports unrooted tree layouts (Figure <span class="quarto-unresolved-ref">?fig-striplabelA</span>).</p>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree, <span class="at">layout=</span><span class="st">"daylight"</span>) <span class="sc">+</span> </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">35</span>, <span class="at">label=</span><span class="st">"test label"</span>, <span class="at">angle=</span><span class="dv">0</span>, </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fontsize=</span><span class="dv">8</span>, <span class="at">offset=</span>.<span class="dv">5</span>, <span class="at">vjust=</span>.<span class="dv">5</span>)  <span class="sc">+</span> </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_cladelab</span>(<span class="at">node=</span><span class="dv">55</span>, <span class="at">label=</span><span class="st">'another clade'</span>, </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">angle=</span><span class="sc">-</span><span class="dv">95</span>, <span class="at">hjust=</span>.<span class="dv">5</span>, <span class="at">fontsize=</span><span class="dv">8</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The <code>geom_cladelab()</code> is designed for labeling Monophyletic (Clade) while there are related taxa that do not form a clade. In <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a>, we provide another layer, <code>geom_strip()</code>, to add a strip/bar to indicate the association with an optional label for Polyphyletic or Paraphyletic (Figure <span class="quarto-unresolved-ref">?fig-striplabelB</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_strip</span>(<span class="st">'t10'</span>, <span class="st">'t30'</span>, <span class="at">barsize=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">'red'</span>, </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">label=</span><span class="st">"associated taxa"</span>, <span class="at">offset.text=</span>.<span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_strip</span>(<span class="st">'t1'</span>, <span class="st">'t18'</span>, <span class="at">barsize=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">'blue'</span>, </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="st">"another label"</span>, <span class="at">offset.text=</span>.<span class="dv">1</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Labeling associated taxa.</strong> The <code>geom_cladelab()</code> is designed for labeling Monophyletic and supports unrooted layouts (A). The <code>geom_strip()</code> is designed for labeling all types of associated taxa, including Monophyletic, Polyphyletic, and Paraphyletic (B).</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="highlight-clades" class="level3" data-number="4.2.2">
<h3 data-number="4.2.2" class="anchored" data-anchor-id="highlight-clades"><span class="header-section-number">4.2.2</span> Highlight clades</h3>
<p>The <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> implements the <code>geom_hilight()</code> layer, which accepts an internal node number and adds a layer of a rectangle to highlight the selected clade (Figure <span class="quarto-unresolved-ref">?fig-hilight</span>)<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>nwk <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"sample.nwk"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(nwk)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree) <span class="sc">+</span> </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hilight</span>(<span class="at">node=</span><span class="dv">21</span>, <span class="at">fill=</span><span class="st">"steelblue"</span>, <span class="at">alpha=</span>.<span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hilight</span>(<span class="at">node=</span><span class="dv">17</span>, <span class="at">fill=</span><span class="st">"darkgreen"</span>, <span class="at">alpha=</span>.<span class="dv">6</span>) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="480"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree, <span class="at">layout=</span><span class="st">"circular"</span>) <span class="sc">+</span> </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hilight</span>(<span class="at">node=</span><span class="dv">21</span>, <span class="at">fill=</span><span class="st">"steelblue"</span>, <span class="at">alpha=</span>.<span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hilight</span>(<span class="at">node=</span><span class="dv">23</span>, <span class="at">fill=</span><span class="st">"darkgreen"</span>, <span class="at">alpha=</span>.<span class="dv">6</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>The <code>geom_hilight</code> layer also supports highlighting clades for unrooted layout trees with round (‘encircle’) or rectangular (‘rect’) shape (Figure <span class="quarto-unresolved-ref">?fig-hilightC</span>).</p>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="do">## type can be 'encircle' or 'rect'</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>pg <span class="sc">+</span> <span class="fu">geom_hilight</span>(<span class="at">node=</span><span class="dv">55</span>, <span class="at">linetype =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hilight</span>(<span class="at">node=</span><span class="dv">35</span>, <span class="at">fill=</span><span class="st">'darkgreen'</span>, <span class="at">type=</span><span class="st">"rect"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>Another way to highlight selected clades is by setting the clades with different colors and/or line types as demonstrated in Figure <span class="quarto-unresolved-ref">?fig-scaleClade</span>.</p>
<p>In addition to <code>geom_hilight()</code>, <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> also implements <code>geom_balance()</code> which is designed to highlight neighboring subclades of a given internal node (Figure <span class="quarto-unresolved-ref">?fig-hilightD</span>).</p>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree) <span class="sc">+</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_balance</span>(<span class="at">node=</span><span class="dv">16</span>, <span class="at">fill=</span><span class="st">'steelblue'</span>, <span class="at">color=</span><span class="st">'white'</span>, <span class="at">alpha=</span><span class="fl">0.6</span>, <span class="at">extend=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_balance</span>(<span class="at">node=</span><span class="dv">19</span>, <span class="at">fill=</span><span class="st">'darkgreen'</span>, <span class="at">color=</span><span class="st">'white'</span>, <span class="at">alpha=</span><span class="fl">0.6</span>, <span class="at">extend=</span><span class="dv">1</span>) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The <code>geom_hilight()</code> layer supports using aesthetic mapping to automatically highlight clades as demonstrated in Figures <span class="quarto-unresolved-ref">?fig-hilightE-F</span>. For plot in Cartesian coordinates (<em>e.g.</em>, rectangular layout), the rectangle can be rounded (Figure <span class="quarto-unresolved-ref">?fig-hilightE</span>) or filled with gradient colors (Figure <span class="quarto-unresolved-ref">?fig-hilightF</span>).</p>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="do">## using external data</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">node=</span><span class="fu">c</span>(<span class="dv">17</span>, <span class="dv">21</span>), <span class="at">type=</span><span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>))</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree) <span class="sc">+</span> <span class="fu">geom_hilight</span>(<span class="at">data=</span>d, <span class="fu">aes</span>(<span class="at">node=</span>node, <span class="at">fill=</span>type),</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">type =</span> <span class="st">"roundrect"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="480"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="do">## using data stored in the tree object</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">read.nhx</span>(<span class="fu">system.file</span>(<span class="st">"extdata/NHX/ADH.nhx"</span>, <span class="at">package=</span><span class="st">"treeio"</span>))</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(x) <span class="sc">+</span> <span class="fu">geom_hilight</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">subset =</span> node <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">12</span>), </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">fill =</span> S),</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">type =</span> <span class="st">"gradient"</span>, <span class="at">gradient.direction =</span> <span class="st">'rt'</span>,</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">alpha =</span> .<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"steelblue"</span>, <span class="st">"darkgreen"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Highlight selected clades.</strong> Rectangular layout (A), circular/fan (B), and unrooted layouts. Highlight neighboring subclades simultaneously (D). Highlight selected clades using associated data (E and F).</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="taxa-connection" class="level3" data-number="4.2.3">
<h3 data-number="4.2.3" class="anchored" data-anchor-id="taxa-connection"><span class="header-section-number">4.2.3</span> Taxa connection</h3>
<p>Some evolutionary events (<em>e.g.</em>, reassortment, horizontal gene transfer) cannot be modeled by a simple tree. The <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> provides the <code>geom_taxalink()</code> layer that allows drawing straight or curved lines between any of two nodes in the tree, allowing it to represent evolutionary events by connecting taxa. It works with rectangular (Figure <span class="quarto-unresolved-ref">?fig-taxalinkA</span>), circular (Figure <span class="quarto-unresolved-ref">?fig-taxalinkB</span>), and inward circular (Figure <span class="quarto-unresolved-ref">?fig-taxalinkC</span>) layouts. The <code>geom_taxalink()</code> is not only useful for presenting evolutionary events, but it can also be used to combine evolutionary trees to present relationships or interactions between species <span class="citation" data-cites="ggtreeExtra_2021">(<a href="#ref-ggtreeExtra_2021" role="doc-biblioref">Xu et al., 2021</a>)</span>.</p>
<p>The <code>geom_taxalink()</code> layout supports aesthetic mapping, which requires a <code>data.frame</code> that stores association information with/without meta-data as input (Figure <span class="quarto-unresolved-ref">?fig-taxalinkD</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> <span class="fu">geom_taxalink</span>(<span class="at">taxa1=</span><span class="st">'A'</span>, <span class="at">taxa2=</span><span class="st">'E'</span>) <span class="sc">+</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_taxalink</span>(<span class="at">taxa1=</span><span class="st">'F'</span>, <span class="at">taxa2=</span><span class="st">'K'</span>, <span class="at">color=</span><span class="st">'red'</span>, <span class="at">linetype =</span> <span class="st">'dashed'</span>,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.02</span>, <span class="st">"npc"</span>)))</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree, <span class="at">layout=</span><span class="st">"circular"</span>) <span class="sc">+</span> </span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_taxalink</span>(<span class="at">taxa1=</span><span class="st">'A'</span>, <span class="at">taxa2=</span><span class="st">'E'</span>, <span class="at">color=</span><span class="st">"grey"</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, </span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">offset=</span><span class="fl">0.05</span>, <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.01</span>, <span class="st">"npc"</span>))) <span class="sc">+</span> </span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_taxalink</span>(<span class="at">taxa1=</span><span class="st">'F'</span>, <span class="at">taxa2=</span><span class="st">'K'</span>, <span class="at">color=</span><span class="st">'red'</span>, </span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">offset=</span><span class="fl">0.05</span>,</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.01</span>, <span class="st">"npc"</span>))) <span class="sc">+</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_taxalink</span>(<span class="at">taxa1=</span><span class="st">"L"</span>, <span class="at">taxa2=</span><span class="st">"M"</span>, <span class="at">color=</span><span class="st">"blue"</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, </span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">offset=</span><span class="fl">0.05</span>, <span class="at">hratio=</span><span class="fl">0.8</span>, </span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.01</span>, <span class="st">"npc"</span>))) <span class="sc">+</span> </span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_tiplab</span>()</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a><span class="co"># when the tree was created using reverse x, </span></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a><span class="co"># we can set outward to FALSE, which will generate the inward curve lines.</span></span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree, <span class="at">layout=</span><span class="st">"inward_circular"</span>, <span class="at">xlim=</span><span class="dv">150</span>) <span class="sc">+</span></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_taxalink</span>(<span class="at">taxa1=</span><span class="st">'A'</span>, <span class="at">taxa2=</span><span class="st">'E'</span>, <span class="at">color=</span><span class="st">"grey"</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, </span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>                    <span class="at">offset=</span><span class="sc">-</span><span class="fl">0.2</span>, <span class="at">outward=</span><span class="cn">FALSE</span>,</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>                    <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.01</span>, <span class="st">"npc"</span>))) <span class="sc">+</span></span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_taxalink</span>(<span class="at">taxa1=</span><span class="st">'F'</span>, <span class="at">taxa2=</span><span class="st">'K'</span>, <span class="at">color=</span><span class="st">'red'</span>, <span class="at">linetype =</span> <span class="st">'dashed'</span>, </span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>                    <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">offset=</span><span class="sc">-</span><span class="fl">0.2</span>, <span class="at">outward=</span><span class="cn">FALSE</span>,</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a>                    <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.01</span>, <span class="st">"npc"</span>))) <span class="sc">+</span></span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_taxalink</span>(<span class="at">taxa1=</span><span class="st">"L"</span>, <span class="at">taxa2=</span><span class="st">"M"</span>, <span class="at">color=</span><span class="st">"blue"</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, </span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>                    <span class="at">offset=</span><span class="sc">-</span><span class="fl">0.2</span>, <span class="at">outward=</span><span class="cn">FALSE</span>,</span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a>                    <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.01</span>, <span class="st">"npc"</span>))) <span class="sc">+</span></span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_tiplab</span>(<span class="at">hjust=</span><span class="dv">1</span>) </span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">from=</span><span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"F"</span>, <span class="st">"L"</span>), </span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>                  <span class="at">to=</span><span class="fu">c</span>(<span class="st">"E"</span>, <span class="st">"K"</span>, <span class="st">"M"</span>), </span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a>                  <span class="at">h=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.1</span>), </span>
<span id="cb26-33"><a href="#cb26-33" aria-hidden="true" tabindex="-1"></a>                  <span class="at">type=</span><span class="fu">c</span>(<span class="st">"t1"</span>, <span class="st">"t2"</span>, <span class="st">"t3"</span>), </span>
<span id="cb26-34"><a href="#cb26-34" aria-hidden="true" tabindex="-1"></a>                  <span class="at">s=</span><span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb26-35"><a href="#cb26-35" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree, <span class="at">layout=</span><span class="st">"inward_circular"</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">150</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb26-36"><a href="#cb26-36" aria-hidden="true" tabindex="-1"></a>          <span class="fu">geom_taxalink</span>(<span class="at">data=</span>dat, </span>
<span id="cb26-37"><a href="#cb26-37" aria-hidden="true" tabindex="-1"></a>                         <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">taxa1=</span>from, </span>
<span id="cb26-38"><a href="#cb26-38" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">taxa2=</span>to, </span>
<span id="cb26-39"><a href="#cb26-39" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">color=</span>type, </span>
<span id="cb26-40"><a href="#cb26-40" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">size=</span>s), </span>
<span id="cb26-41"><a href="#cb26-41" aria-hidden="true" tabindex="-1"></a>                         <span class="at">ncp=</span><span class="dv">10</span>,</span>
<span id="cb26-42"><a href="#cb26-42" aria-hidden="true" tabindex="-1"></a>                         <span class="at">offset=</span><span class="fl">0.15</span>) <span class="sc">+</span> </span>
<span id="cb26-43"><a href="#cb26-43" aria-hidden="true" tabindex="-1"></a>          <span class="fu">geom_tiplab</span>(<span class="at">hjust=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb26-44"><a href="#cb26-44" aria-hidden="true" tabindex="-1"></a>          <span class="fu">scale_size_continuous</span>(<span class="at">range=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb26-45"><a href="#cb26-45" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_list</span>(p1, p2, p3, p4, <span class="at">ncol=</span><span class="dv">2</span>, <span class="at">tag_levels=</span><span class="st">'A'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Linking related taxa.</strong> This can be used to indicate evolutionary events or relationships between species. Rectangular layout (A), circular layout (B), and inward circular layout (C and D). It supports aesthetic mapping to map variables to set line sizes and colors (D).</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="uncertainty-of-evolutionary-inference" class="level3" data-number="4.2.4">
<h3 data-number="4.2.4" class="anchored" data-anchor-id="uncertainty-of-evolutionary-inference"><span class="header-section-number">4.2.4</span> Uncertainty of evolutionary inference</h3>
<p>The <code>geom_range()</code> layer supports displaying interval (highest posterior density, confidence interval, range) as horizontal bars on tree nodes. The center of the interval will anchor to the corresponding node. The center by default is the mean value of the interval (Figure <span class="quarto-unresolved-ref">?fig-geomRangeA</span>). We can set the <code>center</code> to the estimated mean or median value (Figure <span class="quarto-unresolved-ref">?fig-geomRangeB</span>), or the observed value. As the tree branch and the interval may not be on the same scale, <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> provides <code>scale_x_range</code> to add a second <em>x</em>-axis for the range (Figure <span class="quarto-unresolved-ref">?fig-geomRangeC</span>). Note that <em>x</em>-axis is disabled by the default theme, and we need to enable it if we want to display it (<em>e.g.</em>, using <code>theme_tree2()</code>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/MEGA7"</span>, <span class="st">"mtCDNA_timetree.nex"</span>, <span class="at">package =</span> <span class="st">"treeio"</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">read.mega</span>(file)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(x) <span class="sc">+</span> <span class="fu">geom_range</span>(<span class="st">'reltime_0.95_CI'</span>, <span class="at">color=</span><span class="st">'red'</span>, <span class="at">size=</span><span class="dv">3</span>, <span class="at">alpha=</span>.<span class="dv">3</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(x) <span class="sc">+</span> <span class="fu">geom_range</span>(<span class="st">'reltime_0.95_CI'</span>, <span class="at">color=</span><span class="st">'red'</span>, <span class="at">size=</span><span class="dv">3</span>, </span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>                              <span class="at">alpha=</span>.<span class="dv">3</span>, <span class="at">center=</span><span class="st">'reltime'</span>)  </span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> p2 <span class="sc">+</span> <span class="fu">scale_x_range</span>() <span class="sc">+</span> <span class="fu">theme_tree2</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Displaying uncertainty of evolutionary inference.</strong> The center (mean value of the range (A) or estimated value (B)) is anchored to the tree nodes. A second <em>x</em>-axis was used for range scaling (C).</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="tree-annotation-with-output-from-evolution-software" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="tree-annotation-with-output-from-evolution-software"><span class="header-section-number">4.3</span> Tree Annotation with Output from Evolution Software</h2>
<section id="tree-annotation-using-data-from-evolutionary-analysis-software" class="level3" data-number="4.3.1">
<h3 data-number="4.3.1" class="anchored" data-anchor-id="tree-annotation-using-data-from-evolutionary-analysis-software"><span class="header-section-number">4.3.1</span> Tree annotation using data from evolutionary analysis software</h3>
<p><a href="#chapter">Chapter 1</a> introduced using <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> package <span class="citation" data-cites="wang_treeio_2020">(<a href="#ref-wang_treeio_2020" role="doc-biblioref">Wang et al., 2020</a>)</span> to parse different tree formats and commonly used software outputs to obtain phylogeny-associated data. These imported data, as <code>S4</code> objects, can be visualized directly using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a>. Figure <span class="quarto-unresolved-ref">?fig-ggtreeNHX</span> demonstrates a tree annotated using the information (species classification, duplication event, and bootstrap value) stored in the NHX file. <a href="http://pbil.univ-lyon1.fr/software/phyldog/"><strong>PHYLDOG</strong></a> and <code>pkg_revbayes</code> output NHX files that can be parsed by <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> and visualized by <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> with annotation using their inference data.</p>
<p>Furthermore, the evolutionary data from the inference of <a href="http://beast2.org/"><strong>BEAST</strong></a>, <a href="http://nbisweden.github.io/MrBayes/"><strong>MrBayes</strong></a>, and <a href="http://revbayes.github.io/intro.html"><strong>RevBayes</strong></a>, <em>d<sub>N</sub>/d<sub>S</sub></em> values inferred by <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>, ancestral sequences inferred by <a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a>, <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a>, or <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> and short read placement by <a href="http://sco.h-its.org/exelixis/web/software/epa/index.html"><strong>EPA</strong></a> and <a href="http://matsen.fhcrc.org/pplacer/"><strong>PPLACER</strong></a> can be used to annotate the tree directly.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/BEAST"</span>, <span class="st">"beast_mcc.tree"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>beast <span class="ot">&lt;-</span> <span class="fu">read.beast</span>(file)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(beast, <span class="fu">aes</span>(<span class="at">color=</span>rate))  <span class="sc">+</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_range</span>(<span class="at">range=</span><span class="st">'length_0.95_HPD'</span>, <span class="at">color=</span><span class="st">'red'</span>, <span class="at">alpha=</span>.<span class="dv">6</span>, <span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_nodelab</span>(<span class="fu">aes</span>(<span class="at">x=</span>branch, <span class="at">label=</span><span class="fu">round</span>(posterior, <span class="dv">2</span>)), <span class="at">vjust=</span><span class="sc">-</span>.<span class="dv">5</span>, <span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_continuous</span>(<span class="at">low=</span><span class="st">"darkgreen"</span>, <span class="at">high=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="fu">c</span>(.<span class="dv">1</span>, .<span class="dv">8</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Annotating <code>r pkg_beast</code> tree with <em>length_95%_HPD</em> and posterior.</strong> Branch length credible intervals (95% HPD) were displayed as red horizontal bars and clade posterior values were shown on the middle of branches.</figcaption>
</figure>
</div>
</div>
</div>
<p>In Figure <span class="quarto-unresolved-ref">?fig-beast</span>, the tree was visualized and annotated with posterior &gt;0.9 and demonstrated length uncertainty (95% Highest Posterior Density (HPD) interval).</p>
<p>Ancestral sequences inferred by <a href="https://veg.github.io/hyphy-site/"><strong>HyPhy</strong></a> can be parsed using <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a>, whereas the substitutions along each tree branch were automatically computed and stored inside the phylogenetic tree object (<em>i.e.</em>, <code>S4</code> object). The <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> package can utilize this information stored in the object to annotate the tree, as demonstrated in Figure <span class="quarto-unresolved-ref">?fig-hyphy</span>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>nwk <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/HYPHY"</span>, <span class="st">"labelledtree.tree"</span>, </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>ancseq <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/HYPHY"</span>, <span class="st">"ancseq.nex"</span>, </span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>tipfas <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"pa.fas"</span>, <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>hy <span class="ot">&lt;-</span> <span class="fu">read.hyphy</span>(nwk, ancseq, tipfas)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(hy) <span class="sc">+</span> </span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x=</span>branch, <span class="at">label=</span>AA_subs), <span class="at">size=</span><span class="dv">2</span>, </span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">vjust=</span><span class="sc">-</span>.<span class="dv">3</span>, <span class="at">color=</span><span class="st">"firebrick"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Annotating tree with amino acid substitution determined by ancestral sequences inferred by <code>r pkg_hyphy</code>.</strong> Amino acid substitutions were displayed in the middle of branches.</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>PAML</strong></a>’s <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>BASEML</strong></a> and <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> can also be used to infer ancestral sequences, whereas <a href="http://abacus.gene.ucl.ac.uk/software/paml.html"><strong>CODEML</strong></a> can infer selection pressure. After parsing this information using <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a>, <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> can integrate this information into the same tree structure and be used for annotation as illustrated in Figure <span class="quarto-unresolved-ref">?fig-codeml</span>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>rstfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/PAML_Codeml"</span>, <span class="st">"rst"</span>, </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>mlcfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata/PAML_Codeml"</span>, <span class="st">"mlc"</span>, </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">package=</span><span class="st">"treeio"</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>ml <span class="ot">&lt;-</span> <span class="fu">read.codeml</span>(rstfile, mlcfile)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(ml, <span class="fu">aes</span>(<span class="at">color=</span>dN_vs_dS), <span class="at">branch.length=</span><span class="st">'dN_vs_dS'</span>) <span class="sc">+</span> </span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_continuous</span>(<span class="at">name=</span><span class="st">'dN/dS'</span>, <span class="at">limits=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">1.5</span>),</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>                         <span class="at">oob=</span>scales<span class="sc">::</span>squish,</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>                         <span class="at">low=</span><span class="st">'darkgreen'</span>, <span class="at">high=</span><span class="st">'red'</span>) <span class="sc">+</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x=</span>branch, <span class="at">label=</span>AA_subs), </span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">vjust=</span><span class="sc">-</span>.<span class="dv">5</span>, <span class="at">color=</span><span class="st">'steelblue'</span>, <span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tree2</span>(<span class="at">legend.position=</span><span class="fu">c</span>(.<span class="dv">9</span>, .<span class="dv">3</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="05_ggtree_annotation_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Annotating tree with amino acid substitution and <em>d<sub>N</sub>/d<sub>S</sub></em> inferred by <code>r pkg_codeml</code>.</strong> Branches were rescaled and colored by <em>d<sub>N</sub>/d<sub>S</sub></em> values, and amino acid substitutions were displayed on the middle of branches.</figcaption>
</figure>
</div>
</div>
</div>
<p>Not only all the tree data parsed by <a href="http://bioconductor.org/packages/treeio"><strong>treeio</strong></a> can be used to visualize and annotate the phylogenetic tree using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a>, but also other trees and tree-like objects defined in the R community are supported. The <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> plays a unique role in the R ecosystem to facilitate phylogenetic analysis, and it can be easily integrated into other packages and pipelines. For more details of working with other tree-like structures, please refer to <a href="#chapter9">Chapter 9</a>. In addition to direct support of tree objects, <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> also allows users to plot a tree with different types of external data (see also <a href="#chapter7">Chapter 7</a> and <span class="citation" data-cites="yu_two_2018">(<a href="#ref-yu_two_2018" role="doc-biblioref">Yu et al., 2018</a>)</span>).</p>
</section>
</section>
<section id="sec-summary5" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="sec-summary5"><span class="header-section-number">4.4</span> Summary</h2>
<p>The <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> package implements the grammar of graphics for annotating phylogenetic trees. Users can use the <a href="https://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a> syntax to combine different annotation layers to produce complex tree annotation. If you are familiar with <a href="https://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a>, tree annotation with a high level of customization can be intuitive and flexible using <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a>. The <a href="http://bioconductor.org/packages/ggtree"><strong>ggtree</strong></a> can collect information in the <code>treedata</code> object or link external data to the structure of the tree. This will enable us to use the phylogenetic tree for data integration analysis and comparative studies, and will greatly expand the application of the phylogenetic tree in different fields.</p>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-wang_treeio_2020" class="csl-entry" role="listitem">
Wang, L.-G., Lam, T. T.-Y., Xu, S., Dai, Z., Zhou, L., Feng, T., Guo, P., Dunn, C. W., Jones, B. R., Bradley, T., Zhu, H., Guan, Y., Jiang, Y., &amp; Yu, G. (2020). Treeio: An r package for phylogenetic tree input and output with richly annotated and associated data. <em>Molecular Biology and Evolution</em>, <em>37</em>(2), 599–603. <a href="https://doi.org/10.1093/molbev/msz240">https://doi.org/10.1093/molbev/msz240</a>
</div>
<div id="ref-ggtreeExtra_2021" class="csl-entry" role="listitem">
Xu, S., Dai, Z., Guo, P., Fu, X., Liu, S., Zhou, L., Tang, W., Feng, T., Chen, M., Zhan, L., Wu, T., Hu, E., Jiang, Y., Bo, X., &amp; Yu, G. (2021). <span class="nocase">ggtreeExtra: Compact Visualization of Richly Annotated Phylogenetic Data</span>. <em>Molecular Biology and Evolution</em>, <em>38</em>(9), 4039–4042. <a href="https://doi.org/10.1093/molbev/msab166">https://doi.org/10.1093/molbev/msab166</a>
</div>
<div id="ref-yu_two_2018" class="csl-entry" role="listitem">
Yu, G., Lam, T. T.-Y., Zhu, H., &amp; Guan, Y. (2018). Two methods for mapping and visualizing associated data on phylogeny using ggtree. <em>Molecular Biology and Evolution</em>, <em>35</em>(12), 3041–3043. <a href="https://doi.org/10.1093/molbev/msy194">https://doi.org/10.1093/molbev/msy194</a>
</div>
<div id="ref-yu_ggtree:_2017" class="csl-entry" role="listitem">
Yu, G., Smith, D. K., Zhu, H., Guan, Y., &amp; Lam, T. T.-Y. (2017). Ggtree: An r package for visualization and annotation of phylogenetic trees with their covariates and other associated data. <em>Methods in Ecology and Evolution</em>, <em>8</em>(1), 28–36. <a href="https://doi.org/10.1111/2041-210X.12628">https://doi.org/10.1111/2041-210X.12628</a>
</div>
</div>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>If you want to plot the tree above the highlighting area, visit <a href="#faq-under-the-tree">FAQ</a> for details.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./04_ggtree_visualization.html" class="pagination-link" aria-label="(PART\*) Part II: Tree data visualization and annotation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">(PART*) Part II: Tree data visualization and annotation</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./06_ggtree_manipulation.html" class="pagination-link" aria-label="Visual Exploration of Phylogenetic Trees">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Visual Exploration of Phylogenetic Trees</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/YuLab-SMU/treedata-book/edit/main/05_ggtree_annotation.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/blob/main/05_ggtree_annotation.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div></div></footer></body></html>