<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Master ggtree package suite to handle tree with data.">

<title>10&nbsp; Gallery of Reproducible Examples – Data Integration, Manipulation and Visualization of Phylogenetic Trees</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./A-app-faq.html" rel="next">
<link href="./12_ggtree_utilities.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-0dd2bd5de344125cf763a379ddc3eb04.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="css/style.css">
</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./13_ggtree_gallery.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Gallery of Reproducible Examples</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Data Integration, Manipulation and Visualization of Phylogenetic Trees</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/YuLab-SMU/treedata-book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Data Integration, Manipulation and Visualization of Phylogenetic Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./00_author.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About the Author</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_treeio_importing_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part I: Tree data input, output, and manipulation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_tidytree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Manipulating Tree with Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_treeio_exporting_tree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Exporting tree with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_ggtree_visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part II: Tree data visualization and annotation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_ggtree_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Phylogenetic Tree Annotation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_ggtree_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Visual Exploration of Phylogenetic Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_ggtree_tree_with_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Plotting tree with data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_ggtree_with_silhouette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Annotating Tree with Silhouette Images and Sub-plots</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_ggtree_tree_objects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">ggtree for Other Tree-like Objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_ggtree_exts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part III: ggtree extensions</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11_ggtree_exts-others.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Other ggtree Extensions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12_ggtree_utilities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(PART*) Part IV: Miscellaneous topics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13_ggtree_gallery.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Gallery of Reproducible Examples</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendix</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-faq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">(APPENDIX) Appendix</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Related Tools</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A-app-fig-tab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Figures and Tables</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-hpv58" id="toc-sec-hpv58" class="nav-link active" data-scroll-target="#sec-hpv58"><span class="header-section-number">10.1</span> Visualizing pairwise nucleotide sequence distance with a phylogenetic tree</a></li>
  <li><a href="#sec-symbolic-bootstrap" id="toc-sec-symbolic-bootstrap" class="nav-link" data-scroll-target="#sec-symbolic-bootstrap"><span class="header-section-number">10.2</span> Displaying Different Symbolic Points for Bootstrap Values.</a></li>
  <li><a href="#sec-phylo-grouping" id="toc-sec-phylo-grouping" class="nav-link" data-scroll-target="#sec-phylo-grouping"><span class="header-section-number">10.3</span> Highlighting Different Groups</a></li>
  <li><a href="#sec-genome-locus" id="toc-sec-genome-locus" class="nav-link" data-scroll-target="#sec-genome-locus"><span class="header-section-number">10.4</span> Phylogenetic Tree with Genome Locus Structure</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/YuLab-SMU/treedata-book/edit/main/13_ggtree_gallery.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/blob/main/13_ggtree_gallery.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-chapter13" class="quarto-section-identifier"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Gallery of Reproducible Examples</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div style="page-break-after: always;"></div>
<section id="sec-hpv58" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="sec-hpv58"><span class="header-section-number">10.1</span> Visualizing pairwise nucleotide sequence distance with a phylogenetic tree</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"yulab.utils"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>This example reproduces figure 1 of <span class="citation" data-cites="chen_ancient_2017">(<a href="#ref-chen_ancient_2017" role="doc-biblioref">Chen et al., 2017</a>)</span>. It extracts accession numbers from tip labels of the HPV58 tree and calculates pairwise nucleotide sequence distances. The distance matrix is visualized as dot and line plots. This example demonstrates the ability to add multiple layers to a specific panel. As illustrated in Figure <span class="quarto-unresolved-ref">?fig-jv2017</span>, the <code>geom_facet()</code> function displays sequence distances as a dot plot and then adds a layer of line plot to the same panel, <em>i.e.</em>, sequence distance. In addition, the tree in <code>geom_facet()</code> can be fully annotated with multiple layers (clade labels, bootstrap support values, <em>etc.</em>). The source code is modified from the supplemental file of <span class="citation" data-cites="yu_two_2018">(<a href="#ref-yu_two_2018" role="doc-biblioref">Yu et al., 2018</a>)</span>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TDbook)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:tidyr':

    expand</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: XVector</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biostrings'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:base':

    strsplit</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(treeio)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>treeio v1.35.0 Learn more at https://yulab-smu.top/contribution-tree-data/

Please cite:

LG Wang, TTY Lam, S Xu, Z Dai, L Zhou, T Feng, P Guo, CW Dunn, BR
Jones, T Bradley, H Zhu, Y Guan, Y Jiang, G Yu. treeio: an R package
for phylogenetic tree input and output with richly annotated and
associated data. Molecular Biology and Evolution. 2020, 37(2):599-603.
doi: 10.1093/molbev/msz240</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'treeio'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:Biostrings':

    mask</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>ggtree v4.1.1.003 Learn more at https://yulab-smu.top/contribution-tree-data/

Please cite:

S Xu, Z Dai, P Guo, X Fu, S Liu, L Zhou, W Tang, T Feng, M Chen, L
Zhan, T Wu, E Hu, Y Jiang, X Bo, G Yu. ggtreeExtra: Compact
visualization of richly annotated phylogenetic data. Molecular Biology
and Evolution. 2021, 38(9):4039-4042. doi: 10.1093/molbev/msab166</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ggtree'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:Biostrings':

    collapse</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:IRanges':

    collapse</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:S4Vectors':

    expand</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:tidyr':

    expand</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># loaded from TDbook package</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> tree_HPV58</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>clade <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">A3 =</span> <span class="dv">92</span>, <span class="at">A1 =</span> <span class="dv">94</span>, <span class="at">A2 =</span> <span class="dv">108</span>, <span class="at">B1 =</span> <span class="dv">156</span>, </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">B2 =</span> <span class="dv">159</span>, <span class="at">C =</span> <span class="dv">163</span>, <span class="at">D1 =</span> <span class="dv">173</span>, <span class="at">D2 =</span> <span class="dv">176</span>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">groupClade</span>(tree, clade)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">A1 =</span> <span class="st">"#EC762F"</span>, <span class="at">A2 =</span> <span class="st">"#CA6629"</span>, <span class="at">A3 =</span> <span class="st">"#894418"</span>, <span class="at">B1 =</span> <span class="st">"#0923FA"</span>, </span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">B2 =</span> <span class="st">"#020D87"</span>, <span class="at">C =</span> <span class="st">"#000000"</span>, <span class="at">D1 =</span> <span class="st">"#9ACD32"</span>,<span class="at">D2 =</span> <span class="st">"#08630A"</span>)</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a><span class="do">## visualize the tree with tip labels and tree scale</span></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree, <span class="fu">aes</span>(<span class="at">color =</span> group), <span class="at">ladderize =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rotate</span>(<span class="fu">rootnode</span>(tree)) <span class="sc">+</span> </span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">"italic('"</span>, label, <span class="st">"')"</span>)), </span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>                <span class="at">parse =</span> <span class="cn">TRUE</span>, <span class="at">size =</span> <span class="fl">2.5</span>) <span class="sc">+</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_treescale</span>(<span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">1</span>, <span class="at">width =</span> <span class="fl">0.002</span>) <span class="sc">+</span> </span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(cols, <span class="st">"black"</span>), </span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>                <span class="at">na.value =</span> <span class="st">"black"</span>, <span class="at">name =</span> <span class="st">"Lineage"</span>,</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>                <span class="at">breaks =</span> <span class="fu">c</span>(<span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"A3"</span>, <span class="st">"B1"</span>, <span class="st">"B2"</span>, <span class="st">"C"</span>, <span class="st">"D1"</span>, <span class="st">"D2"</span>)) <span class="sc">+</span></span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">5</span>, <span class="at">shape =</span> <span class="dv">15</span>))) <span class="sc">+</span></span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_tree2</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(.<span class="dv">1</span>, .<span class="dv">88</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.
ℹ The deprecated feature was likely used in the ggtree package.
  Please report the issue at &lt;https://github.com/YuLab-SMU/ggtree/issues&gt;.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Optional</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="do">## add labels for monophyletic (A, C and D) and paraphyletic (B) groups </span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">node =</span> <span class="fu">c</span>(<span class="dv">94</span>, <span class="dv">108</span>, <span class="dv">131</span>, <span class="dv">92</span>, <span class="dv">156</span>, <span class="dv">159</span>, <span class="dv">163</span>, <span class="dv">173</span>, <span class="dv">176</span>,<span class="dv">172</span>),</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"A3"</span>, <span class="st">"A"</span>, <span class="st">"B1"</span>, </span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"B2"</span>, <span class="st">"C"</span>, <span class="st">"D1"</span>, <span class="st">"D2"</span>, <span class="st">"D"</span>),</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">offset =</span> <span class="fu">c</span>(<span class="fl">0.003</span>, <span class="fl">0.003</span>, <span class="fl">0.003</span>, <span class="fl">0.00315</span>, <span class="fl">0.003</span>, </span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>                        <span class="fl">0.003</span>, <span class="fl">0.0031</span>, <span class="fl">0.003</span>, <span class="fl">0.003</span>, <span class="fl">0.00315</span>),</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">offset.text =</span> <span class="fu">c</span>(<span class="sc">-</span>.<span class="dv">001</span>, <span class="sc">-</span>.<span class="dv">001</span>, <span class="sc">-</span>.<span class="dv">001</span>, <span class="fl">0.0002</span>, <span class="sc">-</span>.<span class="dv">001</span>, </span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>                        <span class="sc">-</span>.<span class="dv">001</span>, <span class="fl">0.0002</span>, <span class="sc">-</span>.<span class="dv">001</span>, <span class="sc">-</span>.<span class="dv">001</span>, <span class="fl">0.0002</span>),</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>              <span class="at">barsize =</span> <span class="fu">c</span>(<span class="fl">1.2</span>, <span class="fl">1.2</span>, <span class="fl">1.2</span>, <span class="dv">2</span>, <span class="fl">1.2</span>, <span class="fl">1.2</span>, <span class="fl">3.2</span>, <span class="fl">1.2</span>, <span class="fl">1.2</span>, <span class="dv">2</span>),</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>              <span class="at">extend =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), <span class="fl">0.5</span>, <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">0</span>), <span class="dv">0</span>, <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), </span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">0</span>), <span class="dv">0</span>, <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">0</span>), <span class="dv">0</span>)</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>            ) <span class="sc">%&gt;%</span> </span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>       dplyr<span class="sc">::</span><span class="fu">group_split</span>(barsize)</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_cladelab</span>(</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">data =</span> dat[[<span class="dv">1</span>]],</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>             <span class="at">node =</span> node,</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>             <span class="at">label =</span> name,</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> group,</span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a>             <span class="at">offset =</span> offset,</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>             <span class="at">offset.text =</span> offset.text,</span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>             <span class="at">extend =</span> extend</span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a>         ),</span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a>         <span class="at">barsize =</span> <span class="fl">1.2</span>,</span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontface =</span> <span class="dv">3</span>,</span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a>         <span class="at">align =</span> <span class="cn">TRUE</span></span>
<span id="cb31-30"><a href="#cb31-30" aria-hidden="true" tabindex="-1"></a>     ) <span class="sc">+</span></span>
<span id="cb31-31"><a href="#cb31-31" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_cladelab</span>(</span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a>         <span class="at">data =</span> dat[[<span class="dv">2</span>]],</span>
<span id="cb31-33"><a href="#cb31-33" aria-hidden="true" tabindex="-1"></a>         <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb31-34"><a href="#cb31-34" aria-hidden="true" tabindex="-1"></a>             <span class="at">node =</span> node,</span>
<span id="cb31-35"><a href="#cb31-35" aria-hidden="true" tabindex="-1"></a>             <span class="at">label =</span> name,</span>
<span id="cb31-36"><a href="#cb31-36" aria-hidden="true" tabindex="-1"></a>             <span class="at">offset =</span> offset,</span>
<span id="cb31-37"><a href="#cb31-37" aria-hidden="true" tabindex="-1"></a>             <span class="at">offset.text =</span>offset.text,</span>
<span id="cb31-38"><a href="#cb31-38" aria-hidden="true" tabindex="-1"></a>             <span class="at">extend =</span> extend</span>
<span id="cb31-39"><a href="#cb31-39" aria-hidden="true" tabindex="-1"></a>         ),</span>
<span id="cb31-40"><a href="#cb31-40" aria-hidden="true" tabindex="-1"></a>         <span class="at">barcolor =</span> <span class="st">"darkgrey"</span>,</span>
<span id="cb31-41"><a href="#cb31-41" aria-hidden="true" tabindex="-1"></a>         <span class="at">textcolor =</span> <span class="st">"darkgrey"</span>,</span>
<span id="cb31-42"><a href="#cb31-42" aria-hidden="true" tabindex="-1"></a>         <span class="at">barsize =</span> <span class="dv">2</span>,</span>
<span id="cb31-43"><a href="#cb31-43" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontsize =</span> <span class="dv">5</span>,</span>
<span id="cb31-44"><a href="#cb31-44" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontface =</span> <span class="dv">3</span>,</span>
<span id="cb31-45"><a href="#cb31-45" aria-hidden="true" tabindex="-1"></a>         <span class="at">align =</span> <span class="cn">TRUE</span></span>
<span id="cb31-46"><a href="#cb31-46" aria-hidden="true" tabindex="-1"></a>     ) <span class="sc">+</span></span>
<span id="cb31-47"><a href="#cb31-47" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_cladelab</span>(</span>
<span id="cb31-48"><a href="#cb31-48" aria-hidden="true" tabindex="-1"></a>         <span class="at">data =</span> dat[[<span class="dv">3</span>]],</span>
<span id="cb31-49"><a href="#cb31-49" aria-hidden="true" tabindex="-1"></a>         <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb31-50"><a href="#cb31-50" aria-hidden="true" tabindex="-1"></a>             <span class="at">node =</span> node,</span>
<span id="cb31-51"><a href="#cb31-51" aria-hidden="true" tabindex="-1"></a>             <span class="at">label =</span> name,</span>
<span id="cb31-52"><a href="#cb31-52" aria-hidden="true" tabindex="-1"></a>             <span class="at">offset =</span> offset,</span>
<span id="cb31-53"><a href="#cb31-53" aria-hidden="true" tabindex="-1"></a>             <span class="at">offset.text =</span> offset.text,</span>
<span id="cb31-54"><a href="#cb31-54" aria-hidden="true" tabindex="-1"></a>             <span class="at">extend =</span> extend</span>
<span id="cb31-55"><a href="#cb31-55" aria-hidden="true" tabindex="-1"></a>         ),</span>
<span id="cb31-56"><a href="#cb31-56" aria-hidden="true" tabindex="-1"></a>         <span class="at">barcolor =</span> <span class="st">"darkgrey"</span>,</span>
<span id="cb31-57"><a href="#cb31-57" aria-hidden="true" tabindex="-1"></a>         <span class="at">textcolor =</span> <span class="st">"darkgrey"</span>,</span>
<span id="cb31-58"><a href="#cb31-58" aria-hidden="true" tabindex="-1"></a>         <span class="at">barsize =</span> <span class="fl">3.2</span>,</span>
<span id="cb31-59"><a href="#cb31-59" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontsize =</span> <span class="dv">5</span>,</span>
<span id="cb31-60"><a href="#cb31-60" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontface =</span> <span class="dv">3</span>,</span>
<span id="cb31-61"><a href="#cb31-61" aria-hidden="true" tabindex="-1"></a>         <span class="at">align =</span> <span class="cn">TRUE</span></span>
<span id="cb31-62"><a href="#cb31-62" aria-hidden="true" tabindex="-1"></a>     ) <span class="sc">+</span></span>
<span id="cb31-63"><a href="#cb31-63" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_strip</span>(<span class="dv">65</span>, <span class="dv">71</span>, <span class="st">"italic(B)"</span>, <span class="at">color =</span> <span class="st">"darkgrey"</span>, </span>
<span id="cb31-64"><a href="#cb31-64" aria-hidden="true" tabindex="-1"></a>                <span class="at">offset =</span> <span class="fl">0.00315</span>, <span class="at">align =</span> <span class="cn">TRUE</span>, <span class="at">offset.text =</span> <span class="fl">0.0002</span>, </span>
<span id="cb31-65"><a href="#cb31-65" aria-hidden="true" tabindex="-1"></a>                <span class="at">barsize =</span> <span class="dv">2</span>, <span class="at">fontsize =</span> <span class="dv">5</span>, <span class="at">parse =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-66"><a href="#cb31-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-67"><a href="#cb31-67" aria-hidden="true" tabindex="-1"></a><span class="do">## Optional</span></span>
<span id="cb31-68"><a href="#cb31-68" aria-hidden="true" tabindex="-1"></a><span class="do">## display support values</span></span>
<span id="cb31-69"><a href="#cb31-69" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_nodelab</span>(<span class="fu">aes</span>(<span class="at">subset =</span> (node <span class="sc">==</span> <span class="dv">92</span>), <span class="at">label =</span> <span class="st">"*"</span>), </span>
<span id="cb31-70"><a href="#cb31-70" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">nudge_x =</span> <span class="sc">-</span>.<span class="dv">001</span>, <span class="at">nudge_y =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb31-71"><a href="#cb31-71" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_nodelab</span>(<span class="fu">aes</span>(<span class="at">subset =</span> (node <span class="sc">==</span> <span class="dv">155</span>), <span class="at">label =</span> <span class="st">"*"</span>), </span>
<span id="cb31-72"><a href="#cb31-72" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">nudge_x =</span> <span class="sc">-</span>.<span class="dv">0003</span>, <span class="at">nudge_y =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb31-73"><a href="#cb31-73" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_nodelab</span>(<span class="fu">aes</span>(<span class="at">subset =</span> (node <span class="sc">==</span> <span class="dv">158</span>), <span class="at">label =</span> <span class="st">"95/92/1.00"</span>), </span>
<span id="cb31-74"><a href="#cb31-74" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">nudge_x =</span> <span class="sc">-</span><span class="fl">0.0001</span>, </span>
<span id="cb31-75"><a href="#cb31-75" aria-hidden="true" tabindex="-1"></a>                    <span class="at">nudge_y =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb31-76"><a href="#cb31-76" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_nodelab</span>(<span class="fu">aes</span>(<span class="at">subset =</span> (node <span class="sc">==</span> <span class="dv">162</span>), <span class="at">label =</span> <span class="st">"98/97/1.00"</span>), </span>
<span id="cb31-77"><a href="#cb31-77" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">nudge_x =</span> <span class="sc">-</span><span class="fl">0.0001</span>, </span>
<span id="cb31-78"><a href="#cb31-78" aria-hidden="true" tabindex="-1"></a>                    <span class="at">nudge_y =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb31-79"><a href="#cb31-79" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_nodelab</span>(<span class="fu">aes</span>(<span class="at">subset =</span> (node <span class="sc">==</span> <span class="dv">172</span>), <span class="at">label =</span> <span class="st">"*"</span>), </span>
<span id="cb31-80"><a href="#cb31-80" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">nudge_x =</span> <span class="sc">-</span>.<span class="dv">0003</span>, <span class="at">nudge_y =</span> <span class="sc">-</span><span class="dv">1</span>) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="do">## extract accession numbers from tip labels</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>tl <span class="ot">&lt;-</span> tree<span class="sc">$</span>tip.label</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>acc <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">w+</span><span class="sc">\\</span><span class="st">|"</span>, <span class="st">""</span>, tl)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(tl) <span class="ot">&lt;-</span> acc</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="do">## read sequences from GenBank directly into R</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="do">## and convert the object to DNAStringSet</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>tipseq <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">read.GenBank</span>(acc) <span class="sc">%&gt;%</span> as.character <span class="sc">%&gt;%</span> </span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lapply</span>(., paste0, <span class="at">collapse =</span> <span class="st">""</span>) <span class="sc">%&gt;%</span> unlist <span class="sc">%&gt;%</span> </span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>    DNAStringSet</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="do">## align the sequences using muscle</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>tipseq_aln <span class="ot">&lt;-</span> muscle<span class="sc">::</span><span class="fu">muscle</span>(tipseq)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
MUSCLE v3.8.31 by Robert C. Edgar

http://www.drive5.com/muscle
This software is donated to the public domain.
Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.

file1193d10bd1df0 90 seqs, max length 7863, avg  length 7826
1322 MB(8%)00:00:00                Iter   1    0.02%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1   12.23%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1   24.44%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1   36.65%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1   48.86%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1   61.07%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1   73.28%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1   85.49%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1   97.70%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1  100.00%  K-mer dist pass 1
1322 MB(8%)00:00:00                Iter   1    0.02%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1   12.23%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1   24.44%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1   36.65%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1   48.86%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1   61.07%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1   73.28%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1   85.49%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1   97.70%  K-mer dist pass 2
1322 MB(8%)00:00:00                Iter   1  100.00%  K-mer dist pass 2
1331 MB(8%)00:00:00                Iter   1    1.12%  Align node       
1404 MB(9%)00:00:01                Iter   1    2.25%  Align node
1415 MB(9%)00:00:02                Iter   1    3.37%  Align node
1418 MB(9%)00:00:02                Iter   1    4.49%  Align node
1425 MB(9%)00:00:03                Iter   1    5.62%  Align node
1431 MB(9%)00:00:04                Iter   1    6.74%  Align node
1438 MB(9%)00:00:05                Iter   1    7.87%  Align node
1445 MB(9%)00:00:05                Iter   1    8.99%  Align node
1457 MB(9%)00:00:06                Iter   1   10.11%  Align node
1463 MB(9%)00:00:07                Iter   1   11.24%  Align node
1466 MB(9%)00:00:07                Iter   1   12.36%  Align node
1468 MB(9%)00:00:08                Iter   1   13.48%  Align node
1470 MB(9%)00:00:09                Iter   1   14.61%  Align node
1473 MB(9%)00:00:10                Iter   1   15.73%  Align node
1475 MB(9%)00:00:10                Iter   1   16.85%  Align node
1477 MB(9%)00:00:11                Iter   1   17.98%  Align node
1480 MB(9%)00:00:12                Iter   1   19.10%  Align node
1482 MB(9%)00:00:12                Iter   1   20.22%  Align node
1485 MB(9%)00:00:13                Iter   1   21.35%  Align node
1487 MB(9%)00:00:14                Iter   1   22.47%  Align node
1489 MB(9%)00:00:15                Iter   1   23.60%  Align node
1492 MB(9%)00:00:15                Iter   1   24.72%  Align node
1494 MB(9%)00:00:16                Iter   1   25.84%  Align node
1505 MB(9%)00:00:17                Iter   1   26.97%  Align node
1519 MB(9%)00:00:18                Iter   1   28.09%  Align node
1521 MB(9%)00:00:18                Iter   1   29.21%  Align node
1530 MB(9%)00:00:19                Iter   1   30.34%  Align node
1533 MB(9%)00:00:20                Iter   1   31.46%  Align node
1535 MB(9%)00:00:20                Iter   1   32.58%  Align node
1537 MB(9%)00:00:21                Iter   1   33.71%  Align node
1540 MB(9%)00:00:22                Iter   1   34.83%  Align node
1542 MB(9%)00:00:23                Iter   1   35.96%  Align node
1545 MB(9%)00:00:23                Iter   1   37.08%  Align node
1560 MB(10%)00:00:24                Iter   1   38.20%  Align node
1565 MB(10%)00:00:25                Iter   1   39.33%  Align node
1572 MB(10%)00:00:25                Iter   1   40.45%  Align node
1581 MB(10%)00:00:26                Iter   1   41.57%  Align node
1586 MB(10%)00:00:27                Iter   1   42.70%  Align node
1588 MB(10%)00:00:28                Iter   1   43.82%  Align node
1590 MB(10%)00:00:28                Iter   1   44.94%  Align node
1593 MB(10%)00:00:29                Iter   1   46.07%  Align node
1595 MB(10%)00:00:30                Iter   1   47.19%  Align node
1597 MB(10%)00:00:31                Iter   1   48.31%  Align node
1600 MB(10%)00:00:31                Iter   1   49.44%  Align node
1602 MB(10%)00:00:32                Iter   1   50.56%  Align node
1604 MB(10%)00:00:33                Iter   1   51.69%  Align node
1607 MB(10%)00:00:33                Iter   1   52.81%  Align node
1614 MB(10%)00:00:34                Iter   1   53.93%  Align node
1632 MB(10%)00:00:35                Iter   1   55.06%  Align node
1634 MB(10%)00:00:36                Iter   1   56.18%  Align node
1636 MB(10%)00:00:36                Iter   1   57.30%  Align node
1639 MB(10%)00:00:37                Iter   1   58.43%  Align node
1641 MB(10%)00:00:38                Iter   1   59.55%  Align node
1644 MB(10%)00:00:39                Iter   1   60.67%  Align node
1659 MB(10%)00:00:39                Iter   1   61.80%  Align node
1662 MB(10%)00:00:40                Iter   1   62.92%  Align node
1664 MB(10%)00:00:41                Iter   1   64.04%  Align node
1666 MB(10%)00:00:41                Iter   1   65.17%  Align node
1669 MB(10%)00:00:42                Iter   1   66.29%  Align node
1671 MB(10%)00:00:43                Iter   1   67.42%  Align node
1674 MB(10%)00:00:44                Iter   1   68.54%  Align node
1676 MB(10%)00:00:44                Iter   1   69.66%  Align node
1678 MB(10%)00:00:45                Iter   1   70.79%  Align node
1692 MB(10%)00:00:46                Iter   1   71.91%  Align node
1694 MB(10%)00:00:47                Iter   1   73.03%  Align node
1697 MB(10%)00:00:47                Iter   1   74.16%  Align node
1699 MB(10%)00:00:48                Iter   1   75.28%  Align node
1710 MB(10%)00:00:49                Iter   1   76.40%  Align node
1713 MB(10%)00:00:49                Iter   1   77.53%  Align node
1715 MB(10%)00:00:50                Iter   1   78.65%  Align node
1717 MB(10%)00:00:51                Iter   1   79.78%  Align node
1726 MB(11%)00:00:52                Iter   1   80.90%  Align node
1729 MB(11%)00:00:52                Iter   1   82.02%  Align node
1738 MB(11%)00:00:53                Iter   1   83.15%  Align node
1749 MB(11%)00:00:54                Iter   1   84.27%  Align node
1752 MB(11%)00:00:54                Iter   1   85.39%  Align node
1754 MB(11%)00:00:55                Iter   1   86.52%  Align node
1756 MB(11%)00:00:56                Iter   1   87.64%  Align node
1759 MB(11%)00:00:57                Iter   1   88.76%  Align node
1761 MB(11%)00:00:57                Iter   1   89.89%  Align node
1770 MB(11%)00:00:58                Iter   1   91.01%  Align node
1772 MB(11%)00:00:59                Iter   1   92.13%  Align node
1779 MB(11%)00:01:00                Iter   1   93.26%  Align node
1786 MB(11%)00:01:00                Iter   1   94.38%  Align node
1789 MB(11%)00:01:01                Iter   1   95.51%  Align node
1791 MB(11%)00:01:02                Iter   1   96.63%  Align node
1793 MB(11%)00:01:02                Iter   1   97.75%  Align node
1796 MB(11%)00:01:03                Iter   1   98.88%  Align node
1798 MB(11%)00:01:04                Iter   1  100.00%  Align node
1800 MB(11%)00:01:05                Iter   1  100.00%  Align node
1800 MB(11%)00:01:05                Iter   1    1.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1    2.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1    3.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1    4.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1    5.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1    6.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1    7.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1    8.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   10.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   11.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   12.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   13.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   14.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   15.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   16.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   17.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   18.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   20.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   21.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   22.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   23.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   24.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   25.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   26.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   27.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   28.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   30.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   31.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   32.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   33.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   34.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   35.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   36.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   37.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   38.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   40.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   41.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   42.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   43.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   44.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   45.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   46.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   47.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   48.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   50.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   51.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   52.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   53.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   54.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   55.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   56.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   57.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   58.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   60.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   61.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   62.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   63.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   64.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   65.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   66.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   67.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   68.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   70.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   71.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   72.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   73.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   74.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   75.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   76.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   77.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   78.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   80.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   81.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   82.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   83.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   84.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   85.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   86.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   87.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   88.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1   90.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1   91.11%  Root alignment
1801 MB(11%)00:01:05                Iter   1   92.22%  Root alignment
1801 MB(11%)00:01:05                Iter   1   93.33%  Root alignment
1801 MB(11%)00:01:05                Iter   1   94.44%  Root alignment
1801 MB(11%)00:01:05                Iter   1   95.56%  Root alignment
1801 MB(11%)00:01:05                Iter   1   96.67%  Root alignment
1801 MB(11%)00:01:05                Iter   1   97.78%  Root alignment
1801 MB(11%)00:01:05                Iter   1   98.89%  Root alignment
1801 MB(11%)00:01:05                Iter   1  100.00%  Root alignment
1801 MB(11%)00:01:05                Iter   1  100.00%  Root alignment
1801 MB(11%)00:01:05                Iter   2    1.14%  Refine tree   
1803 MB(11%)00:01:06                Iter   2    2.27%  Refine tree
1803 MB(11%)00:01:06                Iter   2    3.41%  Refine tree
1803 MB(11%)00:01:07                Iter   2    4.55%  Refine tree
1803 MB(11%)00:01:08                Iter   2    5.68%  Refine tree
1803 MB(11%)00:01:08                Iter   2    6.82%  Refine tree
1803 MB(11%)00:01:09                Iter   2    7.95%  Refine tree
1803 MB(11%)00:01:10                Iter   2    9.09%  Refine tree
1803 MB(11%)00:01:11                Iter   2   10.23%  Refine tree
1803 MB(11%)00:01:11                Iter   2   11.36%  Refine tree
1803 MB(11%)00:01:12                Iter   2   12.50%  Refine tree
1803 MB(11%)00:01:13                Iter   2   13.64%  Refine tree
1803 MB(11%)00:01:13                Iter   2   14.77%  Refine tree
1803 MB(11%)00:01:14                Iter   2   15.91%  Refine tree
1803 MB(11%)00:01:15                Iter   2   17.05%  Refine tree
1803 MB(11%)00:01:16                Iter   2   18.18%  Refine tree
1803 MB(11%)00:01:16                Iter   2   19.32%  Refine tree
1803 MB(11%)00:01:17                Iter   2   20.45%  Refine tree
1803 MB(11%)00:01:18                Iter   2   21.59%  Refine tree
1803 MB(11%)00:01:19                Iter   2   22.73%  Refine tree
1803 MB(11%)00:01:19                Iter   2   23.86%  Refine tree
1803 MB(11%)00:01:20                Iter   2   25.00%  Refine tree
1803 MB(11%)00:01:21                Iter   2   26.14%  Refine tree
1803 MB(11%)00:01:21                Iter   2   27.27%  Refine tree
1803 MB(11%)00:01:22                Iter   2   28.41%  Refine tree
1803 MB(11%)00:01:23                Iter   2   29.55%  Refine tree
1803 MB(11%)00:01:24                Iter   2   30.68%  Refine tree
1803 MB(11%)00:01:24                Iter   2   31.82%  Refine tree
1803 MB(11%)00:01:25                Iter   2   32.95%  Refine tree
1803 MB(11%)00:01:26                Iter   2   34.09%  Refine tree
1803 MB(11%)00:01:26                Iter   2   35.23%  Refine tree
1803 MB(11%)00:01:27                Iter   2   36.36%  Refine tree
1803 MB(11%)00:01:28                Iter   2   37.50%  Refine tree
1803 MB(11%)00:01:29                Iter   2   38.64%  Refine tree
1803 MB(11%)00:01:29                Iter   2   39.77%  Refine tree
1803 MB(11%)00:01:30                Iter   2   40.91%  Refine tree
1803 MB(11%)00:01:31                Iter   2   42.05%  Refine tree
1803 MB(11%)00:01:32                Iter   2   43.18%  Refine tree
1803 MB(11%)00:01:32                Iter   2   44.32%  Refine tree
1803 MB(11%)00:01:33                Iter   2   45.45%  Refine tree
1803 MB(11%)00:01:34                Iter   2   46.59%  Refine tree
1803 MB(11%)00:01:34                Iter   2   47.73%  Refine tree
1803 MB(11%)00:01:35                Iter   2   48.86%  Refine tree
1803 MB(11%)00:01:36                Iter   2   50.00%  Refine tree
1803 MB(11%)00:01:37                Iter   2   51.14%  Refine tree
1803 MB(11%)00:01:37                Iter   2   52.27%  Refine tree
1803 MB(11%)00:01:38                Iter   2   53.41%  Refine tree
1803 MB(11%)00:01:39                Iter   2   54.55%  Refine tree
1803 MB(11%)00:01:40                Iter   2  100.00%  Refine tree
1803 MB(11%)00:01:40                Iter   2    1.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2    2.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2    3.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2    4.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2    5.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2    6.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2    7.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2    8.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   10.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   11.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   12.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   13.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   14.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   15.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   16.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   17.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   18.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   20.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   21.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   22.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   23.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   24.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   25.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   26.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   27.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   28.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   30.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   31.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   32.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   33.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   34.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   35.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   36.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   37.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   38.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   40.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   41.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   42.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   43.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   44.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   45.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   46.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   47.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   48.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   50.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   51.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   52.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   53.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   54.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   55.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   56.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   57.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   58.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   60.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   61.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   62.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   63.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   64.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   65.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   66.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   67.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   68.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   70.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   71.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   72.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   73.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   74.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   75.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   76.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   77.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   78.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   80.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   81.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   82.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   83.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   84.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   85.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   86.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   87.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   88.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2   90.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2   91.11%  Root alignment
1803 MB(11%)00:01:40                Iter   2   92.22%  Root alignment
1803 MB(11%)00:01:40                Iter   2   93.33%  Root alignment
1803 MB(11%)00:01:40                Iter   2   94.44%  Root alignment
1803 MB(11%)00:01:40                Iter   2   95.56%  Root alignment
1803 MB(11%)00:01:40                Iter   2   96.67%  Root alignment
1803 MB(11%)00:01:40                Iter   2   97.78%  Root alignment
1803 MB(11%)00:01:40                Iter   2   98.89%  Root alignment
1803 MB(11%)00:01:40                Iter   2  100.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2  100.00%  Root alignment
1803 MB(11%)00:01:40                Iter   2  100.00%  Root alignment
1803 MB(11%)00:01:40                Iter   3    1.13%  Refine biparts
1803 MB(11%)00:01:41                Iter   3    1.69%  Refine biparts
1803 MB(11%)00:01:42                Iter   3    2.26%  Refine biparts
1803 MB(11%)00:01:43                Iter   3    2.82%  Refine biparts
1803 MB(11%)00:01:43                Iter   3    3.39%  Refine biparts
1803 MB(11%)00:01:44                Iter   3    3.95%  Refine biparts
1803 MB(11%)00:01:45                Iter   3    4.52%  Refine biparts
1803 MB(11%)00:01:46                Iter   3    5.08%  Refine biparts
1803 MB(11%)00:01:46                Iter   3    5.65%  Refine biparts
1803 MB(11%)00:01:47                Iter   3    6.21%  Refine biparts
1803 MB(11%)00:01:48                Iter   3    6.78%  Refine biparts
1803 MB(11%)00:01:49                Iter   3    7.34%  Refine biparts
1803 MB(11%)00:01:49                Iter   3    7.91%  Refine biparts
1803 MB(11%)00:01:50                Iter   3    8.47%  Refine biparts
1803 MB(11%)00:01:51                Iter   3    9.04%  Refine biparts
1803 MB(11%)00:01:52                Iter   3    9.60%  Refine biparts
1803 MB(11%)00:01:52                Iter   3   10.17%  Refine biparts
1803 MB(11%)00:01:53                Iter   3   10.73%  Refine biparts
1803 MB(11%)00:01:54                Iter   3   11.30%  Refine biparts
1803 MB(11%)00:01:55                Iter   3   11.86%  Refine biparts
1803 MB(11%)00:01:56                Iter   3   12.43%  Refine biparts
1803 MB(11%)00:01:56                Iter   3   12.99%  Refine biparts
1803 MB(11%)00:01:57                Iter   3   13.56%  Refine biparts
1803 MB(11%)00:01:58                Iter   3   14.12%  Refine biparts
1803 MB(11%)00:01:59                Iter   3   14.69%  Refine biparts
1803 MB(11%)00:01:59                Iter   3   15.25%  Refine biparts
1803 MB(11%)00:02:00                Iter   3   15.82%  Refine biparts
1803 MB(11%)00:02:01                Iter   3   16.38%  Refine biparts
1803 MB(11%)00:02:02                Iter   3   16.95%  Refine biparts
1803 MB(11%)00:02:02                Iter   3   17.51%  Refine biparts
1803 MB(11%)00:02:03                Iter   3   18.08%  Refine biparts
1803 MB(11%)00:02:04                Iter   3   18.64%  Refine biparts
1803 MB(11%)00:02:05                Iter   3   19.21%  Refine biparts
1803 MB(11%)00:02:05                Iter   3   19.77%  Refine biparts
1803 MB(11%)00:02:06                Iter   3   20.34%  Refine biparts
1803 MB(11%)00:02:07                Iter   3   20.90%  Refine biparts
1803 MB(11%)00:02:08                Iter   3   21.47%  Refine biparts
1803 MB(11%)00:02:08                Iter   3   22.03%  Refine biparts
1803 MB(11%)00:02:09                Iter   3   22.60%  Refine biparts
1803 MB(11%)00:02:10                Iter   3   23.16%  Refine biparts
1803 MB(11%)00:02:11                Iter   3   23.73%  Refine biparts
1803 MB(11%)00:02:12                Iter   3   24.29%  Refine biparts
1803 MB(11%)00:02:12                Iter   3   24.86%  Refine biparts
1803 MB(11%)00:02:13                Iter   3   25.42%  Refine biparts
1803 MB(11%)00:02:14                Iter   3   25.99%  Refine biparts
1803 MB(11%)00:02:15                Iter   3   26.55%  Refine biparts
1803 MB(11%)00:02:15                Iter   3   27.12%  Refine biparts
1803 MB(11%)00:02:16                Iter   3   27.68%  Refine biparts
1803 MB(11%)00:02:17                Iter   3   28.25%  Refine biparts
1803 MB(11%)00:02:18                Iter   3   28.81%  Refine biparts
1803 MB(11%)00:02:18                Iter   3   29.38%  Refine biparts
1803 MB(11%)00:02:19                Iter   3   29.94%  Refine biparts
1803 MB(11%)00:02:20                Iter   3   30.51%  Refine biparts
1803 MB(11%)00:02:21                Iter   3   31.07%  Refine biparts
1803 MB(11%)00:02:21                Iter   3   31.64%  Refine biparts
1803 MB(11%)00:02:22                Iter   3   32.20%  Refine biparts
1803 MB(11%)00:02:23                Iter   3   32.77%  Refine biparts
1803 MB(11%)00:02:24                Iter   3   33.33%  Refine biparts
1803 MB(11%)00:02:24                Iter   3   33.90%  Refine biparts
1803 MB(11%)00:02:25                Iter   3   34.46%  Refine biparts
1803 MB(11%)00:02:26                Iter   3   35.03%  Refine biparts
1803 MB(11%)00:02:27                Iter   3   35.59%  Refine biparts
1803 MB(11%)00:02:27                Iter   3   36.16%  Refine biparts
1803 MB(11%)00:02:28                Iter   3   36.72%  Refine biparts
1803 MB(11%)00:02:29                Iter   3   37.29%  Refine biparts
1803 MB(11%)00:02:30                Iter   3   37.85%  Refine biparts
1803 MB(11%)00:02:31                Iter   3   38.42%  Refine biparts
1803 MB(11%)00:02:31                Iter   3   38.98%  Refine biparts
1803 MB(11%)00:02:32                Iter   3   39.55%  Refine biparts
1803 MB(11%)00:02:33                Iter   3   40.11%  Refine biparts
1803 MB(11%)00:02:34                Iter   3   40.68%  Refine biparts
1803 MB(11%)00:02:34                Iter   3   41.24%  Refine biparts
1803 MB(11%)00:02:35                Iter   3   41.81%  Refine biparts
1803 MB(11%)00:02:36                Iter   3   42.37%  Refine biparts
1803 MB(11%)00:02:37                Iter   3   42.94%  Refine biparts
1803 MB(11%)00:02:37                Iter   3   43.50%  Refine biparts
1803 MB(11%)00:02:38                Iter   3   44.07%  Refine biparts
1803 MB(11%)00:02:39                Iter   3   44.63%  Refine biparts
1803 MB(11%)00:02:40                Iter   3   45.20%  Refine biparts
1803 MB(11%)00:02:41                Iter   3   45.76%  Refine biparts
1803 MB(11%)00:02:41                Iter   3   46.33%  Refine biparts
1803 MB(11%)00:02:42                Iter   3   46.89%  Refine biparts
1803 MB(11%)00:02:43                Iter   3   47.46%  Refine biparts
1803 MB(11%)00:02:44                Iter   3   48.02%  Refine biparts
1803 MB(11%)00:02:44                Iter   3   48.59%  Refine biparts
1803 MB(11%)00:02:45                Iter   3   49.15%  Refine biparts
1803 MB(11%)00:02:46                Iter   3   49.72%  Refine biparts
1803 MB(11%)00:02:47                Iter   3   50.28%  Refine biparts
1803 MB(11%)00:02:47                Iter   3   50.85%  Refine biparts
1803 MB(11%)00:02:48                Iter   3   51.41%  Refine biparts
1803 MB(11%)00:02:49                Iter   3   51.98%  Refine biparts
1803 MB(11%)00:02:50                Iter   3   52.54%  Refine biparts
1803 MB(11%)00:02:51                Iter   3   53.11%  Refine biparts
1803 MB(11%)00:02:51                Iter   3   53.67%  Refine biparts
1803 MB(11%)00:02:52                Iter   3   54.24%  Refine biparts
1803 MB(11%)00:02:53                Iter   3   54.80%  Refine biparts
1803 MB(11%)00:02:54                Iter   3   55.37%  Refine biparts
1803 MB(11%)00:02:54                Iter   3   55.93%  Refine biparts
1803 MB(11%)00:02:55                Iter   3   56.50%  Refine biparts
1803 MB(11%)00:02:56                Iter   3   57.06%  Refine biparts
1803 MB(11%)00:02:57                Iter   3   57.63%  Refine biparts
1803 MB(11%)00:02:57                Iter   3   58.19%  Refine biparts
1803 MB(11%)00:02:58                Iter   3   58.76%  Refine biparts
1803 MB(11%)00:02:59                Iter   3   59.32%  Refine biparts
1803 MB(11%)00:03:00                Iter   3   59.89%  Refine biparts
1803 MB(11%)00:03:00                Iter   3   60.45%  Refine biparts
1803 MB(11%)00:03:01                Iter   3   61.02%  Refine biparts
1803 MB(11%)00:03:02                Iter   3   61.58%  Refine biparts
1803 MB(11%)00:03:03                Iter   3   62.15%  Refine biparts
1803 MB(11%)00:03:03                Iter   3   62.71%  Refine biparts
1803 MB(11%)00:03:04                Iter   3   63.28%  Refine biparts
1803 MB(11%)00:03:05                Iter   3   63.84%  Refine biparts
1803 MB(11%)00:03:06                Iter   3   64.41%  Refine biparts
1803 MB(11%)00:03:06                Iter   3   64.97%  Refine biparts
1803 MB(11%)00:03:07                Iter   3   65.54%  Refine biparts
1803 MB(11%)00:03:08                Iter   3   66.10%  Refine biparts
1803 MB(11%)00:03:09                Iter   3   66.67%  Refine biparts
1803 MB(11%)00:03:10                Iter   3   67.23%  Refine biparts
1803 MB(11%)00:03:10                Iter   3   67.80%  Refine biparts
1803 MB(11%)00:03:11                Iter   3   68.36%  Refine biparts
1803 MB(11%)00:03:12                Iter   3   68.93%  Refine biparts
1803 MB(11%)00:03:13                Iter   3   69.49%  Refine biparts
1803 MB(11%)00:03:13                Iter   3   70.06%  Refine biparts
1803 MB(11%)00:03:14                Iter   3   70.62%  Refine biparts
1803 MB(11%)00:03:15                Iter   3   71.19%  Refine biparts
1803 MB(11%)00:03:16                Iter   3   71.75%  Refine biparts
1803 MB(11%)00:03:16                Iter   3   72.32%  Refine biparts
1803 MB(11%)00:03:17                Iter   3   72.88%  Refine biparts
1803 MB(11%)00:03:18                Iter   3   73.45%  Refine biparts
1803 MB(11%)00:03:19                Iter   3   74.01%  Refine biparts
1803 MB(11%)00:03:19                Iter   3   74.58%  Refine biparts
1803 MB(11%)00:03:20                Iter   3   75.14%  Refine biparts
1803 MB(11%)00:03:21                Iter   3   75.71%  Refine biparts
1803 MB(11%)00:03:22                Iter   3   76.27%  Refine biparts
1803 MB(11%)00:03:22                Iter   3   76.84%  Refine biparts
1803 MB(11%)00:03:23                Iter   3   77.40%  Refine biparts
1803 MB(11%)00:03:24                Iter   3   77.97%  Refine biparts
1803 MB(11%)00:03:25                Iter   3   78.53%  Refine biparts
1803 MB(11%)00:03:25                Iter   3   79.10%  Refine biparts
1803 MB(11%)00:03:26                Iter   3   79.66%  Refine biparts
1803 MB(11%)00:03:27                Iter   3   80.23%  Refine biparts
1803 MB(11%)00:03:28                Iter   3   80.79%  Refine biparts
1803 MB(11%)00:03:29                Iter   3   81.36%  Refine biparts
1803 MB(11%)00:03:29                Iter   3   81.92%  Refine biparts
1803 MB(11%)00:03:30                Iter   3   82.49%  Refine biparts
1803 MB(11%)00:03:31                Iter   3   83.05%  Refine biparts
1803 MB(11%)00:03:32                Iter   3   83.62%  Refine biparts
1803 MB(11%)00:03:32                Iter   3   84.18%  Refine biparts
1803 MB(11%)00:03:33                Iter   3   84.75%  Refine biparts
1803 MB(11%)00:03:34                Iter   3   85.31%  Refine biparts
1803 MB(11%)00:03:35                Iter   3   85.88%  Refine biparts
1803 MB(11%)00:03:35                Iter   3   86.44%  Refine biparts
1803 MB(11%)00:03:36                Iter   3   87.01%  Refine biparts
1803 MB(11%)00:03:37                Iter   3   87.57%  Refine biparts
1803 MB(11%)00:03:38                Iter   3   88.14%  Refine biparts
1803 MB(11%)00:03:38                Iter   3   88.70%  Refine biparts
1803 MB(11%)00:03:39                Iter   3   89.27%  Refine biparts
1803 MB(11%)00:03:40                Iter   3   89.83%  Refine biparts
1803 MB(11%)00:03:41                Iter   3   90.40%  Refine biparts
1803 MB(11%)00:03:42                Iter   3   90.96%  Refine biparts
1803 MB(11%)00:03:42                Iter   3   91.53%  Refine biparts
1803 MB(11%)00:03:43                Iter   3   92.09%  Refine biparts
1803 MB(11%)00:03:44                Iter   3   92.66%  Refine biparts
1803 MB(11%)00:03:45                Iter   3   93.22%  Refine biparts
1803 MB(11%)00:03:45                Iter   3   93.79%  Refine biparts
1803 MB(11%)00:03:46                Iter   3   94.35%  Refine biparts
1803 MB(11%)00:03:47                Iter   3   94.92%  Refine biparts
1803 MB(11%)00:03:48                Iter   3   95.48%  Refine biparts
1803 MB(11%)00:03:48                Iter   3   96.05%  Refine biparts
1803 MB(11%)00:03:49                Iter   3   96.61%  Refine biparts
1803 MB(11%)00:03:50                Iter   3   97.18%  Refine biparts
1803 MB(11%)00:03:51                Iter   3   97.74%  Refine biparts
1803 MB(11%)00:03:51                Iter   3   98.31%  Refine biparts
1803 MB(11%)00:03:52                Iter   3   98.87%  Refine biparts
1803 MB(11%)00:03:53                Iter   3   99.44%  Refine biparts
1803 MB(11%)00:03:54                Iter   3  100.00%  Refine biparts
1803 MB(11%)00:03:55                Iter   3  100.56%  Refine biparts
1803 MB(11%)00:03:55                Iter   3  100.00%  Refine biparts
1803 MB(11%)00:03:55                Iter   4    1.13%  Refine biparts
1803 MB(11%)00:03:56                Iter   4    1.69%  Refine biparts
1803 MB(11%)00:03:57                Iter   4    2.26%  Refine biparts
1803 MB(11%)00:03:58                Iter   4    2.82%  Refine biparts
1803 MB(11%)00:03:58                Iter   4    3.39%  Refine biparts
1803 MB(11%)00:03:59                Iter   4    3.95%  Refine biparts
1803 MB(11%)00:04:00                Iter   4    4.52%  Refine biparts
1803 MB(11%)00:04:01                Iter   4    5.08%  Refine biparts
1803 MB(11%)00:04:01                Iter   4    5.65%  Refine biparts
1803 MB(11%)00:04:02                Iter   4    6.21%  Refine biparts
1803 MB(11%)00:04:03                Iter   4    6.78%  Refine biparts
1803 MB(11%)00:04:04                Iter   4    7.34%  Refine biparts
1803 MB(11%)00:04:05                Iter   4    7.91%  Refine biparts
1803 MB(11%)00:04:05                Iter   4    8.47%  Refine biparts
1803 MB(11%)00:04:06                Iter   4    9.04%  Refine biparts
1803 MB(11%)00:04:07                Iter   4    9.60%  Refine biparts
1803 MB(11%)00:04:08                Iter   4   10.17%  Refine biparts
1803 MB(11%)00:04:08                Iter   4   10.73%  Refine biparts
1803 MB(11%)00:04:09                Iter   4   11.30%  Refine biparts
1803 MB(11%)00:04:10                Iter   4   11.86%  Refine biparts
1803 MB(11%)00:04:11                Iter   4   12.43%  Refine biparts
1803 MB(11%)00:04:11                Iter   4   12.99%  Refine biparts
1803 MB(11%)00:04:12                Iter   4   13.56%  Refine biparts
1803 MB(11%)00:04:13                Iter   4   14.12%  Refine biparts
1803 MB(11%)00:04:14                Iter   4   14.69%  Refine biparts
1803 MB(11%)00:04:14                Iter   4   15.25%  Refine biparts
1803 MB(11%)00:04:15                Iter   4   15.82%  Refine biparts
1803 MB(11%)00:04:16                Iter   4   16.38%  Refine biparts
1803 MB(11%)00:04:17                Iter   4   16.95%  Refine biparts
1803 MB(11%)00:04:17                Iter   4   17.51%  Refine biparts
1803 MB(11%)00:04:18                Iter   4   18.08%  Refine biparts
1803 MB(11%)00:04:19                Iter   4   18.64%  Refine biparts
1803 MB(11%)00:04:20                Iter   4   19.21%  Refine biparts
1803 MB(11%)00:04:21                Iter   4   19.77%  Refine biparts
1803 MB(11%)00:04:21                Iter   4   20.34%  Refine biparts
1803 MB(11%)00:04:22                Iter   4   20.90%  Refine biparts
1803 MB(11%)00:04:23                Iter   4   21.47%  Refine biparts
1803 MB(11%)00:04:24                Iter   4   22.03%  Refine biparts
1803 MB(11%)00:04:24                Iter   4   22.60%  Refine biparts
1803 MB(11%)00:04:25                Iter   4   23.16%  Refine biparts
1803 MB(11%)00:04:26                Iter   4   23.73%  Refine biparts
1803 MB(11%)00:04:27                Iter   4   24.29%  Refine biparts
1803 MB(11%)00:04:27                Iter   4   24.86%  Refine biparts
1803 MB(11%)00:04:28                Iter   4   25.42%  Refine biparts
1803 MB(11%)00:04:29                Iter   4   25.99%  Refine biparts
1803 MB(11%)00:04:30                Iter   4   26.55%  Refine biparts
1803 MB(11%)00:04:30                Iter   4   27.12%  Refine biparts
1803 MB(11%)00:04:31                Iter   4   27.68%  Refine biparts
1803 MB(11%)00:04:32                Iter   4   28.25%  Refine biparts
1803 MB(11%)00:04:33                Iter   4   28.81%  Refine biparts
1803 MB(11%)00:04:33                Iter   4   29.38%  Refine biparts
1803 MB(11%)00:04:34                Iter   4   29.94%  Refine biparts
1803 MB(11%)00:04:35                Iter   4   30.51%  Refine biparts
1803 MB(11%)00:04:36                Iter   4   31.07%  Refine biparts
1803 MB(11%)00:04:37                Iter   4   31.64%  Refine biparts
1803 MB(11%)00:04:37                Iter   4   32.20%  Refine biparts
1803 MB(11%)00:04:38                Iter   4   32.77%  Refine biparts
1803 MB(11%)00:04:39                Iter   4   33.33%  Refine biparts
1803 MB(11%)00:04:40                Iter   4   33.90%  Refine biparts
1803 MB(11%)00:04:40                Iter   4   34.46%  Refine biparts
1803 MB(11%)00:04:41                Iter   4   35.03%  Refine biparts
1803 MB(11%)00:04:42                Iter   4   35.59%  Refine biparts
1803 MB(11%)00:04:43                Iter   4   36.16%  Refine biparts
1803 MB(11%)00:04:43                Iter   4   36.72%  Refine biparts
1803 MB(11%)00:04:44                Iter   4   37.29%  Refine biparts
1803 MB(11%)00:04:45                Iter   4   37.85%  Refine biparts
1803 MB(11%)00:04:46                Iter   4   38.42%  Refine biparts
1803 MB(11%)00:04:46                Iter   4   38.98%  Refine biparts
1803 MB(11%)00:04:47                Iter   4   39.55%  Refine biparts
1803 MB(11%)00:04:48                Iter   4   40.11%  Refine biparts
1803 MB(11%)00:04:49                Iter   4   40.68%  Refine biparts
1803 MB(11%)00:04:49                Iter   4   41.24%  Refine biparts
1803 MB(11%)00:04:50                Iter   4   41.81%  Refine biparts
1803 MB(11%)00:04:51                Iter   4   42.37%  Refine biparts
1803 MB(11%)00:04:52                Iter   4   42.94%  Refine biparts
1803 MB(11%)00:04:53                Iter   4   43.50%  Refine biparts
1803 MB(11%)00:04:53                Iter   4   44.07%  Refine biparts
1803 MB(11%)00:04:54                Iter   4   44.63%  Refine biparts
1803 MB(11%)00:04:55                Iter   4   45.20%  Refine biparts
1803 MB(11%)00:04:56                Iter   4   45.76%  Refine biparts
1803 MB(11%)00:04:56                Iter   4   46.33%  Refine biparts
1803 MB(11%)00:04:57                Iter   4   46.89%  Refine biparts
1803 MB(11%)00:04:58                Iter   4   47.46%  Refine biparts
1803 MB(11%)00:04:59                Iter   4   48.02%  Refine biparts
1803 MB(11%)00:04:59                Iter   4   48.59%  Refine biparts
1803 MB(11%)00:05:00                Iter   4   49.15%  Refine biparts
1803 MB(11%)00:05:01                Iter   4   49.72%  Refine biparts
1803 MB(11%)00:05:02                Iter   4   50.28%  Refine biparts
1803 MB(11%)00:05:02                Iter   4   50.85%  Refine biparts
1803 MB(11%)00:05:03                Iter   4   51.41%  Refine biparts
1803 MB(11%)00:05:04                Iter   4   51.98%  Refine biparts
1803 MB(11%)00:05:05                Iter   4   52.54%  Refine biparts
1803 MB(11%)00:05:05                Iter   4   53.11%  Refine biparts
1803 MB(11%)00:05:06                Iter   4   53.67%  Refine biparts
1803 MB(11%)00:05:07                Iter   4   54.24%  Refine biparts
1803 MB(11%)00:05:08                Iter   4   54.80%  Refine biparts
1803 MB(11%)00:05:09                Iter   4   55.37%  Refine biparts
1803 MB(11%)00:05:09                Iter   4   55.93%  Refine biparts
1803 MB(11%)00:05:10                Iter   4   56.50%  Refine biparts
1803 MB(11%)00:05:11                Iter   4   57.06%  Refine biparts
1803 MB(11%)00:05:12                Iter   4   57.63%  Refine biparts
1803 MB(11%)00:05:12                Iter   4   58.19%  Refine biparts
1803 MB(11%)00:05:13                Iter   4   58.76%  Refine biparts
1803 MB(11%)00:05:14                Iter   4   59.32%  Refine biparts
1803 MB(11%)00:05:15                Iter   4   59.89%  Refine biparts
1803 MB(11%)00:05:15                Iter   4   60.45%  Refine biparts
1803 MB(11%)00:05:16                Iter   4   61.02%  Refine biparts
1803 MB(11%)00:05:17                Iter   4   61.58%  Refine biparts
1803 MB(11%)00:05:18                Iter   4   62.15%  Refine biparts
1803 MB(11%)00:05:18                Iter   4   62.71%  Refine biparts
1803 MB(11%)00:05:19                Iter   4   63.28%  Refine biparts
1803 MB(11%)00:05:20                Iter   4   63.84%  Refine biparts
1803 MB(11%)00:05:21                Iter   4   64.41%  Refine biparts
1803 MB(11%)00:05:21                Iter   4   64.97%  Refine biparts
1803 MB(11%)00:05:22                Iter   4   65.54%  Refine biparts
1803 MB(11%)00:05:23                Iter   4   66.10%  Refine biparts
1803 MB(11%)00:05:24                Iter   4   66.67%  Refine biparts
1803 MB(11%)00:05:24                Iter   4   67.23%  Refine biparts
1803 MB(11%)00:05:25                Iter   4   67.80%  Refine biparts
1803 MB(11%)00:05:26                Iter   4   68.36%  Refine biparts
1803 MB(11%)00:05:27                Iter   4   68.93%  Refine biparts
1803 MB(11%)00:05:28                Iter   4   69.49%  Refine biparts
1803 MB(11%)00:05:28                Iter   4   70.06%  Refine biparts
1803 MB(11%)00:05:29                Iter   4   70.62%  Refine biparts
1803 MB(11%)00:05:30                Iter   4   71.19%  Refine biparts
1803 MB(11%)00:05:31                Iter   4   71.75%  Refine biparts
1803 MB(11%)00:05:31                Iter   4   72.32%  Refine biparts
1803 MB(11%)00:05:32                Iter   4   72.88%  Refine biparts
1803 MB(11%)00:05:33                Iter   4   73.45%  Refine biparts
1803 MB(11%)00:05:34                Iter   4   74.01%  Refine biparts
1803 MB(11%)00:05:34                Iter   4   74.58%  Refine biparts
1803 MB(11%)00:05:35                Iter   4   75.14%  Refine biparts
1803 MB(11%)00:05:36                Iter   4   75.71%  Refine biparts
1803 MB(11%)00:05:37                Iter   4   76.27%  Refine biparts
1803 MB(11%)00:05:37                Iter   4   76.84%  Refine biparts
1803 MB(11%)00:05:38                Iter   4   77.40%  Refine biparts
1803 MB(11%)00:05:39                Iter   4   77.97%  Refine biparts
1803 MB(11%)00:05:40                Iter   4   78.53%  Refine biparts
1803 MB(11%)00:05:40                Iter   4   79.10%  Refine biparts
1803 MB(11%)00:05:41                Iter   4   79.66%  Refine biparts
1803 MB(11%)00:05:42                Iter   4   80.23%  Refine biparts
1803 MB(11%)00:05:43                Iter   4   80.79%  Refine biparts
1803 MB(11%)00:05:43                Iter   4   81.36%  Refine biparts
1803 MB(11%)00:05:44                Iter   4   81.92%  Refine biparts
1803 MB(11%)00:05:45                Iter   4   82.49%  Refine biparts
1803 MB(11%)00:05:46                Iter   4   83.05%  Refine biparts
1803 MB(11%)00:05:46                Iter   4   83.62%  Refine biparts
1803 MB(11%)00:05:47                Iter   4   84.18%  Refine biparts
1803 MB(11%)00:05:48                Iter   4   84.75%  Refine biparts
1803 MB(11%)00:05:49                Iter   4   85.31%  Refine biparts
1803 MB(11%)00:05:49                Iter   4   85.88%  Refine biparts
1803 MB(11%)00:05:50                Iter   4   86.44%  Refine biparts
1803 MB(11%)00:05:51                Iter   4   87.01%  Refine biparts
1803 MB(11%)00:05:52                Iter   4   87.57%  Refine biparts
1803 MB(11%)00:05:52                Iter   4   88.14%  Refine biparts
1803 MB(11%)00:05:53                Iter   4   88.70%  Refine biparts
1803 MB(11%)00:05:54                Iter   4   89.27%  Refine biparts
1803 MB(11%)00:05:55                Iter   4   89.83%  Refine biparts
1803 MB(11%)00:05:56                Iter   4   90.40%  Refine biparts
1803 MB(11%)00:05:56                Iter   4   90.96%  Refine biparts
1803 MB(11%)00:05:57                Iter   4   91.53%  Refine biparts
1803 MB(11%)00:05:58                Iter   4   92.09%  Refine biparts
1803 MB(11%)00:05:59                Iter   4   92.66%  Refine biparts
1803 MB(11%)00:05:59                Iter   4   93.22%  Refine biparts
1803 MB(11%)00:06:00                Iter   4   93.79%  Refine biparts
1803 MB(11%)00:06:01                Iter   4   94.35%  Refine biparts
1803 MB(11%)00:06:02                Iter   4   94.92%  Refine biparts
1803 MB(11%)00:06:02                Iter   4   95.48%  Refine biparts
1803 MB(11%)00:06:03                Iter   4   96.05%  Refine biparts
1803 MB(11%)00:06:04                Iter   4   96.61%  Refine biparts
1803 MB(11%)00:06:05                Iter   4   97.18%  Refine biparts
1803 MB(11%)00:06:05                Iter   4   97.74%  Refine biparts
1803 MB(11%)00:06:06                Iter   4   98.31%  Refine biparts
1803 MB(11%)00:06:07                Iter   4   98.87%  Refine biparts
1803 MB(11%)00:06:08                Iter   4   99.44%  Refine biparts
1803 MB(11%)00:06:08                Iter   4  100.00%  Refine biparts
1803 MB(11%)00:06:09                Iter   4  100.56%  Refine biparts
1803 MB(11%)00:06:09                Iter   4  100.00%  Refine biparts</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>tipseq_aln <span class="ot">&lt;-</span> <span class="fu">DNAStringSet</span>(tipseq_aln)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="do">## extract accession numbers from tip labels</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>tl <span class="ot">&lt;-</span> tree<span class="sc">$</span>tip.label</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>acc <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">w+</span><span class="sc">\\</span><span class="st">|"</span>, <span class="st">""</span>, tl)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(tl) <span class="ot">&lt;-</span> acc</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="do">## writeXStringSet(tipseq_aln, file = "data/HPV58_aln.fas")</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co">#tipseq_aln &lt;- readDNAStringSet("data/HPV58_aln.fas")</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>tipseq_aln <span class="ot">&lt;-</span> TDbook<span class="sc">::</span>dna_HPV58_aln <span class="sc">%&gt;%</span> </span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>              as.character <span class="sc">%&gt;%</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>              <span class="fu">lapply</span>(., paste0, <span class="at">collapse =</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>              <span class="fu">unlist</span>() <span class="sc">%&gt;%</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>              Biostrings<span class="sc">::</span><span class="fu">DNAStringSet</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="do">## calculate pairwise hamming distances among sequences</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>tipseq_dist <span class="ot">&lt;-</span> pwalign<span class="sc">::</span><span class="fu">stringDist</span>(tipseq_aln, <span class="at">method =</span> <span class="st">"hamming"</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="do">## calculate the percentage of differences</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>tipseq_d <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tipseq_dist) <span class="sc">/</span> <span class="fu">width</span>(tipseq_aln[<span class="dv">1</span>]) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="do">## convert the matrix to a tidy data frame for facet_plot</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>dd <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(tipseq_d)</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>dd<span class="sc">$</span>seq1 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(tipseq_d)</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>td <span class="ot">&lt;-</span> <span class="fu">gather</span>(dd,seq2, dist, <span class="sc">-</span>seq1)</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>td<span class="sc">$</span>seq1 <span class="ot">&lt;-</span> tl[td<span class="sc">$</span>seq1]</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>td<span class="sc">$</span>seq2 <span class="ot">&lt;-</span> tl[td<span class="sc">$</span>seq2]</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> p<span class="sc">$</span>data<span class="sc">$</span>group</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(g) <span class="ot">&lt;-</span> p<span class="sc">$</span>data<span class="sc">$</span>label</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>td<span class="sc">$</span>clade <span class="ot">&lt;-</span> g[td<span class="sc">$</span>seq2] </span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a><span class="do">## visualize the sequence differences using dot plot and line plot</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a><span class="do">## and align the sequence difference plot to the tree using facet_plot</span></span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_facet</span>(<span class="at">panel =</span> <span class="st">"Sequence Distance"</span>, </span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>            <span class="at">data =</span> td, <span class="at">geom =</span> geom_point, <span class="at">alpha =</span> .<span class="dv">6</span>, </span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a>            <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> dist, <span class="at">color =</span> clade, <span class="at">shape =</span> clade)) <span class="sc">+</span></span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_facet</span>(<span class="at">panel =</span> <span class="st">"Sequence Distance"</span>, </span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a>            <span class="at">data =</span> td, <span class="at">geom =</span> geom_path, <span class="at">alpha =</span> .<span class="dv">6</span>, </span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a>            <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x =</span> dist, <span class="at">group =</span> seq2, <span class="at">color =</span> clade)) <span class="sc">+</span> </span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_shape_manual</span>(<span class="at">values =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="at">guide =</span> <span class="cn">FALSE</span>) </span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p2)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="13_ggtree_gallery_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Phylogeny of HPV58 complete genomes with dot and line plots of pairwise nucleotide sequence distances</strong>.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-symbolic-bootstrap" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="sec-symbolic-bootstrap"><span class="header-section-number">10.2</span> Displaying Different Symbolic Points for Bootstrap Values.</h2>
<p>We can cut the bootstrap values into several intervals, <em>e.g.</em>, to indicate whether the clade is of high, moderate, or low support. Then we can use these intervals as categorical variables to set different colors or shapes of symbolic points to indicate the bootstrap values belong to which category (Figure <span class="quarto-unresolved-ref">?fig-bpinterval</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="do">## phytools also have a read.newick function</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>read.newick <span class="ot">&lt;-</span> treeio<span class="sc">::</span>read.newick</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(treeio)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TDbook)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">read.newick</span>(<span class="at">text=</span>text_RMI_tree, <span class="at">node.label =</span> <span class="st">"support"</span>)</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>root <span class="ot">&lt;-</span> <span class="fu">rootnode</span>(tree)  </span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree, <span class="at">color=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">1.5</span>, <span class="at">linetype=</span><span class="dv">1</span>,  <span class="at">right=</span><span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>(<span class="at">size=</span><span class="fl">4.5</span>, <span class="at">hjust =</span> <span class="sc">-</span><span class="fl">0.060</span>, <span class="at">fontface=</span><span class="st">"bold"</span>) <span class="sc">+</span>  <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="fl">0.09</span>) <span class="sc">+</span> </span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point2</span>(<span class="fu">aes</span>(<span class="at">subset=</span><span class="sc">!</span>isTip <span class="sc">&amp;</span> node <span class="sc">!=</span> root, </span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fill=</span><span class="fu">cut</span>(support, <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">700</span>, <span class="dv">900</span>, <span class="dv">1000</span>))), </span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>                    <span class="at">shape=</span><span class="dv">21</span>, <span class="at">size=</span><span class="dv">4</span>) <span class="sc">+</span> </span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_tree</span>(<span class="at">legend.position=</span><span class="fu">c</span>(<span class="fl">0.2</span>, <span class="fl">0.2</span>)) <span class="sc">+</span> </span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"grey"</span>, <span class="st">"black"</span>), <span class="at">guide=</span><span class="st">'legend'</span>, </span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>                    <span class="at">name=</span><span class="st">'Bootstrap Percentage(BP)'</span>, </span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>                    <span class="at">breaks=</span><span class="fu">c</span>(<span class="st">'(900,1e+03]'</span>, <span class="st">'(700,900]'</span>, <span class="st">'(0,700]'</span>), </span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>                    <span class="at">labels=</span><span class="fu">expression</span>(BP<span class="sc">&gt;=</span><span class="dv">90</span>,<span class="dv">70</span> <span class="sc">&lt;=</span> BP <span class="sc">*</span> <span class="st">" &lt; 90"</span>, BP <span class="sc">&lt;</span> <span class="dv">70</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="13_ggtree_gallery_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Partitioning bootstrap values</strong>. Bootstrap values were divided into three categories and this information was used to color circle points.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-phylo-grouping" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="sec-phylo-grouping"><span class="header-section-number">10.3</span> Highlighting Different Groups</h2>
<p>This example reproduces Figure 1 of <span class="citation" data-cites="larsen_identification_2019">(<a href="#ref-larsen_identification_2019" role="doc-biblioref">Larsen et al., 2019</a>)</span>. It used <code>groupOTU()</code> to add grouping information of chicken CTLDcps. The branch line type and color are defined based on this grouping information. Two groups of CTLDcps are highlighted in different background colors using <code>geom_hilight</code> (red for Group II and green for Group V). The avian-specific expansion of Group V with the subgroups of A and B- are labeled using <code>geom_cladelab</code> (Figure <span class="quarto-unresolved-ref">?fig-treeLarsen</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TDbook)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>mytree <span class="ot">&lt;-</span> tree_treenwk_30.<span class="fl">4.19</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Define nodes for coloring later on</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>tiplab <span class="ot">&lt;-</span> mytree<span class="sc">$</span>tip.label</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>cls <span class="ot">&lt;-</span> tiplab[<span class="fu">grep</span>(<span class="st">"^ch"</span>, tiplab)] </span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>labeltree <span class="ot">&lt;-</span> <span class="fu">groupOTU</span>(mytree, cls)</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(labeltree, <span class="fu">aes</span>(<span class="at">color=</span>group, <span class="at">linetype=</span>group), <span class="at">layout=</span><span class="st">"circular"</span>) <span class="sc">+</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#efad29"</span>, <span class="st">"#63bbd4"</span>)) <span class="sc">+</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_nodepoint</span>(<span class="at">color=</span><span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>(<span class="at">size=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">"black"</span>)</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">flip</span>(p, <span class="dv">136</span>, <span class="dv">110</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">flip</span>(<span class="dv">141</span>, <span class="dv">145</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rotate</span>(<span class="dv">141</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rotate</span>(<span class="dv">142</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rotate</span>(<span class="dv">160</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rotate</span>(<span class="dv">164</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rotate</span>(<span class="dv">131</span>)</span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a><span class="do">### Group V and II coloring </span></span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a>           <span class="at">node =</span> <span class="fu">c</span>(<span class="dv">110</span>, <span class="dv">88</span>, <span class="dv">156</span>,<span class="dv">136</span>),</span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"#229f8a"</span>, <span class="st">"#229f8a"</span>, <span class="st">"#229f8a"</span>, <span class="st">"#f9311f"</span>)</span>
<span id="cb39-26"><a href="#cb39-26" aria-hidden="true" tabindex="-1"></a>       )</span>
<span id="cb39-27"><a href="#cb39-27" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> p2 <span class="sc">+</span></span>
<span id="cb39-28"><a href="#cb39-28" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_hilight</span>(</span>
<span id="cb39-29"><a href="#cb39-29" aria-hidden="true" tabindex="-1"></a>          <span class="at">data =</span> dat,</span>
<span id="cb39-30"><a href="#cb39-30" aria-hidden="true" tabindex="-1"></a>          <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb39-31"><a href="#cb39-31" aria-hidden="true" tabindex="-1"></a>              <span class="at">node =</span> node,</span>
<span id="cb39-32"><a href="#cb39-32" aria-hidden="true" tabindex="-1"></a>              <span class="at">fill =</span> <span class="fu">I</span>(fill)</span>
<span id="cb39-33"><a href="#cb39-33" aria-hidden="true" tabindex="-1"></a>          ),</span>
<span id="cb39-34"><a href="#cb39-34" aria-hidden="true" tabindex="-1"></a>          <span class="at">alpha =</span> <span class="fl">0.2</span>,</span>
<span id="cb39-35"><a href="#cb39-35" aria-hidden="true" tabindex="-1"></a>          <span class="at">extendto =</span> <span class="fl">1.4</span></span>
<span id="cb39-36"><a href="#cb39-36" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb39-37"><a href="#cb39-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-38"><a href="#cb39-38" aria-hidden="true" tabindex="-1"></a><span class="do">### Putting on a label on the avian specific expansion </span></span>
<span id="cb39-39"><a href="#cb39-39" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> p3 <span class="sc">+</span></span>
<span id="cb39-40"><a href="#cb39-40" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_cladelab</span>(</span>
<span id="cb39-41"><a href="#cb39-41" aria-hidden="true" tabindex="-1"></a>          <span class="at">node =</span> <span class="dv">113</span>,</span>
<span id="cb39-42"><a href="#cb39-42" aria-hidden="true" tabindex="-1"></a>          <span class="at">label =</span> <span class="st">"Avian-specific expansion"</span>,</span>
<span id="cb39-43"><a href="#cb39-43" aria-hidden="true" tabindex="-1"></a>          <span class="at">align =</span> <span class="cn">TRUE</span>,</span>
<span id="cb39-44"><a href="#cb39-44" aria-hidden="true" tabindex="-1"></a>          <span class="at">angle =</span> <span class="sc">-</span><span class="dv">35</span>,</span>
<span id="cb39-45"><a href="#cb39-45" aria-hidden="true" tabindex="-1"></a>          <span class="at">offset.text =</span> <span class="fl">0.05</span>,</span>
<span id="cb39-46"><a href="#cb39-46" aria-hidden="true" tabindex="-1"></a>          <span class="at">hjust =</span> <span class="st">"center"</span>,</span>
<span id="cb39-47"><a href="#cb39-47" aria-hidden="true" tabindex="-1"></a>          <span class="at">fontsize =</span> <span class="dv">2</span>,</span>
<span id="cb39-48"><a href="#cb39-48" aria-hidden="true" tabindex="-1"></a>          <span class="at">offset =</span> .<span class="dv">2</span>,</span>
<span id="cb39-49"><a href="#cb39-49" aria-hidden="true" tabindex="-1"></a>          <span class="at">barsize =</span> .<span class="dv">2</span></span>
<span id="cb39-50"><a href="#cb39-50" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb39-51"><a href="#cb39-51" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb39-52"><a href="#cb39-52" aria-hidden="true" tabindex="-1"></a><span class="do">### Adding the bootstrap values with subset used to remove all bootstraps &lt; 50  </span></span>
<span id="cb39-53"><a href="#cb39-53" aria-hidden="true" tabindex="-1"></a>p5 <span class="ot">&lt;-</span> p4 <span class="sc">+</span></span>
<span id="cb39-54"><a href="#cb39-54" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_nodelab</span>(</span>
<span id="cb39-55"><a href="#cb39-55" aria-hidden="true" tabindex="-1"></a>          <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb39-56"><a href="#cb39-56" aria-hidden="true" tabindex="-1"></a>              <span class="at">x =</span> branch,</span>
<span id="cb39-57"><a href="#cb39-57" aria-hidden="true" tabindex="-1"></a>              <span class="at">label =</span> label,</span>
<span id="cb39-58"><a href="#cb39-58" aria-hidden="true" tabindex="-1"></a>              <span class="at">subset =</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">as.numeric</span>(label)) <span class="sc">&amp;</span> <span class="fu">as.numeric</span>(label) <span class="sc">&gt;</span> <span class="dv">50</span></span>
<span id="cb39-59"><a href="#cb39-59" aria-hidden="true" tabindex="-1"></a>          ),</span>
<span id="cb39-60"><a href="#cb39-60" aria-hidden="true" tabindex="-1"></a>          <span class="at">size =</span> <span class="dv">2</span>,</span>
<span id="cb39-61"><a href="#cb39-61" aria-hidden="true" tabindex="-1"></a>          <span class="at">color =</span> <span class="st">"black"</span>,</span>
<span id="cb39-62"><a href="#cb39-62" aria-hidden="true" tabindex="-1"></a>          <span class="at">nudge_y =</span> <span class="fl">0.6</span></span>
<span id="cb39-63"><a href="#cb39-63" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb39-64"><a href="#cb39-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-65"><a href="#cb39-65" aria-hidden="true" tabindex="-1"></a><span class="do">### Putting labels on the subgroups </span></span>
<span id="cb39-66"><a href="#cb39-66" aria-hidden="true" tabindex="-1"></a>p6 <span class="ot">&lt;-</span> p5 <span class="sc">+</span></span>
<span id="cb39-67"><a href="#cb39-67" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_cladelab</span>(</span>
<span id="cb39-68"><a href="#cb39-68" aria-hidden="true" tabindex="-1"></a>          <span class="at">data =</span> <span class="fu">data.frame</span>(</span>
<span id="cb39-69"><a href="#cb39-69" aria-hidden="true" tabindex="-1"></a>              <span class="at">node =</span> <span class="fu">c</span>(<span class="dv">114</span>, <span class="dv">121</span>),</span>
<span id="cb39-70"><a href="#cb39-70" aria-hidden="true" tabindex="-1"></a>              <span class="at">name =</span> <span class="fu">c</span>(<span class="st">"Subgroup A"</span>, <span class="st">"Subgroup B"</span>)</span>
<span id="cb39-71"><a href="#cb39-71" aria-hidden="true" tabindex="-1"></a>          ),</span>
<span id="cb39-72"><a href="#cb39-72" aria-hidden="true" tabindex="-1"></a>          <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb39-73"><a href="#cb39-73" aria-hidden="true" tabindex="-1"></a>              <span class="at">node =</span> node,</span>
<span id="cb39-74"><a href="#cb39-74" aria-hidden="true" tabindex="-1"></a>              <span class="at">label =</span> name</span>
<span id="cb39-75"><a href="#cb39-75" aria-hidden="true" tabindex="-1"></a>          ),</span>
<span id="cb39-76"><a href="#cb39-76" aria-hidden="true" tabindex="-1"></a>          <span class="at">align =</span> <span class="cn">TRUE</span>,</span>
<span id="cb39-77"><a href="#cb39-77" aria-hidden="true" tabindex="-1"></a>          <span class="at">offset =</span> .<span class="dv">05</span>,</span>
<span id="cb39-78"><a href="#cb39-78" aria-hidden="true" tabindex="-1"></a>          <span class="at">offset.text =</span> .<span class="dv">03</span>,</span>
<span id="cb39-79"><a href="#cb39-79" aria-hidden="true" tabindex="-1"></a>          <span class="at">hjust =</span> <span class="st">"center"</span>,</span>
<span id="cb39-80"><a href="#cb39-80" aria-hidden="true" tabindex="-1"></a>          <span class="at">barsize =</span> .<span class="dv">2</span>,</span>
<span id="cb39-81"><a href="#cb39-81" aria-hidden="true" tabindex="-1"></a>          <span class="at">fontsize =</span> <span class="dv">2</span>,</span>
<span id="cb39-82"><a href="#cb39-82" aria-hidden="true" tabindex="-1"></a>          <span class="at">angle =</span> <span class="st">"auto"</span>,</span>
<span id="cb39-83"><a href="#cb39-83" aria-hidden="true" tabindex="-1"></a>          <span class="at">horizontal =</span> <span class="cn">FALSE</span></span>
<span id="cb39-84"><a href="#cb39-84" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb39-85"><a href="#cb39-85" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(</span>
<span id="cb39-86"><a href="#cb39-86" aria-hidden="true" tabindex="-1"></a>          <span class="at">legend.position =</span> <span class="st">"none"</span>,</span>
<span id="cb39-87"><a href="#cb39-87" aria-hidden="true" tabindex="-1"></a>          <span class="at">plot.margin =</span> grid<span class="sc">::</span><span class="fu">unit</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">15</span>, <span class="sc">-</span><span class="dv">15</span>, <span class="sc">-</span><span class="dv">15</span>, <span class="sc">-</span><span class="dv">15</span>), <span class="st">"mm"</span>)</span>
<span id="cb39-88"><a href="#cb39-88" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb39-89"><a href="#cb39-89" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(p6)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): NAs introduced by coercion
Warning in FUN(X[[i]], ...): NAs introduced by coercion
Warning in FUN(X[[i]], ...): NAs introduced by coercion
Warning in FUN(X[[i]], ...): NAs introduced by coercion</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Unknown or uninitialised column: `subgroup`.
Unknown or uninitialised column: `subgroup`.
Unknown or uninitialised column: `subgroup`.
Unknown or uninitialised column: `subgroup`.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="13_ggtree_gallery_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Phylogenetic tree of CTLDcps</strong>. Using different background colors, line types and colors, and clade labels to distinguish groups.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-genome-locus" class="level2" data-number="10.4">
<h2 data-number="10.4" class="anchored" data-anchor-id="sec-genome-locus"><span class="header-section-number">10.4</span> Phylogenetic Tree with Genome Locus Structure</h2>
<p>The <code>geom_motif()</code> is defined in <a href="http://bioconductor.org/packages/ggtree">ggtree</a> and it is a wrapper layer of the <code>gggenes::geom_gene_arrow()</code>. The <code>geom_motif()</code> can automatically adjust genomic alignment by selective gene (via the <code>on</code> parameter) and can label genes via the <code>label</code> parameter. In the following example, we use <code>example_genes</code> dataset provided by <a href="https://CRAN.R-project.org/package=gggenes">gggenes</a>. As the dataset only provides genomic coordination of a set of genes, a phylogeny for the genomes needs to be constructed first. We calculate Jaccard similarity based on the ratio of overlapping genes among genomes and correspondingly determine genome distance. The BioNJ algorithm was applied to construct the tree. Then we can use <code>geom_facet()</code> to visualize the tree with the genomic structures (Figure <span class="quarto-unresolved-ref">?fig-gggenes</span>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:Biostrings':

    collapse, intersect, setdiff, setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:GenomeInfoDb':

    intersect</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:XVector':

    slice</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:IRanges':

    collapse, desc, intersect, setdiff, slice, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:S4Vectors':

    first, intersect, rename, setdiff, setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:BiocGenerics':

    combine, intersect, setdiff, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gggenes)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>get_genes <span class="ot">&lt;-</span> <span class="cf">function</span>(data, genome) {</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(data, molecule <span class="sc">==</span> genome) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(gene)</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">unique</span>(example_genes[,<span class="dv">1</span>])</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">length</span>(g)</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> n, <span class="at">ncol =</span> n)</span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(d) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(d) <span class="ot">&lt;-</span> g</span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">lapply</span>(g, get_genes, <span class="at">data =</span> example_genes)</span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>n) {</span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>i) {</span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a>        jaccard_sim <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">intersect</span>(genes[[i]], genes[[j]])) <span class="sc">/</span> </span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a>                       <span class="fu">length</span>(<span class="fu">union</span>(genes[[i]], genes[[j]]))</span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a>        d[j, i] <span class="ot">&lt;-</span> d[i, j] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> jaccard_sim</span>
<span id="cb52-20"><a href="#cb52-20" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb52-21"><a href="#cb52-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-22"><a href="#cb52-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-23"><a href="#cb52-23" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">bionj</span>(d) </span>
<span id="cb52-24"><a href="#cb52-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-25"><a href="#cb52-25" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree, <span class="at">branch.length=</span><span class="st">'none'</span>) <span class="sc">+</span> </span>
<span id="cb52-26"><a href="#cb52-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tiplab</span>() <span class="sc">+</span> <span class="fu">xlim_tree</span>(<span class="fl">5.5</span>) <span class="sc">+</span> </span>
<span id="cb52-27"><a href="#cb52-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_facet</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">xmin =</span> start, <span class="at">xmax =</span> end, <span class="at">fill =</span> gene),</span>
<span id="cb52-28"><a href="#cb52-28" aria-hidden="true" tabindex="-1"></a>               <span class="at">data =</span> example_genes, <span class="at">geom =</span> geom_motif, <span class="at">panel =</span> <span class="st">'Alignment'</span>,</span>
<span id="cb52-29"><a href="#cb52-29" aria-hidden="true" tabindex="-1"></a>               <span class="at">on =</span> <span class="st">'genE'</span>, <span class="at">label =</span> <span class="st">'gene'</span>, <span class="at">align =</span> <span class="st">'left'</span>) <span class="sc">+</span></span>
<span id="cb52-30"><a href="#cb52-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set3"</span>) <span class="sc">+</span> </span>
<span id="cb52-31"><a href="#cb52-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">expand=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb52-32"><a href="#cb52-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">strip.text=</span><span class="fu">element_blank</span>(),</span>
<span id="cb52-33"><a href="#cb52-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.spacing=</span><span class="fu">unit</span>(<span class="dv">0</span>, <span class="st">'cm'</span>))</span>
<span id="cb52-34"><a href="#cb52-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-35"><a href="#cb52-35" aria-hidden="true" tabindex="-1"></a><span class="fu">facet_widths</span>(p, <span class="at">widths=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in x[i] &lt;- value: number of items to replace is not a multiple of
replacement length</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="13_ggtree_gallery_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Genomic features with a phylogenetic tree.</strong></figcaption>
</figure>
</div>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-chen_ancient_2017" class="csl-entry" role="listitem">
Chen, Z., Ho, W. C. S., Boon, S. S., Law, P. T. Y., Chan, M. C. W., DeSalle, R., Burk, R. D., &amp; Chan, P. K. S. (2017). Ancient evolution and dispersion of human papillomavirus 58 variants. <em>Journal of Virology</em>, <em>91</em>(21), e01285–17. <a href="https://doi.org/10.1128/JVI.01285-17">https://doi.org/10.1128/JVI.01285-17</a>
</div>
<div id="ref-larsen_identification_2019" class="csl-entry" role="listitem">
Larsen, F. T., Bed’Hom, B., Guldbrandtsen, B., &amp; Dalgaard, T. S. (2019). Identification and tissue-expression profiling of novel chicken c-type lectin-like domain containing proteins as potential targets for carbohydrate-based vaccine strategies. <em>Molecular Immunology</em>, <em>114</em>, 216–225. <a href="https://doi.org/10.1016/j.molimm.2019.07.022">https://doi.org/10.1016/j.molimm.2019.07.022</a>
</div>
<div id="ref-yu_two_2018" class="csl-entry" role="listitem">
Yu, G., Lam, T. T.-Y., Zhu, H., &amp; Guan, Y. (2018). Two methods for mapping and visualizing associated data on phylogeny using ggtree. <em>Molecular Biology and Evolution</em>, <em>35</em>(12), 3041–3043. <a href="https://doi.org/10.1093/molbev/msy194">https://doi.org/10.1093/molbev/msy194</a>
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./12_ggtree_utilities.html" class="pagination-link" aria-label="(PART\*) Part IV: Miscellaneous topics">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">(PART*) Part IV: Miscellaneous topics</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./A-app-faq.html" class="pagination-link" aria-label="(APPENDIX) Appendix">
        <span class="nav-page-text">(APPENDIX) Appendix</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/YuLab-SMU/treedata-book/edit/main/13_ggtree_gallery.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/YuLab-SMU/treedata-book/blob/main/13_ggtree_gallery.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div></div></footer></body></html>